<!DOCTYPE html>
<html lang="en"><head>
<script src="popgen2022_files/libs/clipboard/clipboard.min.js"></script>
<script src="popgen2022_files/libs/quarto-html/tabby.min.js"></script>
<script src="popgen2022_files/libs/quarto-html/popper.min.js"></script>
<script src="popgen2022_files/libs/quarto-html/tippy.umd.min.js"></script>
<link href="popgen2022_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="popgen2022_files/libs/quarto-html/quarto-html.min.css" rel="stylesheet" data-mode="light">
<link href="popgen2022_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles"><meta charset="utf-8">
  <meta name="generator" content="quarto-1.0.38">

  <meta name="author" content="Martin Petr (mp@bodkan.net)">
  <title>Simulations in population genetics</title>
  <meta name="apple-mobile-web-app-capable" content="yes">
  <meta name="apple-mobile-web-app-status-bar-style" content="black-translucent">
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0, user-scalable=no, minimal-ui">
  <link rel="stylesheet" href="popgen2022_files/libs/revealjs/dist/reset.css">
  <link rel="stylesheet" href="popgen2022_files/libs/revealjs/dist/reveal.css">
  <style>
    code{white-space: pre-wrap;}
    span.smallcaps{font-variant: small-caps;}
    span.underline{text-decoration: underline;}
    div.column{display: inline-block; vertical-align: top; width: 50%;}
    div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
    ul.task-list{list-style: none;}
    pre > code.sourceCode { white-space: pre; position: relative; }
    pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
    pre > code.sourceCode > span:empty { height: 1.2em; }
    .sourceCode { overflow: visible; }
    code.sourceCode > span { color: inherit; text-decoration: inherit; }
    div.sourceCode { margin: 1em 0; }
    pre.sourceCode { margin: 0; }
    @media screen {
    div.sourceCode { overflow: auto; }
    }
    @media print {
    pre > code.sourceCode { white-space: pre-wrap; }
    pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
    }
    pre.numberSource code
      { counter-reset: source-line 0; }
    pre.numberSource code > span
      { position: relative; left: -4em; counter-increment: source-line; }
    pre.numberSource code > span > a:first-child::before
      { content: counter(source-line);
        position: relative; left: -1em; text-align: right; vertical-align: baseline;
        border: none; display: inline-block;
        -webkit-touch-callout: none; -webkit-user-select: none;
        -khtml-user-select: none; -moz-user-select: none;
        -ms-user-select: none; user-select: none;
        padding: 0 4px; width: 4em;
        color: #aaaaaa;
      }
    pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
    div.sourceCode
      { color: #003b4f; background-color: #f1f3f5; }
    @media screen {
    pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
    }
    code span { color: #003b4f; } /* Normal */
    code span.al { color: #ad0000; } /* Alert */
    code span.an { color: #5e5e5e; } /* Annotation */
    code span.at { color: #657422; } /* Attribute */
    code span.bn { color: #ad0000; } /* BaseN */
    code span.bu { } /* BuiltIn */
    code span.cf { color: #003b4f; } /* ControlFlow */
    code span.ch { color: #20794d; } /* Char */
    code span.cn { color: #8f5902; } /* Constant */
    code span.co { color: #5e5e5e; } /* Comment */
    code span.cv { color: #5e5e5e; font-style: italic; } /* CommentVar */
    code span.do { color: #5e5e5e; font-style: italic; } /* Documentation */
    code span.dt { color: #ad0000; } /* DataType */
    code span.dv { color: #ad0000; } /* DecVal */
    code span.er { color: #ad0000; } /* Error */
    code span.ex { } /* Extension */
    code span.fl { color: #ad0000; } /* Float */
    code span.fu { color: #4758ab; } /* Function */
    code span.im { color: #00769e; } /* Import */
    code span.in { color: #5e5e5e; } /* Information */
    code span.kw { color: #003b4f; } /* Keyword */
    code span.op { color: #5e5e5e; } /* Operator */
    code span.ot { color: #003b4f; } /* Other */
    code span.pp { color: #ad0000; } /* Preprocessor */
    code span.sc { color: #5e5e5e; } /* SpecialChar */
    code span.ss { color: #20794d; } /* SpecialString */
    code span.st { color: #20794d; } /* String */
    code span.va { color: #111111; } /* Variable */
    code span.vs { color: #20794d; } /* VerbatimString */
    code span.wa { color: #5e5e5e; font-style: italic; } /* Warning */
  </style>
  <link rel="stylesheet" href="popgen2022_files/libs/revealjs/dist/theme/quarto.css" id="theme">
  <link href="popgen2022_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.css" rel="stylesheet">
  <link href="popgen2022_files/libs/revealjs/plugin/reveal-menu/menu.css" rel="stylesheet">
  <link href="popgen2022_files/libs/revealjs/plugin/reveal-menu/quarto-menu.css" rel="stylesheet">
  <link href="popgen2022_files/libs/revealjs/plugin/quarto-support/footer.css" rel="stylesheet">
  <style type="text/css">

  .callout {
    margin-top: 1em;
    margin-bottom: 1em;  
    border-radius: .25rem;
  }

  .callout.callout-style-simple { 
    padding: 0em 0.5em;
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
    display: flex;
  }

  .callout.callout-style-default {
    border-left: solid #acacac .3rem;
    border-right: solid 1px silver;
    border-top: solid 1px silver;
    border-bottom: solid 1px silver;
  }

  .callout .callout-body-container {
    flex-grow: 1;
  }

  .callout.callout-style-simple .callout-body {
    font-size: 1rem;
    font-weight: 400;
  }

  .callout.callout-style-default .callout-body {
    font-size: 0.9rem;
    font-weight: 400;
  }

  .callout.callout-captioned.callout-style-simple .callout-body {
    margin-top: 0.2em;
  }

  .callout:not(.callout-captioned) .callout-body {
      display: flex;
  }

  .callout:not(.no-icon).callout-captioned.callout-style-simple .callout-content {
    padding-left: 1.6em;
  }

  .callout.callout-captioned .callout-header {
    padding-top: 0.2em;
    margin-bottom: -0.2em;
  }

  .callout.callout-captioned .callout-caption  p {
    margin-top: 0.5em;
    margin-bottom: 0.5em;
  }
    
  .callout.callout-captioned.callout-style-simple .callout-content  p {
    margin-top: 0;
  }

  .callout.callout-captioned.callout-style-default .callout-content  p {
    margin-top: 0.7em;
  }

  .callout.callout-style-simple div.callout-caption {
    border-bottom: none;
    font-size: .9rem;
    font-weight: 600;
    opacity: 75%;
  }

  .callout.callout-style-default  div.callout-caption {
    border-bottom: none;
    font-weight: 600;
    opacity: 85%;
    font-size: 0.9rem;
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-default div.callout-content {
    padding-left: 0.5em;
    padding-right: 0.5em;
  }

  .callout.callout-style-simple .callout-icon::before {
    height: 1rem;
    width: 1rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 1rem 1rem;
  }

  .callout.callout-style-default .callout-icon::before {
    height: 0.9rem;
    width: 0.9rem;
    display: inline-block;
    content: "";
    background-repeat: no-repeat;
    background-size: 0.9rem 0.9rem;
  }

  .callout-caption {
    display: flex
  }
    
  .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  .callout.no-icon::before {
    display: none !important;
  }

  .callout.callout-captioned .callout-body > .callout-content > :last-child {
    margin-bottom: 0.5rem;
  }

  .callout.callout-captioned .callout-icon::before {
    margin-top: .5rem;
    padding-right: .5rem;
  }

  .callout:not(.callout-captioned) .callout-icon::before {
    margin-top: 1rem;
    padding-right: .5rem;
  }

  /* Callout Types */

  div.callout-note {
    border-left-color: #4582ec !important;
  }

  div.callout-note .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEU0lEQVRYCcVXTWhcVRQ+586kSUMMxkyaElstCto2SIhitS5Ek8xUKV2poatCcVHtUlFQk8mbaaziwpWgglJwVaquitBOfhQXFlqlzSJpFSpIYyXNjBNiTCck7x2/8/LeNDOZxDuEkgOXe++553zfefee+/OYLOXFk3+1LLrRdiO81yNqZ6K9cG0P3MeFaMIQjXssE8Z1JzLO9ls20MBZX7oG8w9GxB0goaPrW5aNMp1yOZIa7Wv6o2ykpLtmAPs/vrG14Z+6d4jpbSKuhdcSyq9wGMPXjonwmESXrriLzFGOdDBLB8Y6MNYBu0dRokSygMA/mrun8MGFN3behm6VVAwg4WR3i6FvYK1T7MHo9BK7ydH+1uurECoouk5MPRyVSBrBHMYwVobG2aOXM07sWrn5qgB60rc6mcwIDJtQrnrEr44kmy+UO9r0u9O5/YbkS9juQckLed3DyW2XV/qWBBB3ptvI8EUY3I9p/67OW+g967TNr3Sotn3IuVlfMLVnsBwH4fsnebJvyGm5GeIUA3jljERmrv49SizPYuq+z7c2H/jlGC+Ghhupn/hcapqmcudB9jwJ/3jvnvu6vu5lVzF1fXyZuZZ7U8nRmVzytvT+H3kilYvH09mLWrQdwFSsFEsxFVs5fK7A0g8gMZjbif4ACpKbjv7gNGaD8bUrlk8x+KRflttr22JEMRUbTUwwDQScyzPgedQHZT0xnx7ujw2jfVfExwYHwOsDTjLdJ2ebmeQIlJ7neo41s/DrsL3kl+W2lWvAga0tR3zueGr6GL78M3ifH0rGXrBC2aAR8uYcIA5gwV8zIE8onoh8u0Fca/ciF7j1uOzEnqcIm59sEXoGc0+z6+H45V1CvAvHcD7THztu669cnp+L0okAeIc6zjbM/24LgGM1gZk7jnRu1aQWoU9sfUOuhrmtaPIO3YY1KLLWZaEO5TKUbMY5zx8W9UJ6elpLwKXbsaZ4EFl7B4bMtDv0iRipKoDQT2sNQI9b1utXFdYisi+wzZ/ri/1m7QfDgEuvgUUEIJPq3DhX/5DWNqIXDOweC2wvIR90Oq3lDpdMIgD2r0dXvGdsEW5H6x6HLRJYU7C69VefO1x8Gde1ZFSJLfWS1jbCnhtOPxmpfv2LXOA2Xk2tvnwKKPFuZ/oRmwBwqRQDcKNeVQkYcOjtWVBuM/JuYw5b6isojIkYxyYAFn5K7ZBF10fea52y8QltAg6jnMqNHFBmGkQ1j+U43HMi2xMar1Nv0zGsf1s8nUsmUtPOOrbFIR8bHFDMB5zL13Gmr/kGlCkUzedTzzmzsaJXhYawnA3UmARpiYj5ooJZiUoxFRtK3X6pgNPv+IZVPcnwbOl6f+aBaO1CNvPW9n9LmCp01nuSaTRF2YxHqZ8DYQT6WsXT+RD6eUztwYLZ8rM+rcPxamv1VQzFUkzFXvkiVrySGQgJNvXHJAxiU3/NwiC03rSf05VBaPtu/Z7/B8Yn/w7eguloAAAAAElFTkSuQmCC');
  }

  div.callout-note.callout-style-default .callout-caption {
    background-color: #dae6fb
  }

  div.callout-important {
    border-left-color: #d9534f !important;
  }

  div.callout-important .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAEKklEQVRYCcVXTWhcVRS+575MJym48A+hSRFr00ySRQhURRfd2HYjk2SSTokuBCkU2o0LoSKKraKIBTcuFCoidGFD08nkBzdREbpQ1EDNIv8qSGMFUboImMSZd4/f9zJv8ibJMC8xJQfO3HPPPef7zrvvvnvviIkpC9nsw0UttFunbUhpFzFtarSd6WJkStVMw5xyVqYTvkwfzuf/5FgtkVoB0729j1rjXwThS7Vio+Mo6DNnvLfahoZ+i/o32lULuJ3NNiz7q6+pyAUkJaFF6JwaM2lUJlV0MlnQn5aTRbEu0SEqHUa0A4AdiGuB1kFXRfVyg5d87+Dg4DL6m2TLAub60ilj7A1Ec4odSAc8X95sHh7+ZRPCFo6Fnp7HfU/fBng/hi10CjCnWnJjsxvDNxWw0NfV6Rv5GgP3I3jGWXumdTD/3cbEOP2ZbOZp69yniG3FQ9z1jD7bnBu9Fc2tKGC2q+uAJOQHBDRiZX1x36o7fWBs7J9ownbtO+n0/qWkvW7UPIfc37WgT6ZGR++EOJyeQDSb9UB+DZ1G6DdLDzyS+b/kBCYGsYgJbSQHuThGKRcw5xdeQf8YdNHsc6ePXrlSYMBuSIAFTGAtQo+VuALo4BX83N190NWZWbynBjhOHsmNfFWLeL6v+ynsA58zDvvAC8j5PkbOcXCMg2PZFk3q8MjI7WAG/Dp9AwP7jdGBOOQkAvlFUB+irtm16I1Zw9YBcpGTGXYmk3kQIC/Cds55l+iMI3jqhjAuaoe+am2Jw5GT3Nbz3CkE12NavmzN5+erJW7046n/CH1RO/RVa8lBLozXk9uqykkGAyRXLWlLv5jyp4RFsG5vGVzpDLnIjTWgnRy2Rr+tDKvRc7Y8AyZq10jj8DqXdnIRNtFZb+t/ZRtXcDiVnzpqx8mPcDWxgARUqx0W1QB9MeUZiNrV4qP+Ehc+BpNgATsTX8ozYKL2NtFYAHc84fG7ndxUPr+AR/iQSns7uSUufAymwDOb2+NjK27lEFocm/EE2WpyIy/Hi66MWuMKJn8RvxIcj87IM5Vh9663ziW36kR0HNenXuxmfaD8JC7tfKbrhFr7LiZCrMjrzTeGx+PmkosrkNzW94ObzwocJ7A1HokLolY+AvkTiD/q1H0cN48c5EL8Crkttsa/AXQVDmutfyku0E7jShx49XqV3MFK8IryDhYVbj7Sj2P2eBxwcXoe8T8idsKKPRcnZw1b+slFTubwUwhktrfnAt7J++jwQtLZcm3sr9LQrjRzz6cfMv9aLvgmnAGvpoaGLxM4mAEaLV7iAzQ3oU0IvD5x9ix3yF2RAAuYAOO2f7PEFWCXZ4C9Pb2UsgDeVnFSpbFK7/IWu7TPTvBqzbGdCHOJQSxiEjt6IyZmxQyEJHv6xyQsYk//moVFsN2zP6fRImjfq7/n/wFDguUQFNEwugAAAABJRU5ErkJggg==');
  }

  div.callout-important.callout-style-default .callout-caption {
    background-color: #f7dddc
  }

  div.callout-warning {
    border-left-color: #f0ad4e !important;
  }

  div.callout-warning .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAAETklEQVRYCeVWW2gcVRg+58yaTUnizqbipZeX4uWhBEniBaoUX1Ioze52t7sRq6APio9V9MEaoWlVsFasRq0gltaAPuxms8lu0gcviE/FFOstVbSIxgcv6SU7EZqmdc7v9+9mJtNks51NTUH84ed889/PP+cmxP+d5FIbMJmNbpREu4WUkiTtCicKny0l1pIKmBzovF2S+hIJHX8iEu3hZJ5lNZGqyRrGSIQpq15AzF28jgpeY6yk6GVdrfFqdrD6Iw+QlB8g0YS2g7dyQmXM/IDhBhT0UCiRf59lfqmmDvzRt6kByV/m4JjtzuaujMUM2c5Z2d6JdKrRb3K2q6mA+oYVz8JnDdKPmmNthzkAk/lN63sYPgevrguc72aZX/L9C6x09GYyxBgCX4NlvyGUHOKELlm5rXeR1kchuChJt4SSwyddZRXgvwMGvYo4QSlk3/zkHD8UHxwVJA6zjZZqP8v8kK8OWLnIZtLyCAJagYC4rTGW/9Pqj92N/c+LUaAj27movwbi19tk/whRCIE7Q9vyI6yvRpftAKVTdUjOW40X3h5OXsKCdmFcx0xlLJoSuQngnrJe7Kcjm4OMq9FlC7CMmScQANuNvjfP3PjGXDBaUQmbp296S5L4DrpbrHN1T87ZVEZVCzg1FF0Ft+dKrlLukI+/c9ENo+TvlTDbYFvuKPtQ9+l052rXrgKoWkDAFnvh0wTOmYn8R5f4k/jN/fZiCM1tQx9jQQ4ANhqG4hiL0qIFTGViG9DKB7GYzgubnpofgYRwO+DFjh0Zin2m4b/97EDkXkc+f6xYAPX0KK2I/7fUQuwzuwo/L3AkcjugPNixC8cHf0FyPjWlItmLxWw4Ou9YsQCr5fijMGoD/zpdRy95HRysyXA74MWOnscpO4j2y3HAVisw85hX5+AFBRSHt4ShfLFkIMXTqyKFc46xdzQM6XbAi702a7sy04J0+feReMFKp5q9esYLCqAZYw/k14E/xcLLsFElaornTuJB0svMuJINy8xkIYuL+xPAlWRceH6+HX7THJ0djLUom46zREu7tTkxwmf/FdOZ/sh6Q8qvEAiHpm4PJ4a/doJe0gH1t+aHRgCzOvBvJedEK5OFE5jpm4AGP2a8Dxe3gGJ/pAutug9Gp6he92CsSsWBaEcxGx0FHytmIpuqGkOpldqNYQK8cSoXvd+xLxXADw0kf6UkJNFtdo5MOgaLjiQOQHcn+A6h5NuL2s0qsC2LOM75PcF3yr5STuBSAcGG+meA14K/CI21HcS4LBT6tv0QAh8Dr5l93AhZzG5ZJ4VxAqdZUEl9z7WJ4aN+svMvwHHL21UKTd1mqvChH7/Za5xzXBBKrUcB0TQ+Ulgkfbi/H/YT5EptrGzsEK7tR1B7ln9BBwckYfMiuSqklSznIuoIIOM42MQO+QnduCoFCI0bpkzjCjddHPN/F+2Yu+sd9bKNpVwHhbS3LluK/0zgfwD0xYI5dXuzlQAAAABJRU5ErkJggg==');
  }

  div.callout-warning.callout-style-default .callout-caption {
    background-color: #fcefdc
  }

  div.callout-tip {
    border-left-color: #02b875 !important;
  }

  div.callout-tip .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAADr0lEQVRYCe1XTWgTQRj9ZjZV8a9SPIkKgj8I1bMHsUWrqYLVg4Ue6v9BwZOxSYsIerFao7UiUryIqJcqgtpimhbBXoSCVxUFe9CTiogUrUp2Pt+3aUI2u5vdNh4dmMzOzHvvezuz8xNFM0mjnbXaNu1MvFWRXkXEyE6aYOYJpdW4IXuA4r0fo8qqSMDBU0v1HJUgVieAXxzCsdE/YJTdFcVIZQNMyhruOMJKXYFoLfIfIvVIMWdsrd+Rpd86ZmyzzjJmLStqRn0v8lzkb4rVIXvnpScOJuAn2ACC65FkPzEdEy4TPWRLJ2h7z4cArXzzaOdKlbOvKKX25Wl00jSnrwVxAg3o4dRxhO13RBSdNvH0xSARv3adTXbBdTf64IWO2vH0LT+cv4GR1DJt+DUItaQogeBX/chhbTBxEiZ6gftlDNXTrvT7co4ub5A6gp9HIcHvzTa46OS5fBeP87Qm0fQkr4FsYgVQ7Qg+ZayaDg9jhg1GkWj8RG6lkeSacrrHgDaxdoBiZPg+NXV/KifMuB6//JmYH4CntVEHy/keA6x4h4CU5oFy8GzrBS18cLJMXcljAKB6INjWsRcuZBWVaS3GDrqB7rdapVIeA+isQ57Eev9eCqzqOa81CY05VLd6SamW2wA2H3SiTbnbSxmzfp7WtKZkqy4mdyAlGx7ennghYf8voqp9cLSgKdqNfa6RdRsAAkPwRuJZNbpByn+RrJi1RXTwdi8RQF6ymDwGMAtZ6TVE+4uoKh+MYkcLsT0Hk8eAienbiGdjJHZTpmNjlbFJNKDVAp2fJlYju6IreQxQ08UJDNYdoLSl6AadO+fFuCQqVMB1NJwPm69T04Wv5WhfcWyfXQB+wXRs1pt+nCknRa0LVzSA/2B+a9+zQJadb7IyyV24YAxKp2Jqs3emZTuNnKxsah+uabKbMk7CbTgJx/zIgQYErIeTKRQ9yD9wxVof5YolPHqaWo7TD6tJlh7jQnK5z2n3+fGdggIOx2kaa2YI9QWarc5Ce1ipNWMKeSG4DysFF52KBmTNMmn5HqCFkwy34rDg05gDwgH3bBi+sgFhN/e8QvRn8kbamCOhgrZ9GJhFDgfcMHzFb6BAtjKpFhzTjwv1KCVuxHvCbsSiEz4CANnj84cwHdFXAbAOJ4LTSAawGWFn5tDhLMYz6nWeU2wJfIhmIJBefcd/A5FWQWGgrWzyORZ3Q6HuV+Jf0Bj+BTX69fm1zWgK7By1YTXchFDORywnfQ7GpzOo6S+qECrsx2ifVQAAAABJRU5ErkJggg==');
  }

  div.callout-tip.callout-style-default .callout-caption {
    background-color: #ccf1e3
  }

  div.callout-caution {
    border-left-color: #fd7e14 !important;
  }

  div.callout-caution .callout-icon::before {
    background-image: url('data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAACAAAAAgCAYAAABzenr0AAAAAXNSR0IArs4c6QAAAERlWElmTU0AKgAAAAgAAYdpAAQAAAABAAAAGgAAAAAAA6ABAAMAAAABAAEAAKACAAQAAAABAAAAIKADAAQAAAABAAAAIAAAAACshmLzAAACV0lEQVRYCdVWzWoUQRCuqp2ICBLJXgITZL1EfQDBW/bkzUMUD7klD+ATSHBEfAIfQO+iXsWDxJsHL96EHAwhgzlkg8nBg25XWb0zIb0zs9muYYWkoKeru+vn664fBqElyZNuyh167NXJ8Ut8McjbmEraKHkd7uAnAFku+VWdb3reSmRV8PKSLfZ0Gjn3a6Xlcq9YGb6tADjn+lUfTXtVmaZ1KwBIvFI11rRXlWlatwIAAv2asaa9mlB9wwygiDX26qaw1yYPzFXg2N1GgG0FMF8Oj+VIx7E/03lHx8UhvYyNZLN7BwSPgekXXLribw7w5/c8EF+DBK5idvDVYtEEwMeYefjjLAdEyQ3M9nfOkgnPTEkYU+sxMq0BxNR6jExrAI31H1rzvLEfRIdgcv1XEdj6QTQAS2wtstEALLG1yEZ3QhH6oDX7ExBSFEkFINXH98NTrme5IOaaA7kIfiu2L8A3qhH9zRbukdCqdsA98TdElyeMe5BI8Rs2xHRIsoTSSVFfCFCWGPn9XHb4cdobRIWABNf0add9jakDjQJpJ1bTXOJXnnRXHRf+dNL1ZV1MBRCXhMbaHqGI1JkKIL7+i8uffuP6wVQAzO7+qVEbF6NbS0LJureYcWXUUhH66nLR5rYmva+2tjRFtojkM2aD76HEGAD3tPtKM309FJg5j/K682ywcWJ3PASCcycH/22u+Bh7Aa0ehM2Fu4z0SAE81HF9RkB21c5bEn4Dzw+/qNOyXr3DCTQDMBOdhi4nAgiFDGCinIa2owCEChUwD8qzd03PG+qdW/4fDzjUMcE1ZpIAAAAASUVORK5CYII=');
  }

  div.callout-caution.callout-style-default .callout-caption {
    background-color: #ffe5d0
  }

  </style>
  <style type="text/css">
    .reveal div.sourceCode {
      margin: 0;
      overflow: auto;
    }
    .reveal div.hanging-indent {
      margin-left: 1em;
      text-indent: -1em;
    }
    .reveal .slide:not(.center) {
      height: 100%;
    }
    .reveal .slide.scrollable {
      overflow-y: auto;
    }
    .reveal .footnotes {
      height: 100%;
      overflow-y: auto;
    }
    .reveal .slide .absolute {
      position: absolute;
      display: block;
    }
    .reveal .footnotes ol {
      counter-reset: ol;
      list-style-type: none; 
      margin-left: 0;
    }
    .reveal .footnotes ol li:before {
      counter-increment: ol;
      content: counter(ol) ". "; 
    }
    .reveal .footnotes ol li > p:first-child {
      display: inline-block;
    }
    .reveal .slide ul,
    .reveal .slide ol {
      margin-bottom: 0.5em;
    }
    .reveal .slide ul li,
    .reveal .slide ol li {
      margin-top: 0.4em;
      margin-bottom: 0.2em;
    }
    .reveal .slide ul[role="tablist"] li {
      margin-bottom: 0;
    }
    .reveal .slide ul li > *:first-child,
    .reveal .slide ol li > *:first-child {
      margin-block-start: 0;
    }
    .reveal .slide ul li > *:last-child,
    .reveal .slide ol li > *:last-child {
      margin-block-end: 0;
    }
    .reveal .slide .columns:nth-child(3) {
      margin-block-start: 0.8em;
    }
    .reveal blockquote {
      box-shadow: none;
    }
    .reveal .tippy-content>* {
      margin-top: 0.2em;
      margin-bottom: 0.7em;
    }
    .reveal .tippy-content>*:last-child {
      margin-bottom: 0.2em;
    }
    .reveal .slide > img.stretch.quarto-figure-center,
    .reveal .slide > img.r-stretch.quarto-figure-center {
      display: block;
      margin-left: auto;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-left,
    .reveal .slide > img.r-stretch.quarto-figure-left  {
      display: block;
      margin-left: 0;
      margin-right: auto; 
    }
    .reveal .slide > img.stretch.quarto-figure-right,
    .reveal .slide > img.r-stretch.quarto-figure-right  {
      display: block;
      margin-left: auto;
      margin-right: 0; 
    }
  </style>
</head>
<body class="quarto-light">
  <div class="reveal">
    <div class="slides">

<section id="title-slide" class="center">
  <h1 class="title">Simulations in population genetics</h1>
  <p class="author">Martin Petr (mp@bodkan.net)</p>
</section>

<section id="section" class="title-slide slide level1 center">
<h1></h1>
<div class="columns">
<div class="column" style="width:70%;">
<blockquote>
<p>Many problems in population genetics cannot be solved by a mathematician, no matter how gifted. [It] is already clear that computer methods are very powerful. This is good. It […] <strong>permits people with limited mathematical knowledge to work on important problems</strong> […]</p>
</blockquote>
</div><div class="column" style="width:30%;">
<p><img data-src="images/crow.jpeg"></p>
<p><a href="https://en.wikipedia.org/wiki/James_F._Crow">James F. Crow</a> – <a href="http://www.gnxp.com/blog/2006/06/10-questions-for-jim-crow.php">interview</a></p>
</div>
</div>
</section>

<section>
<section id="why-simulate-genomic-data" class="title-slide slide level1 center">
<h1>Why simulate genomic data?</h1>
<ol type="1">
<li>Exploring expected behavior of statistics</li>
<li>Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</li>
<li>Ground truth for method development</li>
</ol>
</section>
<section id="exploring-expected-behavior-of-statistics" class="slide level2">
<h2>Exploring expected behavior of statistics</h2>
<center>
<img data-src="images/fstats_sims.png">
</center>

<aside><div>
<p>Image from <a href="https://academic.oup.com/genetics/article/202/4/1485/5930214">Peter (2016)</a></p>
</div></aside></section>
<section id="fitting-model-parameters-i.e.-abc" class="slide level2">
<h2>Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</h2>
<center>
<img data-src="images/abc_scheme.png" style="width:50.0%">
</center>

<aside><div>
<p>Image from <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">Wikipedia on ABC</a></p>
</div></aside></section>
<section id="ground-truth-for-testing-and-method-development" class="slide level2">
<h2>Ground truth for testing and method development</h2>
<center>
<img data-src="images/mcmc.png">
</center>

<aside><div>
<p>Image from <a href="https://www.nature.com/articles/ng.3015">Schiffels and Durbin (2014)</a></p>
</div></aside></section></section>
<section id="what-does-it-mean-to-simulate-a-genome" class="title-slide slide level1 center">
<h1>What does it mean to simulate a genome?</h1>
<div class="fragment">
<br> How would you design an algorithm for a popgen simulation?

</div>
<div class="fragment">
<p><br>What minimum components are needed for the program to to be useful?</p>
</div>
</section>

<section id="if-we-want-to-simulate-population-genetics" class="title-slide slide level1 center">
<h1>If we want to simulate population genetics</h1>
<p><br></p>
<div class="fragment">
<p>We need <em>populations</em>.</p>
</div>
<div class="fragment">
<p>We need <em>genetics</em>.</p>
</div>
</section>

<section id="a-chromosome-is" class="title-slide slide level1 center">
<h1>A chromosome is…</h1>
<p><br></p>
<div class="fragment">
<p>…a linear sequence of nucleotides…</p>
<ul>
<li>a list of characters (A/G/C/T nucleotides)</li>
<li>a list of 0 or 1 values (alternative/derived allele)</li>
</ul>
</div>
<div class="fragment">
<p>… which accumulates mutations every generation at a given <em>mutation rate</em>.</p>
</div>
</section>

<section id="a-population-is" class="title-slide slide level1 center">
<h1>A population is…</h1>
<p><br></p>
<p>A collection of <em>individuals</em> at a given point in time.</p>
<div class="fragment">
<p>Each individual carrying two homologous <em>chromosomes</em> inherited from two parents in the previous generation.</p>
</div>
<div class="fragment">
<p>Chromosomes recombine at a certain <em>recombination rate</em>.</p>
</div>
</section>

<section>
<section id="home-brewed-single-locus-simulations-in-r" class="title-slide slide level1 center">
<h1>Home-brewed single-locus simulations in R</h1>
<p><br></p>
<p>Let’s make the algorithm even more minimal.</p>
</section>
<section id="section-1" class="slide level2">
<h2></h2>
<p><br></p>
<p>Some basic setup:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MASS)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="single-locus-simulation" class="slide level2">
<h2>Single-locus simulation</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb2" data-code-line-numbers="|1-3|5-7|9-10|11|13-14|16-18|20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a>N <span class="ot">&lt;-</span> <span class="dv">500</span> <span class="co"># number of (haploid) individuals in the population</span></span>
<span id="cb2-2"><a href="#cb2-2"></a>generations <span class="ot">&lt;-</span> <span class="dv">500</span> <span class="co"># number of generations to run the simulation for</span></span>
<span id="cb2-3"><a href="#cb2-3"></a>p_start <span class="ot">&lt;-</span> <span class="fl">0.5</span> <span class="co"># initial allele frequency</span></span>
<span id="cb2-4"><a href="#cb2-4"></a></span>
<span id="cb2-5"><a href="#cb2-5"></a><span class="co"># initialize an (empty) trajectory vector of allele frequencies</span></span>
<span id="cb2-6"><a href="#cb2-6"></a>p_traj<span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, generations)</span>
<span id="cb2-7"><a href="#cb2-7"></a>p_traj[<span class="dv">1</span>] <span class="ot">&lt;-</span> p_start</span>
<span id="cb2-8"><a href="#cb2-8"></a></span>
<span id="cb2-9"><a href="#cb2-9"></a><span class="co"># in each generation...</span></span>
<span id="cb2-10"><a href="#cb2-10"></a><span class="cf">for</span> (gen_i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>generations) {</span>
<span id="cb2-11"><a href="#cb2-11"></a>  p <span class="ot">&lt;-</span> p_traj[gen_i <span class="sc">-</span> <span class="dv">1</span>] <span class="co"># get the current frequency</span></span>
<span id="cb2-12"><a href="#cb2-12"></a></span>
<span id="cb2-13"><a href="#cb2-13"></a>  <span class="co"># ... calculate the allele frequency in the next generation ...</span></span>
<span id="cb2-14"><a href="#cb2-14"></a>  p_next <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, N, p) <span class="sc">/</span> N</span>
<span id="cb2-15"><a href="#cb2-15"></a></span>
<span id="cb2-16"><a href="#cb2-16"></a>  <span class="co"># ... and save it to the trajectory vector</span></span>
<span id="cb2-17"><a href="#cb2-17"></a>  p_traj[gen_i] <span class="ot">&lt;-</span> p_next</span>
<span id="cb2-18"><a href="#cb2-18"></a>}</span>
<span id="cb2-19"><a href="#cb2-19"></a></span>
<span id="cb2-20"><a href="#cb2-20"></a>p_traj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="n-500-p_0-0.5" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 500, <span class="math inline">\(p_0 = 0.5\)</span></h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(p_traj, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>),</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> p_start, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-2-1.png" width="960" class="r-stretch"></section>
<section id="lets-make-it-a-function" class="slide level2">
<h2>Let’s make it a function</h2>
<p><strong>Input:</strong> <span class="math inline">\(N\)</span>, <span class="math inline">\(p_0\)</span> and the number of generations</p>
<p><strong>Output:</strong> allele frequency trajectory vector</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4" data-code-line-numbers="|1|15"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a>simulate <span class="ot">&lt;-</span> <span class="cf">function</span>(N, p_start, generations) {</span>
<span id="cb4-2"><a href="#cb4-2"></a>  <span class="co"># initialize an (empty) trajectory vector of allele frequencies</span></span>
<span id="cb4-3"><a href="#cb4-3"></a>  p_traj<span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, generations)</span>
<span id="cb4-4"><a href="#cb4-4"></a>  p_traj[<span class="dv">1</span>] <span class="ot">&lt;-</span> p_start</span>
<span id="cb4-5"><a href="#cb4-5"></a></span>
<span id="cb4-6"><a href="#cb4-6"></a>  <span class="co"># in each generation...</span></span>
<span id="cb4-7"><a href="#cb4-7"></a>  <span class="cf">for</span> (gen_i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>generations) {</span>
<span id="cb4-8"><a href="#cb4-8"></a>    p <span class="ot">&lt;-</span> p_traj[gen_i <span class="sc">-</span> <span class="dv">1</span>] <span class="co"># get the current frequency</span></span>
<span id="cb4-9"><a href="#cb4-9"></a>    <span class="co"># ... calculate the allele frequency in the next generation ...</span></span>
<span id="cb4-10"><a href="#cb4-10"></a>    p_next <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, N, p) <span class="sc">/</span> N</span>
<span id="cb4-11"><a href="#cb4-11"></a>    <span class="co"># ... and save it to the trajectory vector</span></span>
<span id="cb4-12"><a href="#cb4-12"></a>    p_traj[gen_i] <span class="ot">&lt;-</span> p_next</span>
<span id="cb4-13"><a href="#cb4-13"></a>  }</span>
<span id="cb4-14"><a href="#cb4-14"></a></span>
<span id="cb4-15"><a href="#cb4-15"></a>  p_traj</span>
<span id="cb4-16"><a href="#cb4-16"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="n-500-p_0-0.5-20-replicates" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 500, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">500</span>, <span class="at">p_start =</span> <span class="fl">0.5</span>, <span class="at">generations =</span> <span class="dv">2000</span>))</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-4-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="n-1000-p_0-0.5-20-replicates" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 1000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">p_start =</span> <span class="fl">0.5</span>, <span class="at">generations =</span> <span class="dv">2000</span>))</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-5-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="n-5000-p_0-0.5-20-replicates" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 5000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">5000</span>, <span class="at">p_start =</span> <span class="fl">0.5</span>, <span class="at">generations =</span> <span class="dv">2000</span>))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-6-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="n-10000-p_0-0.5-20-replicates" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 10000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">p_start =</span> <span class="fl">0.5</span>, <span class="at">generations =</span> <span class="dv">20000</span>))</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-7-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="n-10000-p_0-0.5-20-replicates-1" class="slide level2">
<h2><span class="math inline">\(N\)</span> = 10000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>factors <span class="ot">&lt;-</span> <span class="fu">fractions</span>(<span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">10</span>))</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="dv">10000</span> <span class="sc">*</span> factors, <span class="at">lwd =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-8-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="expected-allele-frequency-distribution" class="slide level2">
<h2>Expected allele frequency distribution</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a>p_start <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>factors <span class="ot">&lt;-</span> <span class="fu">fractions</span>(<span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">10</span>))</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>final_frequencies <span class="ot">&lt;-</span> parallel<span class="sc">::</span><span class="fu">mclapply</span>(</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq_along</span>(factors), <span class="cf">function</span>(i) {</span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true" tabindex="-1"></a>    f <span class="ot">&lt;-</span> factors[i]</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true" tabindex="-1"></a>    t <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(N <span class="sc">*</span> f)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get complete trajectories as a matrix (each column a single trajectory)</span></span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true" tabindex="-1"></a>    reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">10000</span>, <span class="fu">simulate</span>(<span class="at">N =</span> N, <span class="at">p_start =</span> p_start, <span class="at">generations =</span> t))</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># only keep the last slice of the matrix with the final frequencies</span></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(</span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">t =</span> <span class="fu">sprintf</span>(<span class="st">"t = %s * N"</span>, f),</span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">freq =</span> reps[t, ]</span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true" tabindex="-1"></a>  },</span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">mc.cores =</span> parallel<span class="sc">::</span><span class="fu">detectCores</span>()</span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span> <span class="fu">do.call</span>(rbind, .)</span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true" tabindex="-1"></a>final_frequencies<span class="sc">$</span>t <span class="ot">&lt;-</span> <span class="fu">fct_rev</span>(<span class="fu">fct_relevel</span>(final_frequencies<span class="sc">$</span>t, <span class="fu">sprintf</span>(<span class="st">"t = %s * N"</span>, factors)))</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true" tabindex="-1"></a>final_frequencies <span class="sc">%&gt;%</span> .[.<span class="sc">$</span>freq <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> .<span class="sc">$</span>freq <span class="sc">&lt;</span> <span class="dv">1</span>, ] <span class="sc">%&gt;%</span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(freq, <span class="at">y =</span> ..density.., <span class="at">fill =</span> t), <span class="at">position =</span> <span class="st">"identity"</span>, <span class="at">bins =</span> <span class="dv">100</span>, <span class="at">alpha =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"allele frequency"</span>) <span class="sc">+</span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">3</span>)) <span class="sc">+</span></span>
<span id="cb10-26"><a href="#cb10-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(t <span class="sc">~</span> .) <span class="sc">+</span></span>
<span id="cb10-27"><a href="#cb10-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="fu">sprintf</span>(<span class="st">"time since</span><span class="sc">\n</span><span class="st">the start</span><span class="sc">\n</span><span class="st">[assuming</span><span class="sc">\n</span><span class="st">N = %s]"</span>, N))) <span class="sc">+</span></span>
<span id="cb10-28"><a href="#cb10-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb10-29"><a href="#cb10-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">strip.text.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb10-30"><a href="#cb10-30" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb10-31"><a href="#cb10-31" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb10-32"><a href="#cb10-32" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb10-33"><a href="#cb10-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb10-34"><a href="#cb10-34" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb10-35"><a href="#cb10-35" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-9-1.png" width="960"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/kimura.jpeg"></p>
<p></p><figcaption><em>“Diffusion Models in Population Genetics”</em>, <a href="https://www.jstor.org/stable/3211856#metadata_info_tab_contents">Kimura (1964)</a></figcaption><p></p>
</figure>
</div>
</div>
</div>
</div>
</section></section>
<section id="bonus-exercise-1" class="title-slide slide level1 center">
<h1>Bonus exercise #1</h1>
<p>If you’re bored later, try adding selection to this model!</p>
<p>In each generation, weight the frequencies based on fitness of each of the tree genotypes.</p>
<p>If you need a hint, take a look <a href="https://github.com/bodkan/notebooks/blob/master/MLE_single_locus_selection.ipynb">here</a>.</p>
</section>

<section>
<section id="why-bother" class="title-slide slide level1 center">
<h1>Why bother?</h1>
<!-- ## Let's add selection! -->
<!-- Write `simulate_selection(s, N, p_start, generations)`, where `s` is the selection coefficient. -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- simulate_selection <- function(s, N, p_start, generations) { -->
<!--   # we assume additivity -->
<!--   h <- 1/2 -->
<!--   # initialize an (empty) trajectory vector of allele frequencies -->
<!--   p_traj<- rep(NA, generations) -->
<!--   p_traj[1] <- p_start -->
<!--   # in each generation... -->
<!--   for (gen_i in 2:generations) { -->
<!--     p <- p_traj[gen_i - 1] # get the current frequency -->
<!--     fitnesses <- c( -->
<!--       p ^ 2           * (1 + s), -->
<!--       2 * p * (1 - p) * (1 + h * s), -->
<!--       (1 - p) ^ 2     * 1 -->
<!--     ) -->
<!--     p_next <- (fitnesses[1] + h * fitnesses[2]) / mean(fitnesses) -->
<!--     if (p_next >= 1) p_next <- 1 -->
<!--     # ... calculate the allele frequency in the next generation ... -->
<!--     p_next <- rbinom(1, N, p_next) / N -->
<!--     # ... and save it to the trajectory vector -->
<!--     p_traj[gen_i] <- p_next -->
<!--   } -->
<!--   p_traj -->
<!-- } -->
<!-- ``` -->
<!-- ```{r} -->
<!-- deterministic_selection <- function(t, p_start, s) { -->
<!--   p_start * exp(s * t / 2) / (p_start * exp(s * t / 2) + (1 - p_start)) -->
<!-- } -->
<!-- ``` -->
<!-- ```{r} -->
<!-- deterministic_traj <- sapply(1:1000, deterministic_selection, p_start = 0.01, s = 0.1) -->
<!-- plot(deterministic_traj, type = "l") -->
<!-- ``` -->
<!-- (Let's assume additive selection: $h = \frac{1}{2}$). -->
</section>
<section id="section-2" class="slide level2">
<h2></h2>
<div class="columns">
<div class="column" style="width:40%;">
<center>
<p><img data-src="images/feynman.jpeg"></p>
<p><a href="https://en.wikipedia.org/wiki/Richard_Feynman">Richard Feynman</a></p>
</center>
</div><div class="column" style="width:60%;">
<p><img data-src="images/feynman_quote.png"></p>
</div>
</div>
</section></section>
<section id="section-3" class="title-slide slide level1 center" data-background-image="images/montypython.jpeg">
<h1></h1>
<div class="fragment">
<h1 color="black" style="background-color: white">
But now for something completely different.
</h1>
</div>
</section>

<section>
<section id="section-4" class="title-slide slide level1 center" data-background-image="images/montypython.jpeg">
<h1></h1>
<h1 color="black" style="background-color: white">
Let’s do “real” simulations!
</h1>
</section>
<section id="there-are-many-pieces-of-simulation-software" class="slide level2">
<h2>There are many pieces of simulation software</h2>
<p><br></p>
<p>The most famous and widely used are <a href="https://messerlab.org/slim/">SLiM</a> and <a href="https://tskit.dev/msprime/docs/stable/intro.html">msprime</a>.</p>
<div class="fragment">
<p>They are both extremely powerful…</p>
</div>
<div class="fragment">
<p>… but they require a lot of programming knowledge and quite a lot of code to write non-trivial simulations (without bugs).</p>
</div>
<div class="fragment">
<p><br></p>
<center>
<strong>We will learn simulation concepts using <a href="http://www.slendr.net"><em>slendr</em></a>, a convenient R interface to both SLiM and msprime.</strong>
</center>
</div>
</section></section>
<section>
<section id="slim" class="title-slide slide level1 center">
<h1>SLiM</h1>

</section>
<section id="section-5" class="slide level2">
<h2></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2>
What is SLiM?
</h2>
<div class="fragment">
<ul>
<li><strong>Forward-time simulator</strong></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>It’s fully programmable!</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Massive library of functions for:
<ul>
<li>Demographic events</li>
<li>Various mating systems</li>
<li>Selection, quantitative traits, …</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>&gt; 700 pages long <a href="https://github.com/MesserLab/SLiM/releases/download/v3.7.1/SLiM_Manual.pdf">manual</a>!</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/sim_sketches.001.png" style="width:100.0%"></p>
<p></p><figcaption>Modified from Alex Drummond</figcaption><p></p>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="slimgui-ide-for-slim" class="slide level2">
<h2>SLiMgui – <a href="https://en.wikipedia.org/wiki/Integrated_development_environment">IDE</a> for SLiM</h2>
<center>
<img data-src="images/slimgui.png">
</center>
</section>
<section id="simple-neutral-simulation-in-slim" class="slide level2">
<h2>Simple neutral simulation in SLiM</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>initialize() {
    // create a neutral mutation type
    initializeMutationType("m1", 0.5, "f", 0.0);
    // initialize 1Mb segment
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, 999999);
    // set mutation rate and recombination rate of the segment
    initializeMutationRate(1e-8);
    initializeRecombinationRate(1e-8);
}

// create an ancestral population p1 of 10000 diploid individuals
1 early() { sim.addSubpop("p1", 10000); }

// in generation 1000, create two daughter populations p2 and p3
1000 early() {
    sim.addSubpopSplit("p2", 5000, p1);
    sim.addSubpopSplit("p3", 1000, p1);
}

// in generation 10000, stop the simulation and save 100 individuals
// from p2 and p3 to a VCF file
10000 late() {
    p2_subset = sample(p2.individuals, 100);
    p3_subset = sample(p3.individuals, 100);
    c(p2_subset, p3_subset).genomes.outputVCF("/tmp/slim_output.vcf.gz");
    sim.simulationFinished();
    catn("DONE!");
}</code></pre>
</div>
</div>
</section></section>
<section>
<section id="msprime" class="title-slide slide level1 center">
<h1>msprime</h1>

</section>
<section id="section-6" class="slide level2">
<h2></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2>
What is <a href="https://tskit.dev/msprime/docs/stable/intro.html">msprime</a>?
</h2>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/sim_sketches.001.png" style="width:100.0%"></p>
<p></p><figcaption>Modified from Alex Drummond</figcaption><p></p>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="section-7" class="slide level2">
<h2></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2>
What is <a href="https://tskit.dev/msprime/docs/stable/intro.html">msprime</a>?
</h2>
<div class="fragment">
<ul>
<li>A Python module for writing <strong>coalescent simulations</strong></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Extremely fast (genome-scale, population-scale data)</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>You must know Python fairly well to build complex models</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure>
<p><img data-src="images/sim_sketches.002.png" style="width:100.0%"></p>
<p></p><figcaption>Modified from Alex Drummond</figcaption><p></p>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="simple-simulation-using-msprime" class="slide level2">
<h2>Simple simulation using <em>msprime</em></h2>
<p>The following is basically the same model as the SLiM script earlier: &nbsp;</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>import msprime

demography = msprime.Demography()
demography.add_population(name="A", initial_size=10_000)
demography.add_population(name="B", initial_size=5_000)
demography.add_population(name="C", initial_size=1_000)
demography.add_population_split(time=1000, derived=["A", "B"], ancestral="C")

ts = msprime.sim_ancestry(
  sequence_length=10e6,
  recombination_rate=1e-8,
  samples={"A": 100, "B": 100},
  demography=demography
)</code></pre>
</div>
</div>
<p>source: <a href="https://tskit.dev/msprime/docs/stable/demography.html#demographic-models">link</a></p>
</section></section>
<section>
<section id="section-8" class="title-slide slide level1 center">
<h1></h1>
<center>
<p><img data-src="images/slendr_logo.png" style="width:30.0%"></p>
<p><br></p>
<h2>
<a href="www.slendr.net">www.slendr.net</a>
</h2>
</center>
</section>
<section id="why-a-new-package-spatial-simulations" class="slide level2">
<h2>Why a new package? – spatial simulations!</h2>
<div class="fragment">
<center>
<img data-src="images/animation.gif" style="width:70.0%">
</center>
</div>
</section>
<section id="why-a-new-package" class="slide level2">
<h2>Why a new package?</h2>
<div>
<ul>
<li class="fragment"><p>Most researchers are not expert programmers</p></li>
<li class="fragment"><p>All but the most trivial simulations require lots of code</p></li>
<li class="fragment"><p>90% <strong><citation needed=""></citation></strong> of simulations are basically the same!</p>
<ul>
<li class="fragment"><p>create populations (splits and <span class="math inline">\(N_e\)</span> changes)</p></li>
<li class="fragment"><p>specify if/how they should mix (rates and times)</p></li>
<li class="fragment"><p>save output (VCF, EIGENSTRAT)</p></li>
</ul></li>
<li class="fragment"><p>Lot of code duplication across projects</p></li>
</ul>
</div>
<div class="fragment">
<center>
<strong><em>slendr</em> makes “standard” simulations trivial (for newbies <em>and</em> experts) and unlocks new kinds of spatial simulations</strong>
</center>
</div>
</section></section>
<section>
<section id="lets-get-started" class="title-slide slide level1 center">
<h1>Let’s get started</h1>

</section>
<section id="we-will-use-slendr-tidyverse" class="slide level2">
<h2>We will use <code>slendr</code> &amp; <code>tidyverse</code></h2>
<p>First run this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse) <span class="co"># table processing/filtering and plotting</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr) <span class="co"># simulation and tree-sequence analysis</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br>If you get a message about missing Python modules, run:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setup_env</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<center>
<strong>(ignore the message about missing SLiM)</strong>
</center>
<p><br> To check that everything is setup correctly, run:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="fu">check_env</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section></section>
<section>
<section id="section-9" class="title-slide slide level1 center">
<h1></h1>
<h1>
<em>slendr</em> workflow:<br>
</h1>
<b>
<h2>
building (complex) models<br> from simple components
</h2>
</b><p><b></b></p>
</section>
<section id="typical-steps" class="slide level2">
<h2>Typical steps</h2>
<p><br></p>
<ol type="1">
<li>creating populations</li>
<li>scheduling population splits</li>
<li>programming <span class="math inline">\(N_e\)</span> size changes</li>
<li>encoding gene-flow events</li>
<li>simulation sequence of a given size</li>
<li>computing statistics from simulated outputs</li>
</ol>
</section>
<section id="creating-a-population" class="slide level2">
<h2>Creating a population</h2>
<p>A name, size and the time of appearance must be given:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<p>Typing an object prints out a summary in the R console:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>pop1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop1 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 1: created as an ancestral population (N = 1000)</code></pre>
</div>
</div>
</div>
</section>
<section id="programming-population-splits" class="slide level2">
<h2>Programming population splits</h2>
<p>Splits are indicated by the <code>parent = &lt;pop&gt;</code> argument:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<p>The split is reported in the “historical summary”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>pop2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop2 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 50: split from pop1</code></pre>
</div>
</div>
</div>
</section>
<section id="scheduling-resize-events-resize" class="slide level2">
<h2>Scheduling resize events – <code>resize()</code></h2>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb22" data-code-line-numbers="2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb22-2"><a href="#cb22-2"></a>pop1_step <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop1, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb23" data-code-line-numbers="2"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span>
<span id="cb23-2"><a href="#cb23-2"></a>pop2_exp <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop2, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="tidyverse-style-pipe-interface" class="slide level2">
<h2>Tidyverse-style <a href="https://magrittr.tidyverse.org">pipe</a> interface</h2>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>This accomplishes the same thing as the code on the previous slide, but it is a bit more “elegant”.</strong>
</center>
</section>
<section id="more-complex-full-model" class="slide level2">
<h2>More complex full model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb26" data-code-line-numbers="|15-20"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb26-2"><a href="#cb26-2"></a></span>
<span id="cb26-3"><a href="#cb26-3"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb26-4"><a href="#cb26-4"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">300</span>, <span class="at">parent =</span> pop1) <span class="sc">%&gt;%</span></span>
<span id="cb26-5"><a href="#cb26-5"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1000</span>)</span>
<span id="cb26-6"><a href="#cb26-6"></a></span>
<span id="cb26-7"><a href="#cb26-7"></a>pop3 <span class="ot">&lt;-</span></span>
<span id="cb26-8"><a href="#cb26-8"></a>  <span class="fu">population</span>(<span class="st">"pop3"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">400</span>, <span class="at">parent =</span> pop2) <span class="sc">%&gt;%</span></span>
<span id="cb26-9"><a href="#cb26-9"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">2500</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">800</span>)</span>
<span id="cb26-10"><a href="#cb26-10"></a></span>
<span id="cb26-11"><a href="#cb26-11"></a>pop4 <span class="ot">&lt;-</span></span>
<span id="cb26-12"><a href="#cb26-12"></a>  <span class="fu">population</span>(<span class="st">"pop4"</span>, <span class="at">N =</span> <span class="dv">1500</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">parent =</span> pop3) <span class="sc">%&gt;%</span></span>
<span id="cb26-13"><a href="#cb26-13"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">700</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1200</span>, <span class="at">end =</span> <span class="dv">2000</span>)</span>
<span id="cb26-14"><a href="#cb26-14"></a></span>
<span id="cb26-15"><a href="#cb26-15"></a>pop5 <span class="ot">&lt;-</span></span>
<span id="cb26-16"><a href="#cb26-16"></a>  <span class="fu">population</span>(<span class="st">"pop5"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">600</span>, <span class="at">parent =</span> pop4) <span class="sc">%&gt;%</span></span>
<span id="cb26-17"><a href="#cb26-17"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">50</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">900</span>) <span class="sc">%&gt;%</span></span>
<span id="cb26-18"><a href="#cb26-18"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">250</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1200</span>) <span class="sc">%&gt;%</span></span>
<span id="cb26-19"><a href="#cb26-19"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1600</span>, <span class="at">end =</span> <span class="dv">2200</span>) <span class="sc">%&gt;%</span></span>
<span id="cb26-20"><a href="#cb26-20"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">400</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">2400</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="remember-each-object-carries-its-history" class="slide level2">
<h2>Remember: each object carries its history</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>pop5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop5 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 600: split from pop4
  - time 900: resize from 100 to 50 individuals
  - time 1200: resize from 50 to 250 individuals
  - time 1600-2200: exponential resize from 250 to 1000 individuals
  - time 2400: resize from 1000 to 400 individuals</code></pre>
</div>
</div>
</section>
<section id="last-step-before-simulation-compilation" class="slide level2">
<h2>Last step before simulation: compilation</h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span></span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>Compilation takes a list of model components, performs internal consistency checks, returns a single model object.</strong>
</center>
</section>
<section id="model-visualization" class="slide level2">
<h2>Model visualization</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>

</div>
<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-26-1.png" width="672" class="r-stretch quarto-figure-center"></section>
<section id="simulating-data-finally" class="slide level2">
<h2>Simulating data (finally…)</h2>
<p>We have the compiled <code>model</code>, how do we simulate data?</p>
<p><em>slendr</em> has two simulation engines already built-in:</p>
<ul>
<li>SLiM engine</li>
<li>msprime engine</li>
</ul>
<div class="fragment">
<center>
<strong>You don’t have to write any msprime/SLiM code!</strong>
</center>
</div>
<div class="fragment">
<p>Take a model object and use the built-in simulation engine:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment">
<br>
<center>
<code>ts</code> is a so-called “tree sequence”
</center>
</div>
</section></section>
<section>
<section id="the-output-of-a-slendr-simulation-is-a-tree-sequence" class="title-slide slide level1 center">
<h1>The output of a <em>slendr</em> simulation is a <strong>tree sequence</strong></h1>

</section>
<section id="what-is-a-tree-sequence" class="slide level2">
<h2>What is a tree sequence?</h2>
<center>
<img data-src="images/tree_sequence_diagram.png" style="width:90.0%">
</center>
<ul>
<li>A record of full genetic ancestry of a set of samples</li>
<li>An encoding of DNA sequence carried by those samples</li>
<li>An efficient analysis framework</li>
</ul>
<center>
<a href="https://tskit.dev/learn.html#what">from the <em>tskit</em> project</a>
</center>
</section></section>
<section>
<section id="why-a-tree-sequence" class="title-slide slide level1 center">
<h1>Why a tree sequence?</h1>

</section>
<section id="what-we-usually-have" class="slide level2">
<h2>What we usually have</h2>
<center>
<img data-src="images/vcf_screenshot.png" style="width:90.0%">
</center>
</section>
<section id="what-we-usually-want" class="slide level2">
<h2>What we usually <em>want</em></h2>
<p>(As full as possible) a representation of our samples’ history:</p>
<center>
<img data-src="images/tree_sequence_diagram.png">
</center>
<div class="fragment">
<center>
<strong>And this is exactly what tree sequences give us.</strong>
</center>
</div>
</section>
<section id="lets-get-back-to-the-model-we-defined-earlier" class="slide level2">
<h2>Let’s get back to the model we defined earlier</h2>

<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-28-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="simulating-a-tree-sequence-output" class="slide level2">
<h2>Simulating a tree-sequence output</h2>
<p><br> By default, <code>msprime</code> function automatically loads the tree-sequence that the simulation produced in the background:</p>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(</span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>  model,</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">sequence_length =</span> <span class="fl">100e6</span>,</span>
<span id="cb32-4"><a href="#cb32-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">recombination_rate =</span> <span class="fl">1e-8</span></span>
<span id="cb32-5"><a href="#cb32-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<p>If we type <code>ts</code> into an R console, we get…</p>
</div>
</section>
<section id="a-tree-sequence-content-summary" class="slide level2">
<h2>… a tree-sequence content summary</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │     141334║
╟───────────────┼───────────╢
║Sequence Length│  100000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │       9400║
╟───────────────┼───────────╢
║Total Size     │   24.7 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤══════╤═════════╤════════════╗
║Table      │Rows  │Size     │Has Metadata║
╠═══════════╪══════╪═════════╪════════════╣
║Edges      │569170│ 17.4 MiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Individuals│  4700│128.5 KiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Migrations │     0│  8 Bytes│          No║
╟───────────┼──────┼─────────┼────────────╢
║Mutations  │     0│ 16 Bytes│          No║
╟───────────┼──────┼─────────┼────────────╢
║Nodes      │106580│  2.8 MiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Populations│     5│383 Bytes│         Yes║
╟───────────┼──────┼─────────┼────────────╢
║Provenances│     1│  3.9 KiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Sites      │     0│ 16 Bytes│          No║
╚═══════════╧══════╧═════════╧════════════╝</code></pre>
</div>
</div>
</section></section>
<section>
<section id="what-can-we-do-with-this" class="title-slide slide level1 center">
<h1>What can we do with this?</h1>

</section>
<section id="r-interface-to-tskit" class="slide level2">
<h2>R interface to tskit</h2>
<center>
<p><img data-src="images/slendr_tskit.png" style="width:120.0%"></p>
<p>This <a href="https://www.slendr.net/reference/index.html#tree-sequence-loading-and-processing">R interface</a> links to Python methods implemented in <a href="https://tskit.dev/tskit/docs/stable/python-api.html#statistics"><em>tskit</em></a>.</p>
</center>
</section>
<section id="here-is-the-magic" class="slide level2">
<h2>Here is the magic</h2>
<p>Tree sequences make it possible to directly compute many quantities of interest <em>without going via conversion to a genotype table/VCF</em>!</p>
<center>
<img data-src="images/tree_sequence_diagram.png">
</center>
</section></section>
<section id="how-do-we-use-this-in-practice" class="title-slide slide level1 center">
<h1>How do we use this in practice?</h1>

</section>

<section>
<section id="section-10" class="title-slide slide level1 center">
<h1></h1>
<h1>
To analyze tree sequences, we need to refer to “samples”
</h1>
</section>
<section id="extracting-sample-information" class="slide level2">
<h2>Extracting sample information</h2>
<p>Each “sampled” individual in <em>slendr</em> has a symbolic name, a sampling time, and a population assignment:</p>

<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-31-1.png" width="768" class="r-stretch quarto-figure-center"></section>
<section id="extracting-sample-information-1" class="slide level2">
<h2>Extracting sample information</h2>
<p>Each “sampled” individual in <em>slendr</em> has a symbolic name, a sampling time, and a population assignment:</p>
<div class="columns">
<div class="column" style="width:55%;">
<div class="cell" data-output-location="fragment">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,700 × 3
   name     time pop  
   &lt;chr&gt;   &lt;dbl&gt; &lt;chr&gt;
 1 pop1_1      0 pop1 
 2 pop1_2      0 pop1 
 3 pop1_3      0 pop1 
 4 pop1_4      0 pop1 
 5 pop1_5      0 pop1 
 6 pop1_6      0 pop1 
 7 pop1_7      0 pop1 
 8 pop1_8      0 pop1 
 9 pop1_9      0 pop1 
10 pop1_10     0 pop1 
# … with 4,690 more rows
# ℹ Use `print(n = ...)` to see more rows</code></pre>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:43%;">
<div class="cell" data-output-location="fragment">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">count</span>(pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 2
  pop       n
  &lt;chr&gt; &lt;int&gt;
1 pop1   1000
2 pop2    100
3 pop3   2500
4 pop4    700
5 pop5    400</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="analyzing-tree-sequences-with-slendr" class="slide level2">
<h2>Analyzing tree sequences with <em>slendr</em></h2>
<p>Let’s say we have the following model and we simulate a tree sequence from it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39" data-code-line-numbers="|1|2|4"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1"></a>pop <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop"</span>, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb39-2"><a href="#cb39-2"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(pop, <span class="at">generation_time =</span> <span class="dv">1</span>, <span class="at">simulation_length =</span> <span class="dv">10000</span>)</span>
<span id="cb39-3"><a href="#cb39-3"></a></span>
<span id="cb39-4"><a href="#cb39-4"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong><em>slendr</em> provides a large <a href="https://www.slendr.net/reference/index.html#tree-sequence-loading-and-processing">library of functions</a> for computing population genetic statistics on tree sequences</strong>
</center>
</section>
<section id="example-allele-frequency-spectrum" class="slide level2">
<h2>Example: allele frequency spectrum</h2>
<div class="cell">

</div>
<div class="columns">
<div class="column" style="width:55%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb40" data-code-line-numbers="|3-6|11-16"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1"></a><span class="co"># sample 5 individuals</span></span>
<span id="cb40-2"><a href="#cb40-2"></a><span class="co"># (i.e. 10 chromosomes)</span></span>
<span id="cb40-3"><a href="#cb40-3"></a>samples <span class="ot">&lt;-</span></span>
<span id="cb40-4"><a href="#cb40-4"></a>  <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb40-5"><a href="#cb40-5"></a>  <span class="fu">sample_n</span>(<span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb40-6"><a href="#cb40-6"></a>  <span class="fu">pull</span>(name)</span>
<span id="cb40-7"><a href="#cb40-7"></a></span>
<span id="cb40-8"><a href="#cb40-8"></a><span class="co"># compute allele frequency</span></span>
<span id="cb40-9"><a href="#cb40-9"></a><span class="co"># spectrum from the given set</span></span>
<span id="cb40-10"><a href="#cb40-10"></a><span class="co"># of individuals</span></span>
<span id="cb40-11"><a href="#cb40-11"></a>afs <span class="ot">&lt;-</span> <span class="fu">ts_afs</span>(</span>
<span id="cb40-12"><a href="#cb40-12"></a>  ts, <span class="fu">list</span>(samples),</span>
<span id="cb40-13"><a href="#cb40-13"></a>  <span class="at">mode =</span> <span class="st">"branch"</span>,</span>
<span id="cb40-14"><a href="#cb40-14"></a>  <span class="at">polarised =</span> <span class="cn">TRUE</span></span>
<span id="cb40-15"><a href="#cb40-15"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:43%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs, <span class="at">type =</span> <span class="st">"b"</span>,</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"allele count bin"</span>,</span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-37-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="but-wait-we-dont-have-any-mutations" class="slide level2">
<h2>But wait, we don’t have any mutations!</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │     389699║
╟───────────────┼───────────╢
║Sequence Length│  100000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │      20000║
╟───────────────┼───────────╢
║Total Size     │   65.8 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤═══════╤═════════╤════════════╗
║Table      │Rows   │Size     │Has Metadata║
╠═══════════╪═══════╪═════════╪════════════╣
║Edges      │1517216│ 46.3 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Individuals│  10000│273.5 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Migrations │      0│  8 Bytes│          No║
╟───────────┼───────┼─────────┼────────────╢
║Mutations  │      0│ 16 Bytes│          No║
╟───────────┼───────┼─────────┼────────────╢
║Nodes      │ 285922│  7.6 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Populations│      1│222 Bytes│         Yes║
╟───────────┼───────┼─────────┼────────────╢
║Provenances│      1│  1.3 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Sites      │      0│ 16 Bytes│          No║
╚═══════════╧═══════╧═════════╧════════════╝</code></pre>
</div>
</div>
</section>
<section id="how-can-we-compute-statistics" class="slide level2">
<h2>How can we compute statistics?</h2>
<p>There is a duality between mutations and branch lengths in trees (more <a href="https://tskit.dev/tskit/docs/stable/stats.html">here</a>).</p>
<center>
<p><img data-src="images/tree_sequence_diagram.png"></p>
<h3>
But what if we want mutations?
</h3>
</center>
</section>
<section id="coalescent-and-mutation-processes-can-be-decoupled" class="slide level2">
<h2>Coalescent and mutation processes can be decoupled!</h2>
<center>
<p><img data-src="images/tree_sequence_diagram.png"></p>
<h3>
This means we can add mutations<br><em>after</em> the simulation.
</h3>
</center>
</section>
<section id="this-allows-efficient-massive-simulations-especially-with-slim" class="slide level2">
<h2>This allows efficient, massive simulations (especially with SLiM)</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>ts_mutated <span class="ot">&lt;-</span> <span class="fu">ts_mutate</span>(</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  ts,</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">mutation_rate =</span> <span class="fl">1e-8</span></span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment">
<p>Or, with a shortcut:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(</span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>  model,</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">sequence_length =</span> <span class="fl">100e6</span>,</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">recombination_rate =</span> <span class="fl">1e-8</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">mutation_rate =</span> <span class="fl">1e-8</span></span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div><div class="column" style="width:1%;">
<p>&nbsp;</p>
</div><div class="column" style="width:59%;">
<div class="fragment">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │     390860║
╟───────────────┼───────────╢
║Sequence Length│  100000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │      20000║
╟───────────────┼───────────╢
║Total Size     │   90.8 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤═══════╤═════════╤════════════╗
║Table      │Rows   │Size     │Has Metadata║
╠═══════════╪═══════╪═════════╪════════════╣
║Edges      │1521502│ 46.4 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Individuals│  10000│273.5 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Migrations │      0│  8 Bytes│          No║
╟───────────┼───────┼─────────┼────────────╢
║Mutations  │ 419734│ 14.8 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Nodes      │ 286611│  7.7 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Populations│      1│222 Bytes│         Yes║
╟───────────┼───────┼─────────┼────────────╢
║Provenances│      2│  2.1 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Sites      │ 418855│ 10.0 MiB│          No║
╚═══════════╧═══════╧═════════╧════════════╝</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="example-allele-frequency-spectrum-1" class="slide level2">
<h2>Example: allele frequency spectrum</h2>
<div class="cell">

</div>
<div class="columns">
<div class="column" style="width:55%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sample 5 individuals</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (i.e. 10 chromosomes)</span></span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(name)</span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a><span class="co"># compute allele frequency</span></span>
<span id="cb47-9"><a href="#cb47-9" aria-hidden="true" tabindex="-1"></a><span class="co"># spectrum from the given set</span></span>
<span id="cb47-10"><a href="#cb47-10" aria-hidden="true" tabindex="-1"></a><span class="co"># of individuals</span></span>
<span id="cb47-11"><a href="#cb47-11" aria-hidden="true" tabindex="-1"></a>afs <span class="ot">&lt;-</span> <span class="fu">ts_afs</span>(</span>
<span id="cb47-12"><a href="#cb47-12" aria-hidden="true" tabindex="-1"></a>  ts, <span class="fu">list</span>(samples),</span>
<span id="cb47-13"><a href="#cb47-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">polarised =</span> <span class="cn">TRUE</span></span>
<span id="cb47-14"><a href="#cb47-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb47-15"><a href="#cb47-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-16"><a href="#cb47-16" aria-hidden="true" tabindex="-1"></a>afs</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 40380 19809 13063  9861  7884  6690  5525  4979  4350  4185</code></pre>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:43%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs, <span class="at">type =</span> <span class="st">"b"</span>,</span>
<span id="cb49-2"><a href="#cb49-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"allele count bin"</span>,</span>
<span id="cb49-3"><a href="#cb49-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-44-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="what-we-have-learned-so-far" class="slide level2">
<h2>What we have learned so far</h2>
<ol type="1">
<li>creating populations – <a href="https://www.slendr.net/reference/population.html"><code>populations()</code></a></li>
<li>compiling models – <a href="https://www.slendr.net/reference/compile_model.html"><code>compile_model()</code></a></li>
<li>simulating tree sequences – <a href="https://www.slendr.net/reference/msprime.html"><code>msprime()</code></a></li>
<li>extracting samples – <a href="https://www.slendr.net/reference/ts_samples.html"><code>ts_samples()</code></a></li>
<li>computing AFS – <a href="https://www.slendr.net/reference/ts_afs.html"><code>ts_afs()</code></a></li>
</ol>
<p><br></p>
<center>
<h1>
<strong>Let’s put this to use!</strong>
</h1>
</center>
</section></section>
<section>
<section id="exercise-1" class="title-slide slide level1 center">
<h1>Exercise #1</h1>

</section>
<section id="exercise-1-1" class="slide level2">
<h2>Exercise #1</h2>
<div class="cell">

</div>
<p>Collaborator Hestu gave you AFS computed from 20 individuals of a sub-species of the <a href="https://www.zeldadungeon.net/wiki/Bushy-Tailed_Squirrel"><em>bushy-tailed squirrel</em></a> discovered in the Forest of Spirits in the land of Hyrule:</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>c(2520, 1449, 855, 622, 530, 446, 365, 334, 349, 244, 264, 218, 
133, 173, 159, 142, 167, 129, 125, 143)</code></pre>
</div>
</div>
<p>Fossil evidence is consistent with constant size of the population over 100,000 generations of its history. An Oracle you met in the Temple of Time said that the true squirrel <span class="math inline">\(N_e\)</span> has been between 1000 and 30000.</p>
<p>Use <em>slendr</em> to simulate history of this species. Use this to guess the likely value of squirrel’s <span class="math inline">\(N_e\)</span> given the observed AFS.</p>
</section>
<section id="exercise-1-hints" class="slide level2">
<h2>Exercise #1 – hints</h2>
<ol type="1">
<li><p>Write a function that gets <span class="math inline">\(N_e\)</span> as input and returns the AFS.</p></li>
<li><p>Find the <span class="math inline">\(N_e\)</span> value that will give the closest AFS to the one you got from Hestu. Use whatever method you’re comfortable with based on your programming experience:</p>
<ul>
<li><p>i ) Plot simulated AFS for different <span class="math inline">\(N_e\)</span> with the AFS and just eye-ball <span class="math inline">\(N_e\)</span> value that looks correct.</p></li>
<li><p>ii ) Simulate AFS across a grid of <span class="math inline">\(N_e\)</span> values and find the closest matching one (maybe use <a href="https://en.wikipedia.org/wiki/Mean_squared_error">mean-squared error</a>?)</p></li>
<li><p>iii ) Run a mini-Approximate Bayesian Computation, using the Oracle’s range of [10000, 30000] as a uniform prior.</p></li>
</ul></li>
</ol>
</section>
<section id="exercise-1-simulations-on-a-grid" class="slide level2">
<h2>Exercise #1 – simulations on a grid</h2>
<center>
<h3>
<a href="https://github.com/bodkan/popgen-2022/blob/main/solutions/exercise_1_grid.R">solution</a>
</h3>
</center>
</section>
<section id="exercise-1-abc-inference" class="slide level2">
<h2>Exercise #1 – ABC inference</h2>
<center>
<h3>
<a href="https://github.com/bodkan/popgen-2022/blob/main/solutions/exercise_1_abc.R">solution</a>
</h3>
</center>
</section>
<section id="exercise-2" class="slide level2">
<h2>Exercise #2</h2>
<div class="columns">
<div class="column" style="width:60%;">
<p>Tectonic shifts caused by activated Sheikah towers caused speciation of the squirrels. Unfortunately, their future is clouded by the shadow cast over Hyrule by Calamity Ganon.</p>
<p>Can you help the Oracle predict the future shape of the AFS of the squirrels, assuming starting <span class="math inline">\(N_e\)</span> = 6543? Species #1 will remain constant, species #2 will expand 3X, species #3 will get 3X smaller.</p>
</div><div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-47-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</section>
<section id="exercise-2-solution" class="slide level2">
<h2>Exercise #2 – solution</h2>
<p><br><br></p>
<center>
<h3>
<a href="https://github.com/bodkan/popgen-2022/blob/main/solutions/exercise_2.R">link</a>
</h3>
</center>
</section></section>
<section>
<section id="what-else-can-slendr-do" class="title-slide slide level1 center">
<h1>What else can <em>slendr</em> do?</h1>

</section>
<section id="gene-flow-events" class="slide level2">
<h2>Gene flow events</h2>
<p>Gene flow is programmed using the <code>gene_flow()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> pop1, <span class="at">to =</span> pop2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<p>Multiple gene-flow events can be gathered in a list:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> pop1, <span class="at">to =</span> pop2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>),</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> pop2, <span class="at">to =</span> pop1, <span class="at">start =</span> <span class="dv">700</span>, <span class="at">end =</span> <span class="dv">750</span>, <span class="at">rate =</span> <span class="fl">0.1</span>)</span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment">
<br>
<center>
<strong><code>gene_flow()</code> checks admixture events for consistency</strong>
</center>
</div>
</section>
<section id="lets-build-another-toy-model" class="slide level2">
<h2>Let’s build another toy model</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a>o <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"o"</span>, <span class="at">time =</span> <span class="dv">1</span>, <span class="at">N =</span> <span class="dv">100</span>)</span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>c <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"c"</span>, <span class="at">time =</span> <span class="dv">2500</span>, <span class="at">N =</span> <span class="dv">500</span>, <span class="at">parent =</span> o)</span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"a"</span>, <span class="at">time =</span> <span class="dv">3000</span>, <span class="at">N =</span> <span class="dv">2000</span>, <span class="at">parent =</span> c)</span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"b"</span>, <span class="at">time =</span> <span class="dv">3500</span>, <span class="at">N =</span> <span class="dv">4000</span>, <span class="at">parent =</span> a)</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a>x1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"x1"</span>, <span class="at">time =</span> <span class="dv">3800</span>, <span class="at">N =</span> <span class="dv">8000</span>, <span class="at">parent =</span> c)</span>
<span id="cb53-6"><a href="#cb53-6" aria-hidden="true" tabindex="-1"></a>x2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"x2"</span>, <span class="at">time =</span> <span class="dv">4000</span>, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">parent =</span> x1)</span>
<span id="cb53-7"><a href="#cb53-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-8"><a href="#cb53-8" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> b, <span class="at">to =</span> x1, <span class="at">start =</span> <span class="dv">5500</span>, <span class="at">end =</span> <span class="dv">6000</span>, <span class="at">rate =</span> <span class="fl">0.3</span>)</span>
<span id="cb53-9"><a href="#cb53-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-10"><a href="#cb53-10" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb53-11"><a href="#cb53-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(a, b, x1, x2, c, o), <span class="at">gene_flow =</span> gf,</span>
<span id="cb53-12"><a href="#cb53-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>, <span class="at">simulation_length =</span> <span class="dv">7000</span></span>
<span id="cb53-13"><a href="#cb53-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb53-14"><a href="#cb53-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-15"><a href="#cb53-15" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb53-16"><a href="#cb53-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span>
<span id="cb53-17"><a href="#cb53-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb53-18"><a href="#cb53-18" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">proportions =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-50-1.png" width="672" class="r-stretch quarto-figure-center"><div class="cell">

</div>
</section>
<section id="diversity-ts_diversity" class="slide level2">
<h2>Diversity – <code>ts_diversity()</code></h2>
<div class="columns">
<div class="column" style="width:48%;">
<p>Extract a list of lists of individuals’ names for each population:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">split</span>(., .<span class="sc">$</span>pop) <span class="sc">%&gt;%</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(pull, <span class="st">"name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:50%;">
<div class="fragment">
<p>Compute diversity in each population:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_diversity</span>(ts, samples) <span class="sc">%&gt;%</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(diversity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  set    diversity
  &lt;chr&gt;      &lt;dbl&gt;
1 o     0.00000380
2 c     0.0000199 
3 a     0.0000538 
4 b     0.0000686 
5 x2    0.0000725 
6 x1    0.0000763 </code></pre>
</div>
</div>
</div>
</div>
</div>
</section></section>
<section>
<section id="exercise-3" class="title-slide slide level1 center">
<h1>Exercise #3</h1>

</section>
<section id="program-the-following-model-in-slendr" class="slide level2">
<h2>Program the following model in <em>slendr</em></h2>
<div class="cell">

</div>

<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-55-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="parameter-values" class="slide level2">
<h2>Parameter values</h2>
<ul>
<li><p>“CH” outgroup at time 7 Mya (<span class="math inline">\(N_e\)</span> = 7k)</p></li>
<li><p>“AFR” splitting from “CH” at 6 Mya (<span class="math inline">\(N_e\)</span> = 10k)</p></li>
<li><p>“ALT” splitting from “AFR” at 700 kya (<span class="math inline">\(N_e\)</span> = 500)</p></li>
<li><p>“VI” splitting from “ALT” at 120 kya (<span class="math inline">\(N_e\)</span> = 1k)</p></li>
<li><p>“EUR” splitting from “AFR” at 70 kya (<span class="math inline">\(N_e\)</span> = 5k)</p></li>
<li><p>gene flow from “VI” to “EUR” at 3% over 50-40 kya</p></li>
</ul>
<p>Then simulate 100Mb sequence with <code>msprime()</code>, using recombination rate 1e-8 per bp per generation.</p>
</section>
<section id="exercise-3-solution" class="slide level2">
<h2>Exercise #3 – solution</h2>
<p><br><br></p>
<center>
<h3>
<a href="https://github.com/bodkan/popgen-2022/blob/main/solutions/exercise_3.R">link</a>
</h3>
</center>
</section></section>
<section>
<section id="exercise-4" class="title-slide slide level1 center">
<h1>Exercise #4</h1>

</section>
<section id="diversity-and-divergence" class="slide level2">
<h2>Diversity and divergence</h2>
<div class="columns">
<div class="column" style="width:60%;">
<p>Use <code>ts_diversity()</code> to compute diversity in each simulated population (<code>CH</code>, <code>AFR</code>, …).</p>
<p>Does the result correspond to what you would expect based on the demographic history?</p>
<p>What about pairwise genetic divergence between all populations (<code>ts_divergence()</code>)? Do your results recapitulate phylogenetic relationships?</p>
</div><div class="column" style="width:40%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-56-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</section>
<section id="exercise-4-solution-diversity" class="slide level2">
<h2>Exercise #4 – solution (diversity)</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">50e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>sample_names <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">split</span>(., .<span class="sc">$</span>pop) <span class="sc">%&gt;%</span> <span class="fu">lapply</span>(pull, <span class="st">"name"</span>)</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-7"><a href="#cb57-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> sample_names) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(diversity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 2
  set   diversity
  &lt;chr&gt;     &lt;dbl&gt;
1 ALT   0.0000199
2 VI    0.0000365
3 CH    0.000276 
4 EUR   0.000365 
5 AFR   0.000391 </code></pre>
</div>
</div>
</section>
<section id="exercise-4-solution-divergence" class="slide level2">
<h2>Exercise #4 – solution (divergence)</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>divergence <span class="ot">&lt;-</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_divergence</span>(ts, sample_names) <span class="sc">%&gt;%</span></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(divergence)</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>divergence</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 3
   x     y     divergence
   &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt;
 1 ALT   VI      0.000101
 2 AFR   EUR     0.000448
 3 EUR   VI      0.000850
 4 ALT   EUR     0.000851
 5 AFR   VI      0.000865
 6 AFR   ALT     0.000866
 7 CH    VI      0.00427 
 8 CH    EUR     0.00427 
 9 AFR   CH      0.00427 
10 ALT   CH      0.00427 </code></pre>
</div>
</div>
</div><div class="column" style="width:60%;">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>divergence <span class="sc">%&gt;%</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">pair =</span> <span class="fu">paste</span>(x, <span class="st">"-"</span>, y),</span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">pair =</span> <span class="fu">fct_reorder</span>(pair, divergence)) <span class="sc">%&gt;%</span></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(pair, divergence)) <span class="sc">+</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">""</span>, <span class="at">y =</span> <span class="st">"divergence"</span>) <span class="sc">+</span></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Pairwise genetic divergence between populations"</span>) <span class="sc">+</span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-59-1.png" width="960"></p>
</div>
</div>
</div>
</div>
</section>
<section id="sampling-ancient-dna-time-series" class="slide level2">
<h2>Sampling “ancient DNA” time-series</h2>
<p>By default, <em>slendr</em> records every individual living at the end of the simulation. Sampling can be also scheduled explicitly:</p>
<div class="fragment">
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb62" data-code-line-numbers="1-4|6-9|11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb62-1"><a href="#cb62-1"></a>x_schedule <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(</span>
<span id="cb62-2"><a href="#cb62-2"></a>  model, <span class="at">times =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">7000</span>, <span class="at">to =</span> <span class="dv">4000</span>, <span class="at">by =</span> <span class="sc">-</span><span class="dv">100</span>),</span>
<span id="cb62-3"><a href="#cb62-3"></a>  <span class="fu">list</span>(x1, <span class="dv">5</span>), <span class="fu">list</span>(x2, <span class="dv">5</span>)</span>
<span id="cb62-4"><a href="#cb62-4"></a>)</span>
<span id="cb62-5"><a href="#cb62-5"></a></span>
<span id="cb62-6"><a href="#cb62-6"></a>abco_schedule <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(</span>
<span id="cb62-7"><a href="#cb62-7"></a>  model, <span class="at">times =</span> <span class="dv">7000</span>,</span>
<span id="cb62-8"><a href="#cb62-8"></a>  <span class="fu">list</span>(a, <span class="dv">1</span>), <span class="fu">list</span>(b, <span class="dv">1</span>), <span class="fu">list</span>(c, <span class="dv">1</span>), <span class="fu">list</span>(o, <span class="dv">1</span>)</span>
<span id="cb62-9"><a href="#cb62-9"></a>)</span>
<span id="cb62-10"><a href="#cb62-10"></a></span>
<span id="cb62-11"><a href="#cb62-11"></a>schedule <span class="ot">&lt;-</span> <span class="fu">rbind</span>(x_schedule, abco_schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment">
<p>The schedule can be used in an <code>msprime()</code> call like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>,</span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">samples =</span> schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="computing-f_4-statistic-ts_f4" class="slide level2">
<h2>Computing <span class="math inline">\(f_4\)</span> statistic – <code>ts_f4()</code></h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell">
<div class="cell-output-display">
<p><img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-63-1.png" width="960"></p>
</div>
</div>
</div><div class="column" style="width:40%;">
<p>Having simulated data from this model, can we detect gene flow from <em>b</em> into <em>x1</em>?</p>
</div>
</div>
</section>
<section id="computing-f_4-statistic-ts_f4-1" class="slide level2">
<h2>Computing <span class="math inline">\(f_4\)</span> statistic – <code>ts_f4()</code></h2>
<p>Simulate tree sequence, add mutations on it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">samples =</span> schedule,</span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="fragment">
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f4</span>(ts, <span class="st">"x1_1"</span>, <span class="st">"x2_1"</span>, <span class="st">"b_1"</span> , <span class="st">"o_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  W     X     Y     Z              f4
  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;       &lt;dbl&gt;
1 x1_1  x2_1  b_1   o_1   0.000000138</code></pre>
</div>
</div>
</div>
<div class="fragment">
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f4</span>(ts, <span class="st">"x1_100"</span>, <span class="st">"x2_100"</span>, <span class="st">"b_1"</span> , <span class="st">"o_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  W      X      Y     Z             f4
  &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt;
1 x1_100 x2_100 b_1   o_1   0.00000206</code></pre>
</div>
</div>
</div>
</section>
<section id="computing-f_4-statistic-for-all-x-individuals" class="slide level2">
<h2>Computing <span class="math inline">\(f_4\)</span> statistic for all <code>x</code> individuals</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract information about samples from populations x1 and x2</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>x_inds <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(pop <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"x1"</span>, <span class="st">"x2"</span>))</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a><span class="co"># create a vector of individual's names</span></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a><span class="co"># (used as input in ts_&lt;...&gt;() functions</span></span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a>x_names <span class="ot">&lt;-</span> x_inds<span class="sc">$</span>name</span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a><span class="co"># iterate over all sample names, compute</span></span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a><span class="co"># f4(X, C; B, O)</span></span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a>f4_result<span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(</span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>  x_names,</span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(w) <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> w, <span class="at">X =</span> <span class="st">"c_1"</span>, <span class="at">Y =</span> <span class="st">"b_1"</span> , <span class="at">Z =</span> <span class="st">"o_1"</span>)</span>
<span id="cb69-13"><a href="#cb69-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb69-14"><a href="#cb69-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-15"><a href="#cb69-15" aria-hidden="true" tabindex="-1"></a><span class="co"># add the f4 value as a column to the sample information table</span></span>
<span id="cb69-16"><a href="#cb69-16" aria-hidden="true" tabindex="-1"></a>x_inds<span class="sc">$</span>f4 <span class="ot">&lt;-</span> f4_result<span class="sc">$</span>f4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="computing-f_4-statistic-for-all-x-individuals-1" class="slide level2">
<h2>Computing <span class="math inline">\(f_4\)</span> statistic for all <code>x</code> individuals</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(x_inds) <span class="sc">+</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rect</span>(<span class="fu">aes</span>(<span class="at">xmin =</span> <span class="dv">5500</span>, <span class="at">xmax =</span> <span class="dv">6000</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymax =</span> <span class="cn">Inf</span>), <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">fill =</span> <span class="st">"gray"</span>) <span class="sc">+</span></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(time, f4, <span class="at">color =</span> pop)) <span class="sc">+</span></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data =</span> . <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(pop, time) <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="at">f4 =</span> <span class="fu">mean</span>(f4)),</span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(time, f4, <span class="at">color =</span> pop), <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-68-1.png" width="960" class="r-stretch quarto-figure-center"></section></section>
<section>
<section id="exercise-5" class="title-slide slide level1 center">
<h1>Exercise #5</h1>

</section>
<section id="neanderthal-ancestry-trajectory" class="slide level2">
<h2>Neanderthal ancestry trajectory</h2>
<ol type="1">
<li><p>Take your model of Neanderthal introgression</p></li>
<li><p>Implement temporal sampling:</p></li>
</ol>
<ul>
<li>one “European” every 1000 yrs between 40 kya and today</li>
<li>“Altai” (70 ky old) and “Vindija” individuals (40 ky old)</li>
</ul>
<ol start="3" type="1">
<li>Compute <span class="math inline">\(f_4\)</span>-ratio statistic using:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f4ratio</span>(ts, <span class="at">X =</span> <span class="st">'&lt;vector of "European" individuals&gt;'</span>,</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">A =</span> <span class="st">'&lt;"Altai"&gt;'</span>, <span class="at">C =</span> <span class="st">'&lt;"African"&gt;'</span>, <span class="at">O =</span> <span class="st">'&lt;"Chimp"&gt;'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="4" type="1">
<li>Plot the estimated trajectory of Neanderthal ancestry in Europe over time.</li>
</ol>
</section></section>
<section id="exercise-5-solution" class="title-slide slide level1 center">
<h1>Exercise #5 – solution</h1>
<center>
<h3>
<a href="https://github.com/bodkan/popgen-2022/blob/main/solutions/exercise_5.R">solution</a>
</h3>
</center>
<div class="cell">

</div>
</section>

<section>
<section id="tree-sequence-simplification" class="title-slide slide level1 center">
<h1>Tree-sequence simplification</h1>

</section>
<section id="sometimes-a-tree-sequence-is-too-big" class="slide level2">
<h2>Sometimes a tree sequence is too big</h2>

<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-71-1.png" width="960" class="r-stretch"><center>
<strong>Simplification reduces the genealogies only to those nodes that form ancestors of a desired set of samples.</strong>
</center>
</section>
<section id="meet-ts_simplify" class="slide level2">
<h2>Meet <code>ts_simplify()</code></h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>ts_small <span class="ot">&lt;-</span> <span class="fu">ts_simplify</span>(</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>  ts,</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">simplify_to =</span> <span class="fu">c</span>(<span class="st">"CH_1"</span>, <span class="st">"AFR_1"</span>, <span class="st">"AFR_2"</span>, <span class="st">"AFR_3"</span>,</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>                  <span class="st">"ALT_1"</span>, <span class="st">"VI_1"</span>, <span class="st">"EUR_1"</span>, <span class="st">"EUR_2"</span>, <span class="st">"EUR_3"</span>, <span class="st">"EUR_4"</span>, <span class="st">"EUR_5"</span>)</span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- ```{r} -->
<!-- library(ggtree) -->
<!-- pdf("trees.pdf") -->
<!-- for (i in seq_len(ts_small$num_trees)) { -->
<!--   tree <- ts_phylo(ts_small, i, quiet = TRUE) -->
<!-- nodes <- ts_nodes(tree) %>% -->
<!--   as_tibble %>%  -->
<!--   dplyr::select(node = phylo_id, pop) -->
<!-- p_tree <- ggtree(tree) %<+% nodes + -->
<!--   geom_tiplab(aes(color = pop, fill = pop)) + -->
<!--   guides(color = "none") + -->
<!--   ggtitle(i) + -->
<!--   scale_x_continuous(limits = c(-7e6, 900e3)) -->
<!-- print(revts(p_tree)) -->
<!-- } -->
<!-- dev.off() -->
<!-- ``` -->
<p><br></p>
<center>
<strong>Only the coalescent nodes of genealogies involving<br>these samples will be retained.</strong>
</center>
</section>
<section id="extracting-a-tree-1734" class="slide level2">
<h2>Extracting a tree #1734</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">ts_phylo</span>(ts_small, <span class="dv">1734</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this needs another R package</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a><span class="co"># simply using the ape package and running `plot(tree)` will also work</span></span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> <span class="fu">ts_nodes</span>(tree) <span class="sc">%&gt;%</span></span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>  as_tibble <span class="sc">%&gt;%</span></span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">node =</span> phylo_id, pop)</span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a>p_tree <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree) <span class="sc">%&lt;+%</span> nodes <span class="sc">+</span></span>
<span id="cb74-10"><a href="#cb74-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> pop, <span class="at">fill =</span> pop)) <span class="sc">+</span></span>
<span id="cb74-11"><a href="#cb74-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb74-12"><a href="#cb74-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">7e6</span>, <span class="fl">900e3</span>)) <span class="sc">+</span></span>
<span id="cb74-13"><a href="#cb74-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="fl">700e3</span>, <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"red"</span>)</span>
<span id="cb74-14"><a href="#cb74-14" aria-hidden="true" tabindex="-1"></a><span class="fu">revts</span>(p_tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-74-1.png" width="960" class="r-stretch quarto-figure-center"></section>
<section id="extracting-a-tree-1" class="slide level2">
<h2>Extracting a tree #1</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">ts_phylo</span>(ts_small, <span class="dv">1</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this needs another R package</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="co"># simply using the ape package and running `plot(tree)` will also work</span></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> <span class="fu">ts_nodes</span>(tree) <span class="sc">%&gt;%</span></span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>  as_tibble <span class="sc">%&gt;%</span></span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">node =</span> phylo_id, pop)</span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a>p_tree <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree) <span class="sc">%&lt;+%</span> nodes <span class="sc">+</span></span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> pop, <span class="at">fill =</span> pop)) <span class="sc">+</span></span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb76-12"><a href="#cb76-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">7e6</span>, <span class="fl">900e3</span>)) <span class="sc">+</span></span>
<span id="cb76-13"><a href="#cb76-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="fl">700e3</span>, <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"red"</span>)</span>
<span id="cb76-14"><a href="#cb76-14" aria-hidden="true" tabindex="-1"></a><span class="fu">revts</span>(p_tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>

</div>
<img data-src="popgen2022_files/figure-revealjs/unnamed-chunk-76-1.png" width="960" class="r-stretch quarto-figure-center"><div class="fragment">
<center>
<strong>Introgression!</strong>
</center>
</div>
</section></section>
<section>
<section id="bonus-exercise-2" class="title-slide slide level1 center">
<h1>Bonus exercise #2</h1>

</section>
<section id="try-to-implement-something-relevant-for-your-own-work" class="slide level2">
<h2>Try to implement something relevant for your own work</h2>
<ul>
<li>Program a demographic model for your species of interest</li>
<li>Choose a statistic whose value “true value” you know from the literature (divergence? diversity? f4-statistic? ancestry proportion?)</li>
<li>Try to replicate the statistic in a <em>slendr</em> simulation using some of its <code>ts_&lt;...&gt;()</code> tree-sequence functions!</li>
</ul>
</section></section>
<section>
<section id="bonus-exercise-3" class="title-slide slide level1 center">
<h1>Bonus exercise #3</h1>

</section>
<section id="try-to-implement-your-own-d-statistic" class="slide level2">
<h2>Try to implement your own D-statistic</h2>
<p>You can extract data from a tree sequence in the form of a genotype table using the function <code>ts_genotypes()</code>.</p>
<p>This function returns a simple R data frame:</p>
<ul>
<li>each column for an individual</li>
<li>each row for a site</li>
<li>each cell giving the number of derived alleles in that individual at that site</li>
</ul>
<p>Use the genotype table to compute your own ABBA/BABA statistic on the Neanderthal model data and check if you can recover the same signal as you get from <code>ts_f4()</code>.</p>
<div class="footer footer-default">

</div>
</section></section>
    </div>
  </div>

  <script>window.backupDefine = window.define; window.define = undefined;</script>
  <script src="popgen2022_files/libs/revealjs/dist/reveal.js"></script>
  <!-- reveal.js plugins -->
  <script src="popgen2022_files/libs/revealjs/plugin/quarto-line-highlight/line-highlight.js"></script>
  <script src="popgen2022_files/libs/revealjs/plugin/pdf-export/pdfexport.js"></script>
  <script src="popgen2022_files/libs/revealjs/plugin/reveal-menu/menu.js"></script>
  <script src="popgen2022_files/libs/revealjs/plugin/reveal-menu/quarto-menu.js"></script>
  <script src="popgen2022_files/libs/revealjs/plugin/quarto-support/support.js"></script>
  

  <script src="popgen2022_files/libs/revealjs/plugin/notes/notes.js"></script>
  <script src="popgen2022_files/libs/revealjs/plugin/search/search.js"></script>
  <script src="popgen2022_files/libs/revealjs/plugin/zoom/zoom.js"></script>
  <script src="popgen2022_files/libs/revealjs/plugin/math/math.js"></script>
  <script>window.define = window.backupDefine; window.backupDefine = undefined;</script>

  <script>

      // Full list of configuration options available at:
      // https://revealjs.com/config/
      Reveal.initialize({
'controlsAuto': true,
'previewLinksAuto': false,
'smaller': false,
'pdfSeparateFragments': false,
'autoAnimateEasing': "ease",
'autoAnimateDuration': 1,
'autoAnimateUnmatched': true,
'menu': {"side":"left","useTextContentForMissingTitles":true,"markers":false,"loadIcons":false,"custom":[{"title":"Tools","icon":"<i class=\"fas fa-gear\"></i>","content":"<ul class=\"slide-menu-items\">\n<li class=\"slide-tool-item active\" data-item=\"0\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.fullscreen(event)\"><kbd>f</kbd> Fullscreen</a></li>\n<li class=\"slide-tool-item\" data-item=\"1\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.speakerMode(event)\"><kbd>s</kbd> Speaker View</a></li>\n<li class=\"slide-tool-item\" data-item=\"2\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>o</kbd> Slide Overview</a></li>\n<li class=\"slide-tool-item\" data-item=\"3\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.overview(event)\"><kbd>e</kbd> PDF Export Mode</a></li>\n<li class=\"slide-tool-item\" data-item=\"4\"><a href=\"#\" onclick=\"RevealMenuToolHandlers.keyboardHelp(event)\"><kbd>?</kbd> Keyboard Help</a></li>\n</ul>"}],"openButton":true},
'smaller': false,
 
        // Display controls in the bottom right corner
        controls: false,

        // Help the user learn the controls by providing hints, for example by
        // bouncing the down arrow when they first encounter a vertical slide
        controlsTutorial: false,

        // Determines where controls appear, "edges" or "bottom-right"
        controlsLayout: 'edges',

        // Visibility rule for backwards navigation arrows; "faded", "hidden"
        // or "visible"
        controlsBackArrows: 'faded',

        // Display a presentation progress bar
        progress: true,

        // Display the page number of the current slide
        slideNumber: 'c/t',

        // 'all', 'print', or 'speaker'
        showSlideNumber: 'all',

        // Add the current slide number to the URL hash so that reloading the
        // page/copying the URL will return you to the same slide
        hash: true,

        // Start with 1 for the hash rather than 0
        hashOneBasedIndex: false,

        // Flags if we should monitor the hash and change slides accordingly
        respondToHashChanges: true,

        // Push each slide change to the browser history
        history: true,

        // Enable keyboard shortcuts for navigation
        keyboard: true,

        // Enable the slide overview mode
        overview: true,

        // Disables the default reveal.js slide layout (scaling and centering)
        // so that you can use custom CSS layout
        disableLayout: false,

        // Vertical centering of slides
        center: false,

        // Enables touch navigation on devices with touch input
        touch: true,

        // Loop the presentation
        loop: false,

        // Change the presentation direction to be RTL
        rtl: false,

        // see https://revealjs.com/vertical-slides/#navigation-mode
        navigationMode: 'linear',

        // Randomizes the order of slides each time the presentation loads
        shuffle: false,

        // Turns fragments on and off globally
        fragments: true,

        // Flags whether to include the current fragment in the URL,
        // so that reloading brings you to the same fragment position
        fragmentInURL: false,

        // Flags if the presentation is running in an embedded mode,
        // i.e. contained within a limited portion of the screen
        embedded: false,

        // Flags if we should show a help overlay when the questionmark
        // key is pressed
        help: true,

        // Flags if it should be possible to pause the presentation (blackout)
        pause: true,

        // Flags if speaker notes should be visible to all viewers
        showNotes: false,

        // Global override for autoplaying embedded media (null/true/false)
        autoPlayMedia: null,

        // Global override for preloading lazy-loaded iframes (null/true/false)
        preloadIframes: null,

        // Number of milliseconds between automatically proceeding to the
        // next slide, disabled when set to 0, this value can be overwritten
        // by using a data-autoslide attribute on your slides
        autoSlide: 0,

        // Stop auto-sliding after user input
        autoSlideStoppable: true,

        // Use this method for navigation when auto-sliding
        autoSlideMethod: null,

        // Specify the average time in seconds that you think you will spend
        // presenting each slide. This is used to show a pacing timer in the
        // speaker view
        defaultTiming: null,

        // Enable slide navigation via mouse wheel
        mouseWheel: false,

        // The display mode that will be used to show slides
        display: 'block',

        // Hide cursor if inactive
        hideInactiveCursor: true,

        // Time before the cursor is hidden (in ms)
        hideCursorTime: 5000,

        // Opens links in an iframe preview overlay
        previewLinks: false,

        // Transition style (none/fade/slide/convex/concave/zoom)
        transition: 'none',

        // Transition speed (default/fast/slow)
        transitionSpeed: 'default',

        // Transition style for full page slide backgrounds
        // (none/fade/slide/convex/concave/zoom)
        backgroundTransition: 'none',

        // Number of slides away from the current that are visible
        viewDistance: 3,

        // Number of slides away from the current that are visible on mobile
        // devices. It is advisable to set this to a lower number than
        // viewDistance in order to save resources.
        mobileViewDistance: 2,

        // The "normal" size of the presentation, aspect ratio will be preserved
        // when the presentation is scaled to fit different resolutions. Can be
        // specified using percentage units.
        width: 1050,

        height: 700,

        // Factor of the display size that should remain empty around the content
        margin: 0.1,

        math: {
          mathjax: 'https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.0/MathJax.js',
          config: 'TeX-AMS_HTML-full',
          tex2jax: {
            inlineMath: [['\\(','\\)']],
            displayMath: [['\\[','\\]']],
            balanceBraces: true,
            processEscapes: false,
            processRefs: true,
            processEnvironments: true,
            preview: 'TeX',
            skipTags: ['script','noscript','style','textarea','pre','code'],
            ignoreClass: 'tex2jax_ignore',
            processClass: 'tex2jax_process'
          },
        },

        // reveal.js plugins
        plugins: [QuartoLineHighlight, PdfExport, RevealMenu, QuartoSupport,

          RevealMath,
          RevealNotes,
          RevealSearch,
          RevealZoom
        ]
      });
    </script>
    
    <script>
      // htmlwidgets need to know to resize themselves when slides are shown/hidden.
      // Fire the "slideenter" event (handled by htmlwidgets.js) when the current
      // slide changes (different for each slide format).
      (function () {
        function fireSlideChanged(previousSlide, currentSlide) {

          // dispatch for htmlwidgets
          const event = window.document.createEvent("Event");
          event.initEvent("slideenter", true, true);
          window.document.dispatchEvent(event);

          // dispatch for shiny
          if (window.jQuery) {
            if (previousSlide) {
              window.jQuery(previousSlide).trigger("hidden");
            }
            if (currentSlide) {
              window.jQuery(currentSlide).trigger("shown");
            }
          }
        }

        // hookup for reveal
        if (window.Reveal) {
          window.Reveal.addEventListener("slidechanged", function(event) {
            fireSlideChanged(event.previousSlide, event.currentSlide);
          });
        }

        // hookup for slidy
        if (window.w3c_slidy) {
          window.w3c_slidy.add_observer(function (slide_num) {
            // slide_num starts at position 1
            fireSlideChanged(null, w3c_slidy.slides[slide_num - 1]);
          });
        }

      })();
    </script>

    <script id="quarto-html-after-body" type="application/javascript">
    window.document.addEventListener("DOMContentLoaded", function (event) {
      const toggleBodyColorMode = (bsSheetEl) => {
        const mode = bsSheetEl.getAttribute("data-mode");
        const bodyEl = window.document.querySelector("body");
        if (mode === "dark") {
          bodyEl.classList.add("quarto-dark");
          bodyEl.classList.remove("quarto-light");
        } else {
          bodyEl.classList.add("quarto-light");
          bodyEl.classList.remove("quarto-dark");
        }
      }
      const toggleBodyColorPrimary = () => {
        const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
        if (bsSheetEl) {
          toggleBodyColorMode(bsSheetEl);
        }
      }
      toggleBodyColorPrimary();  
      const tabsets =  window.document.querySelectorAll(".panel-tabset-tabby")
      tabsets.forEach(function(tabset) {
        const tabby = new Tabby('#' + tabset.id);
      });
      const clipboard = new window.ClipboardJS('.code-copy-button', {
        target: function(trigger) {
          return trigger.previousElementSibling;
        }
      });
      clipboard.on('success', function(e) {
        // button target
        const button = e.trigger;
        // don't keep focus
        button.blur();
        // flash "checked"
        button.classList.add('code-copy-button-checked');
        var currentTitle = button.getAttribute("title");
        button.setAttribute("title", "Copied!");
        setTimeout(function() {
          button.setAttribute("title", currentTitle);
          button.classList.remove('code-copy-button-checked');
        }, 1000);
        // clear code selection
        e.clearSelection();
      });
      function tippyHover(el, contentFn) {
        const config = {
          allowHTML: true,
          content: contentFn,
          maxWidth: 500,
          delay: 100,
          arrow: false,
          appendTo: function(el) {
              return el.closest('section.slide') || el.parentElement;
          },
          interactive: true,
          interactiveBorder: 10,
          theme: 'quarto-reveal',
          placement: 'bottom-start'
        };
          config['offset'] = [0,0];
          config['maxWidth'] = 700;
        window.tippy(el, config); 
      }
      const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
      for (var i=0; i<noterefs.length; i++) {
        const ref = noterefs[i];
        tippyHover(ref, function() {
          let href = ref.getAttribute('href');
          try { href = new URL(href).hash; } catch {}
          const id = href.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          return note.innerHTML;
        });
      }
      var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
      for (var i=0; i<bibliorefs.length; i++) {
        const ref = bibliorefs[i];
        const cites = ref.parentNode.getAttribute('data-cites').split(' ');
        tippyHover(ref, function() {
          var popup = window.document.createElement('div');
          cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    });
    </script>
    

</body></html>