<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.0.38">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="PopGen 2022 — Martin Petr (mp@bodkan.net)">

<title>Simulations in population genetics</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>


<script src="popgen2022_files/libs/clipboard/clipboard.min.js"></script>
<script src="popgen2022_files/libs/quarto-html/quarto.js"></script>
<script src="popgen2022_files/libs/quarto-html/popper.min.js"></script>
<script src="popgen2022_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="popgen2022_files/libs/quarto-html/anchor.min.js"></script>
<link href="popgen2022_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="popgen2022_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="popgen2022_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="popgen2022_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="popgen2022_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">

  <script src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-chtml-full.js" type="text/javascript"></script>

</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Simulations in population genetics</h1>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p><a href="http://www.popgen.dk/popgen22/">PopGen 2022</a> — Martin Petr (mp@bodkan.net) </p>
          </div>
  </div>
    
    
  </div>
  

</header>

<section id="section" class="level1">
<h1></h1>
<div class="columns">
<div class="column" style="width:70%;">
<blockquote class="blockquote">
<p>Many problems in population genetics cannot be solved by a mathematician, no matter how gifted. [It] is already clear that computer methods are very powerful. This is good. It […] <strong>permits people with limited mathematical knowledge to work on important problems</strong> […]</p>
</blockquote>
</div><div class="column" style="width:30%;">
<p><img src="images/crow.jpeg" class="img-fluid"></p>
<p><a href="https://en.wikipedia.org/wiki/James_F._Crow">James F. Crow</a> – <a href="http://www.gnxp.com/blog/2006/06/10-questions-for-jim-crow.php">interview</a></p>
</div>
</div>
</section>
<section id="why-simulate-genomic-data" class="level1">
<h1>Why simulate genomic data?</h1>
<ol type="1">
<li>Exploring expected behavior of statistics</li>
<li>Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</li>
<li>Ground truth for method development</li>
</ol>
<section id="exploring-expected-behavior-of-statistics" class="level2">
<h2 class="anchored" data-anchor-id="exploring-expected-behavior-of-statistics">Exploring expected behavior of statistics</h2>
<center>
<img src="images/fstats_sims.png" class="img-fluid">
</center>
<div class="aside">
<p>Image from <a href="https://academic.oup.com/genetics/article/202/4/1485/5930214">Peter (2016)</a></p>
</div>
</section>
<section id="fitting-model-parameters-i.e.-abc" class="level2">
<h2 class="anchored" data-anchor-id="fitting-model-parameters-i.e.-abc">Fitting model parameters (i.e.&nbsp;<a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">ABC</a>)</h2>
<center>
<img src="images/abc_scheme.png" class="img-fluid" style="width:50.0%">
</center>
<div class="aside">
<p>Image from <a href="https://en.wikipedia.org/wiki/Approximate_Bayesian_computation">Wikipedia on ABC</a></p>
</div>
</section>
<section id="ground-truth-for-testing-and-method-development" class="level2">
<h2 class="anchored" data-anchor-id="ground-truth-for-testing-and-method-development">Ground truth for testing and method development</h2>
<center>
<img src="images/mcmc.png" class="img-fluid">
</center>
<div class="aside">
<p>Image from <a href="https://www.nature.com/articles/ng.3015">Schiffels and Durbin (2014)</a></p>
</div>
</section>
</section>
<section id="what-does-it-mean-to-simulate-a-genome" class="level1">
<h1>What does it mean to simulate a genome?</h1>
<div class="incremental">
<br> How would you design an algorithm for a popgen simulation?

</div>
<div class="incremental">
<p><br>What minimum components are needed for the program to to be useful?</p>
</div>
</section>
<section id="if-we-want-to-simulate-population-genetics" class="level1">
<h1>If we want to simulate population genetics</h1>
<p><br></p>
<div class="incremental">
<p>We need <em>populations</em>.</p>
</div>
<div class="incremental">
<p>We need <em>genetics</em>.</p>
</div>
</section>
<section id="a-chromosome-is" class="level1">
<h1>A chromosome is…</h1>
<p><br></p>
<div class="incremental">
<p>…a linear sequence of nucleotides…</p>
<ul>
<li>a list of characters (A/G/C/T nucleotides)</li>
<li>a list of 0 or 1 values (alternative/derived allele)</li>
</ul>
</div>
<div class="incremental">
<p>… which accumulates mutations every generation at a given <em>mutation rate</em>.</p>
</div>
</section>
<section id="a-population-is" class="level1">
<h1>A population is…</h1>
<p><br></p>
<p>A collection of <em>individuals</em> at a given point in time.</p>
<div class="incremental">
<p>Each individual carrying two homologous <em>chromosomes</em> inherited from two parents in the previous generation.</p>
</div>
<div class="incremental">
<p>Chromosomes recombine at a certain <em>recombination rate</em>.</p>
</div>
</section>
<section id="home-brewed-single-locus-simulations-in-r" class="level1">
<h1>Home-brewed single-locus simulations in R</h1>
<p><br></p>
<p>Let’s make the algorithm even more minimal by focusing on the evolution of a single mutation across time (i.e.&nbsp;no mutation process, no recombination).</p>
<section id="some-basic-setup" class="level2">
<h2 class="anchored" data-anchor-id="some-basic-setup">Some basic setup</h2>
<p>If you want to try the following couple of examples yourself, you will need to paste the following lines into your R session (either R in the terminal, or better an RStudio session):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>✔ ggplot2 3.3.6     ✔ purrr   0.3.4
✔ tibble  3.1.8     ✔ dplyr   1.0.9
✔ tidyr   1.2.0     ✔ stringr 1.4.0
✔ readr   2.1.2     ✔ forcats 0.5.1</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()</code></pre>
</div>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(MASS)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'MASS'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:dplyr':

    select</code></pre>
</div>
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">42</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="single-locus-simulation" class="level2">
<h2 class="anchored" data-anchor-id="single-locus-simulation">Single-locus simulation</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1"></a>N <span class="ot">&lt;-</span> <span class="dv">500</span> <span class="co"># number of (haploid) individuals in the population</span></span>
<span id="cb9-2"><a href="#cb9-2"></a>generations <span class="ot">&lt;-</span> <span class="dv">500</span> <span class="co"># number of generations to run the simulation for</span></span>
<span id="cb9-3"><a href="#cb9-3"></a>p_start <span class="ot">&lt;-</span> <span class="fl">0.5</span> <span class="co"># initial allele frequency</span></span>
<span id="cb9-4"><a href="#cb9-4"></a></span>
<span id="cb9-5"><a href="#cb9-5"></a><span class="co"># initialize an (empty) trajectory vector of allele frequencies</span></span>
<span id="cb9-6"><a href="#cb9-6"></a>p_traj<span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, generations)</span>
<span id="cb9-7"><a href="#cb9-7"></a>p_traj[<span class="dv">1</span>] <span class="ot">&lt;-</span> p_start</span>
<span id="cb9-8"><a href="#cb9-8"></a></span>
<span id="cb9-9"><a href="#cb9-9"></a><span class="co"># in each generation...</span></span>
<span id="cb9-10"><a href="#cb9-10"></a><span class="cf">for</span> (gen_i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>generations) {</span>
<span id="cb9-11"><a href="#cb9-11"></a>  p <span class="ot">&lt;-</span> p_traj[gen_i <span class="sc">-</span> <span class="dv">1</span>] <span class="co"># get the current frequency</span></span>
<span id="cb9-12"><a href="#cb9-12"></a></span>
<span id="cb9-13"><a href="#cb9-13"></a>  <span class="co"># ... calculate the allele frequency in the next generation ...</span></span>
<span id="cb9-14"><a href="#cb9-14"></a>  p_next <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, N, p) <span class="sc">/</span> N</span>
<span id="cb9-15"><a href="#cb9-15"></a></span>
<span id="cb9-16"><a href="#cb9-16"></a>  <span class="co"># ... and save it to the trajectory vector</span></span>
<span id="cb9-17"><a href="#cb9-17"></a>  p_traj[gen_i] <span class="ot">&lt;-</span> p_next</span>
<span id="cb9-18"><a href="#cb9-18"></a>}</span>
<span id="cb9-19"><a href="#cb9-19"></a></span>
<span id="cb9-20"><a href="#cb9-20"></a>p_traj</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="n-500-p_0-0.5" class="level2">
<h2 class="anchored" data-anchor-id="n-500-p_0-0.5"><span class="math inline">\(N\)</span> = 500, <span class="math inline">\(p_0 = 0.5\)</span></h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(p_traj, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>),</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">h =</span> p_start, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-2-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="lets-make-it-a-function" class="level2">
<h2 class="anchored" data-anchor-id="lets-make-it-a-function">Let’s make it a function</h2>
<p><strong>Input:</strong> <span class="math inline">\(N\)</span>, <span class="math inline">\(p_0\)</span> and the number of generations</p>
<p><strong>Output:</strong> allele frequency trajectory vector</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1"></a>simulate <span class="ot">&lt;-</span> <span class="cf">function</span>(N, p_start, generations) {</span>
<span id="cb11-2"><a href="#cb11-2"></a>  <span class="co"># initialize an (empty) trajectory vector of allele frequencies</span></span>
<span id="cb11-3"><a href="#cb11-3"></a>  p_traj<span class="ot">&lt;-</span> <span class="fu">rep</span>(<span class="cn">NA</span>, generations)</span>
<span id="cb11-4"><a href="#cb11-4"></a>  p_traj[<span class="dv">1</span>] <span class="ot">&lt;-</span> p_start</span>
<span id="cb11-5"><a href="#cb11-5"></a></span>
<span id="cb11-6"><a href="#cb11-6"></a>  <span class="co"># in each generation...</span></span>
<span id="cb11-7"><a href="#cb11-7"></a>  <span class="cf">for</span> (gen_i <span class="cf">in</span> <span class="dv">2</span><span class="sc">:</span>generations) {</span>
<span id="cb11-8"><a href="#cb11-8"></a>    p <span class="ot">&lt;-</span> p_traj[gen_i <span class="sc">-</span> <span class="dv">1</span>] <span class="co"># get the current frequency</span></span>
<span id="cb11-9"><a href="#cb11-9"></a>    <span class="co"># ... calculate the allele frequency in the next generation ...</span></span>
<span id="cb11-10"><a href="#cb11-10"></a>    p_next <span class="ot">&lt;-</span> <span class="fu">rbinom</span>(<span class="dv">1</span>, N, p) <span class="sc">/</span> N</span>
<span id="cb11-11"><a href="#cb11-11"></a>    <span class="co"># ... and save it to the trajectory vector</span></span>
<span id="cb11-12"><a href="#cb11-12"></a>    p_traj[gen_i] <span class="ot">&lt;-</span> p_next</span>
<span id="cb11-13"><a href="#cb11-13"></a>  }</span>
<span id="cb11-14"><a href="#cb11-14"></a></span>
<span id="cb11-15"><a href="#cb11-15"></a>  p_traj</span>
<span id="cb11-16"><a href="#cb11-16"></a>}</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="n-500-p_0-0.5-20-replicates" class="level2">
<h2 class="anchored" data-anchor-id="n-500-p_0-0.5-20-replicates"><span class="math inline">\(N\)</span> = 500, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">500</span>, <span class="at">p_start =</span> <span class="fl">0.5</span>, <span class="at">generations =</span> <span class="dv">2000</span>))</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-4-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="n-1000-p_0-0.5-20-replicates" class="level2">
<h2 class="anchored" data-anchor-id="n-1000-p_0-0.5-20-replicates"><span class="math inline">\(N\)</span> = 1000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">p_start =</span> <span class="fl">0.5</span>, <span class="at">generations =</span> <span class="dv">2000</span>))</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-5-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="n-5000-p_0-0.5-20-replicates" class="level2">
<h2 class="anchored" data-anchor-id="n-5000-p_0-0.5-20-replicates"><span class="math inline">\(N\)</span> = 5000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">5000</span>, <span class="at">p_start =</span> <span class="fl">0.5</span>, <span class="at">generations =</span> <span class="dv">2000</span>))</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-6-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="n-10000-p_0-0.5-20-replicates" class="level2">
<h2 class="anchored" data-anchor-id="n-10000-p_0-0.5-20-replicates"><span class="math inline">\(N\)</span> = 10000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a>reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">20</span>, <span class="fu">simulate</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">p_start =</span> <span class="fl">0.5</span>, <span class="at">generations =</span> <span class="dv">20000</span>))</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-7-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="n-10000-p_0-0.5-20-replicates-1" class="level2">
<h2 class="anchored" data-anchor-id="n-10000-p_0-0.5-20-replicates-1"><span class="math inline">\(N\)</span> = 10000, <span class="math inline">\(p_0 = 0.5\)</span> (20 replicates)</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>factors <span class="ot">&lt;-</span> <span class="fu">fractions</span>(<span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">10</span>))</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">matplot</span>(reps, <span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">1</span>), <span class="at">xlab =</span> <span class="st">"generations"</span>, <span class="at">ylab =</span> <span class="st">"allele frequency"</span>, <span class="at">type =</span> <span class="st">"l"</span>, <span class="at">lty =</span> <span class="dv">1</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">abline</span>(<span class="at">v =</span> <span class="dv">10000</span> <span class="sc">*</span> factors, <span class="at">lwd =</span> <span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-8-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="expected-allele-frequency-distribution" class="level2">
<h2 class="anchored" data-anchor-id="expected-allele-frequency-distribution">Expected allele frequency distribution</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a>p_start <span class="ot">&lt;-</span> <span class="fl">0.5</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>N <span class="ot">&lt;-</span> <span class="dv">10000</span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>factors <span class="ot">&lt;-</span> <span class="fu">fractions</span>(<span class="fu">c</span>(<span class="dv">3</span>, <span class="dv">2</span>, <span class="dv">1</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">2</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">5</span>, <span class="dv">1</span><span class="sc">/</span><span class="dv">10</span>))</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a>final_frequencies <span class="ot">&lt;-</span> parallel<span class="sc">::</span><span class="fu">mclapply</span>(</span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">seq_along</span>(factors), <span class="cf">function</span>(i) {</span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>    f <span class="ot">&lt;-</span> factors[i]</span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>    t <span class="ot">&lt;-</span> <span class="fu">as.integer</span>(N <span class="sc">*</span> f)</span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>    <span class="co"># get complete trajectories as a matrix (each column a single trajectory)</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>    reps <span class="ot">&lt;-</span> <span class="fu">replicate</span>(<span class="dv">10000</span>, <span class="fu">simulate</span>(<span class="at">N =</span> N, <span class="at">p_start =</span> p_start, <span class="at">generations =</span> t))</span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>    <span class="co"># only keep the last slice of the matrix with the final frequencies</span></span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a>    <span class="fu">data.frame</span>(</span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a>      <span class="at">t =</span> <span class="fu">sprintf</span>(<span class="st">"t = %s * N"</span>, f),</span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>      <span class="at">freq =</span> reps[t, ]</span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>  },</span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>  <span class="at">mc.cores =</span> parallel<span class="sc">::</span><span class="fu">detectCores</span>()</span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span> <span class="fu">do.call</span>(rbind, .)</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a>final_frequencies<span class="sc">$</span>t <span class="ot">&lt;-</span> <span class="fu">fct_rev</span>(<span class="fu">fct_relevel</span>(final_frequencies<span class="sc">$</span>t, <span class="fu">sprintf</span>(<span class="st">"t = %s * N"</span>, factors)))</span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>final_frequencies <span class="sc">%&gt;%</span> .[.<span class="sc">$</span>freq <span class="sc">&gt;</span> <span class="dv">0</span> <span class="sc">&amp;</span> .<span class="sc">$</span>freq <span class="sc">&lt;</span> <span class="dv">1</span>, ] <span class="sc">%&gt;%</span></span>
<span id="cb17-22"><a href="#cb17-22" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb17-23"><a href="#cb17-23" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_histogram</span>(<span class="fu">aes</span>(freq, <span class="at">y =</span> ..density.., <span class="at">fill =</span> t), <span class="at">position =</span> <span class="st">"identity"</span>, <span class="at">bins =</span> <span class="dv">100</span>, <span class="at">alpha =</span> <span class="fl">0.75</span>) <span class="sc">+</span></span>
<span id="cb17-24"><a href="#cb17-24" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"allele frequency"</span>) <span class="sc">+</span></span>
<span id="cb17-25"><a href="#cb17-25" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_cartesian</span>(<span class="at">ylim =</span> <span class="fu">c</span>(<span class="dv">0</span>, <span class="dv">3</span>)) <span class="sc">+</span></span>
<span id="cb17-26"><a href="#cb17-26" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(t <span class="sc">~</span> .) <span class="sc">+</span></span>
<span id="cb17-27"><a href="#cb17-27" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">fill =</span> <span class="fu">guide_legend</span>(<span class="fu">sprintf</span>(<span class="st">"time since</span><span class="sc">\n</span><span class="st">the start</span><span class="sc">\n</span><span class="st">[assuming</span><span class="sc">\n</span><span class="st">N = %s]"</span>, N))) <span class="sc">+</span></span>
<span id="cb17-28"><a href="#cb17-28" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb17-29"><a href="#cb17-29" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">strip.text.y =</span> <span class="fu">element_blank</span>(),</span>
<span id="cb17-30"><a href="#cb17-30" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb17-31"><a href="#cb17-31" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.title.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb17-32"><a href="#cb17-32" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb17-33"><a href="#cb17-33" aria-hidden="true" tabindex="-1"></a>        <span class="at">axis.text.y =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb17-34"><a href="#cb17-34" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.title =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>),</span>
<span id="cb17-35"><a href="#cb17-35" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.text =</span> <span class="fu">element_text</span>(<span class="at">size =</span> <span class="dv">15</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-9-1.png" class="img-fluid figure-img" width="960"></p>
</figure>
</div>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="fragment">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/kimura.jpeg" class="img-fluid figure-img"></p>
<p></p><figcaption class="figure-caption"><em>“Diffusion Models in Population Genetics”</em>, <a href="https://www.jstor.org/stable/3211856#metadata_info_tab_contents">Kimura (1964)</a></figcaption><p></p>
</figure>
</div>
</div>
</div>
</div>
<!-- # Why bother? -->
<!-- ## -->
<!-- :::: columns -->
<!-- ::: {.column width="40%"} -->
<!-- <center> -->
<!-- ![](images/feynman.jpeg) -->
<!-- [Richard Feynman](https://en.wikipedia.org/wiki/Richard_Feynman) -->
<!-- </center> -->
<!-- ::: -->
<!-- ::: {.column width="60%"} -->
<!-- ![](images/feynman_quote.png) -->
<!-- ::: -->
<!-- :::: -->
</section>
</section>
<section id="bonus-exercise-1" class="level1">
<h1>Bonus exercise #1</h1>
<p>If you’re bored later, <strong>try adding selection</strong> to this model!</p>
<p>In each generation, weight the frequencies based on fitness of each of the tree genotypes.</p>
<p>If you need a hint, take a look <a href="https://github.com/bodkan/notebooks/blob/master/MLE_single_locus_selection.ipynb">here</a>.</p>
</section>
<section id="section-1" class="level1" data-background-image="images/montypython.jpeg">
<h1 data-background-image="images/montypython.jpeg"></h1>
<div class="fragment">
<h1 color="black" style="background-color: white">
But now for something completely different.
</h1>
</div>
</section>
<section id="section-2" class="level1" data-background-image="images/montypython.jpeg">
<h1 data-background-image="images/montypython.jpeg"></h1>
<h1 color="black" style="background-color: white">
Let’s do “real” simulations!
</h1>
<section id="there-are-many-pieces-of-simulation-software" class="level2">
<h2 class="anchored" data-anchor-id="there-are-many-pieces-of-simulation-software">There are many pieces of simulation software</h2>
<p><br></p>
<p>The most famous and widely used are <a href="https://messerlab.org/slim/">SLiM</a> and <a href="https://tskit.dev/msprime/docs/stable/intro.html">msprime</a>.</p>
<div class="fragment">
<p>They are both extremely powerful…</p>
</div>
<div class="fragment">
<p>… but they require a lot of programming knowledge and quite a lot of code to write non-trivial simulations (without bugs).</p>
</div>
<div class="fragment">
<p><br></p>
<center>
<strong>We will learn simulation concepts using <a href="http://www.slendr.net"><em>slendr</em></a>,<br>a convenient R interface to both SLiM and msprime.</strong>
</center>
</div>
</section>
</section>
<section id="slim" class="level1">
<h1>SLiM</h1>
<section id="section-3" class="level2">
<h2 class="anchored" data-anchor-id="section-3"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is SLiM?
</h2>
<div class="fragment">
<ul>
<li><strong>Forward-time simulator</strong></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>It’s fully programmable!</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Massive library of functions for:
<ul>
<li>Demographic events</li>
<li>Various mating systems</li>
<li>Selection, quantitative traits, …</li>
</ul></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>&gt; 700 pages long <a href="https://github.com/MesserLab/SLiM/releases/download/v3.7.1/SLiM_Manual.pdf">manual</a>!</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/sim_sketches.001.png" class="img-fluid figure-img" style="width:100.0%"></p>
<p></p><figcaption class="figure-caption">Modified from Alex Drummond</figcaption><p></p>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="slimgui-ide-for-slim" class="level2">
<h2 class="anchored" data-anchor-id="slimgui-ide-for-slim">SLiMgui – <a href="https://en.wikipedia.org/wiki/Integrated_development_environment">IDE</a> for SLiM</h2>
<center>
<img src="images/slimgui.png" class="img-fluid">
</center>
</section>
<section id="simple-neutral-simulation-in-slim" class="level2">
<h2 class="anchored" data-anchor-id="simple-neutral-simulation-in-slim">Simple neutral simulation in SLiM</h2>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>initialize() {
    // create a neutral mutation type
    initializeMutationType("m1", 0.5, "f", 0.0);
    // initialize 1Mb segment
    initializeGenomicElementType("g1", m1, 1.0);
    initializeGenomicElement(g1, 0, 999999);
    // set mutation rate and recombination rate of the segment
    initializeMutationRate(1e-8);
    initializeRecombinationRate(1e-8);
}

// create an ancestral population p1 of 10000 diploid individuals
1 early() { sim.addSubpop("p1", 10000); }

// in generation 1000, create two daughter populations p2 and p3
1000 early() {
    sim.addSubpopSplit("p2", 5000, p1);
    sim.addSubpopSplit("p3", 1000, p1);
}

// in generation 10000, stop the simulation and save 100 individuals
// from p2 and p3 to a VCF file
10000 late() {
    p2_subset = sample(p2.individuals, 100);
    p3_subset = sample(p3.individuals, 100);
    c(p2_subset, p3_subset).genomes.outputVCF("/tmp/slim_output.vcf.gz");
    sim.simulationFinished();
    catn("DONE!");
}</code></pre>
</div>
</div>
</section>
</section>
<section id="msprime" class="level1">
<h1>msprime</h1>
<section id="section-4" class="level2">
<h2 class="anchored" data-anchor-id="section-4"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is <a href="https://tskit.dev/msprime/docs/stable/intro.html">msprime</a>?
</h2>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/sim_sketches.001.png" class="img-fluid figure-img" style="width:100.0%"></p>
<p></p><figcaption class="figure-caption">Modified from Alex Drummond</figcaption><p></p>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="section-5" class="level2">
<h2 class="anchored" data-anchor-id="section-5"></h2>
<div class="columns">
<div class="column" style="width:60%;">
<h2 class="anchored">
What is <a href="https://tskit.dev/msprime/docs/stable/intro.html">msprime</a>?
</h2>
<div class="fragment">
<ul>
<li>A Python module for writing <strong>coalescent simulations</strong></li>
</ul>
</div>
<div class="fragment">
<ul>
<li>Extremely fast (genome-scale, population-scale data)</li>
</ul>
</div>
<div class="fragment">
<ul>
<li>You must know Python fairly well to build complex models</li>
</ul>
</div>
</div><div class="column" style="width:40%;">
<center>
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="images/sim_sketches.002.png" class="img-fluid figure-img" style="width:100.0%"></p>
<p></p><figcaption class="figure-caption">Modified from Alex Drummond</figcaption><p></p>
</figure>
</div>
</center>
</div>
</div>
</section>
<section id="simple-simulation-using-msprime" class="level2">
<h2 class="anchored" data-anchor-id="simple-simulation-using-msprime">Simple simulation using <em>msprime</em></h2>
<p>The following is basically the same model as the SLiM script earlier: &nbsp;</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>import msprime

demography = msprime.Demography()
demography.add_population(name="A", initial_size=10_000)
demography.add_population(name="B", initial_size=5_000)
demography.add_population(name="C", initial_size=1_000)
demography.add_population_split(time=1000, derived=["A", "B"], ancestral="C")

ts = msprime.sim_ancestry(
  sequence_length=10e6,
  recombination_rate=1e-8,
  samples={"A": 100, "B": 100},
  demography=demography
)</code></pre>
</div>
</div>
<p>source: <a href="https://tskit.dev/msprime/docs/stable/demography.html#demographic-models">link</a></p>
</section>
</section>
<section id="section-6" class="level1">
<h1></h1>
<center>
<p><img src="images/slendr_logo.png" class="img-fluid" style="width:30.0%"></p>
<p><br></p>
<h2 class="anchored" data-anchor-id="section-6">
<a href="www.slendr.net">www.slendr.net</a>
</h2>
</center>
<section id="why-a-new-package-spatial-simulations" class="level2">
<h2 class="anchored" data-anchor-id="why-a-new-package-spatial-simulations">Why a new package? – spatial simulations!</h2>
<div class="fragment">
<center>
<img src="images/animation.gif" class="img-fluid" style="width:70.0%">
</center>
</div>
</section>
<section id="why-a-new-package" class="level2">
<h2 class="anchored" data-anchor-id="why-a-new-package">Why a new package?</h2>
<div class="incremental">
<ul class="incremental">
<li><p>Most researchers are not expert programmers</p></li>
<li><p>All but the most trivial simulations require lots of code</p></li>
<li><p>90% <strong><citation needed=""></citation></strong> of simulations are basically the same!</p>
<ul class="incremental">
<li><p>create populations (splits and <span class="math inline">\(N_e\)</span> changes)</p></li>
<li><p>specify if/how they should mix (rates and times)</p></li>
<li><p>save output (VCF, EIGENSTRAT)</p></li>
</ul></li>
<li><p>Lot of code duplication across projects</p></li>
</ul>
</div>
<div class="fragment">
<center>
<strong><em>slendr</em> makes “standard” simulations trivial (for newbies <em>and</em> experts) and unlocks new kinds of spatial simulations</strong>
</center>
</div>
</section>
</section>
<section id="lets-get-started" class="level1">
<h1>Let’s get started</h1>
<section id="we-will-use-slendr-tidyverse" class="level2">
<h2 class="anchored" data-anchor-id="we-will-use-slendr-tidyverse">We will use <code>slendr</code> &amp; <code>tidyverse</code></h2>
<p><br></p>
<p>First run this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse) <span class="co"># table processing/filtering and plotting</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(slendr) <span class="co"># simulation and tree-sequence analysis</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'slendr'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:MASS':

    area</code></pre>
</div>
</div>
<center>
<em>(ignore the message about missing SLiM)</em>
</center>
<p><br></p>
<p>Then run this (and wait for the Python installation to finish!):</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="fu">setup_env</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<br>
<center>
<h3 class="anchored">
The entire lecture &amp; exercises will be in R!
</h3>
</center>
</section>
<section id="workaround-for-an-rstudio-bug" class="level2">
<h2 class="anchored" data-anchor-id="workaround-for-an-rstudio-bug">Workaround for an RStudio bug</h2>
<p>RStudio sometimes interferes with Python setup that we need for simulation. To fix this, go to <code>Tools</code> -&gt; <code>Global Options</code> in your RStudio and set the following options:</p>
<center>
<img src="images/rstudio_setting.png" class="img-fluid" style="width:40.0%">
</center>
</section>
</section>
<section id="section-7" class="level1">
<h1></h1>
<h1>
<em>slendr</em> workflow:<br><br>
</h1>
<h3 class="anchored">
Build models from simple components
</h3>
<br>
<h3 class="anchored">
Then simulate data from them
</h3>
<br>
<h3 class="anchored">
All within a single R script
</h3>
<p><br></p>
<section id="typical-steps" class="level2">
<h2 class="anchored" data-anchor-id="typical-steps">Typical steps</h2>
<p><br></p>
<ol type="1">
<li>creating populations</li>
<li>scheduling population splits</li>
<li>programming <span class="math inline">\(N_e\)</span> size changes</li>
<li>encoding gene-flow events</li>
<li>simulation sequence of a given size</li>
<li>computing statistics from simulated outputs</li>
</ol>
</section>
<section id="creating-a-population" class="level2">
<h2 class="anchored" data-anchor-id="creating-a-population">Creating a population</h2>
<p>A name, size and the time of appearance must be given:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p><br></p>
<p>Typing an object prints out a summary in the R console:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>pop1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop1 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 1: created as an ancestral population (N = 1000)</code></pre>
</div>
</div>
</div>
</section>
<section id="for-r-experts-what-are-the-population-objects" class="level2">
<h2 class="anchored" data-anchor-id="for-r-experts-what-are-the-population-objects">For R experts – what are the population objects?</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(pop1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "slendr"     "slendr_pop" "list"      </code></pre>
</div>
</div>
<div class="incremental">
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(pop1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "pop"  "time"</code></pre>
</div>
</div>
</div>
<div class="incremental">
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="fu">str</span>(pop1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>List of 2
 $ pop : chr "pop1"
 $ time: int 1
 - attr(*, "remove")= num -1
 - attr(*, "parent")= chr "ancestor"
 - attr(*, "map")= logi FALSE
 - attr(*, "history")=List of 1
  ..$ :'data.frame':    1 obs. of  8 variables:
  .. ..$ pop          : chr "pop1"
  .. ..$ event        : chr "split"
  .. ..$ time         : int 1
  .. ..$ N            : int 1000
  .. ..$ competition  : logi NA
  .. ..$ mating       : logi NA
  .. ..$ dispersal    : logi NA
  .. ..$ dispersal_fun: chr "normal"
 - attr(*, "class")= chr [1:3] "slendr" "slendr_pop" "list"</code></pre>
</div>
</div>
</div>
</section>
<section id="programming-population-splits" class="level2">
<h2 class="anchored" data-anchor-id="programming-population-splits">Programming population splits</h2>
<p>Splits are indicated by the <code>parent = &lt;pop&gt;</code> argument:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p><br></p>
<p>The split is reported in the “historical summary”:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a>pop2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop2 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 50: split from pop1</code></pre>
</div>
</div>
</div>
</section>
<section id="scheduling-resize-events-resize" class="level2">
<h2 class="anchored" data-anchor-id="scheduling-resize-events-resize">Scheduling resize events – <code>resize()</code></h2>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb37-1"><a href="#cb37-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb37-2"><a href="#cb37-2"></a>pop1_step <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop1, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1"></a>pop2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">50</span>, <span class="at">parent =</span> pop1)</span>
<span id="cb38-2"><a href="#cb38-2"></a>pop2_exp <span class="ot">&lt;-</span> <span class="fu">resize</span>(pop2, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="tidyverse-style-pipe-interface" class="level2">
<h2 class="anchored" data-anchor-id="tidyverse-style-pipe-interface">Tidyverse-style <a href="https://magrittr.tidyverse.org">pipe</a> interface</h2>
<p>Step size decrease:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">how =</span> <span class="st">"step"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<p>Exponential increase:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>) <span class="sc">%&gt;%</span></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">2000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>This accomplishes the same thing as the code on the previous slide, but it is a bit more “elegant”.</strong>
</center>
</section>
<section id="more-complex-full-model" class="level2">
<h2 class="anchored" data-anchor-id="more-complex-full-model">More complex full model</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>pop1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop1"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb41-2"><a href="#cb41-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-3"><a href="#cb41-3" aria-hidden="true" tabindex="-1"></a>pop2 <span class="ot">&lt;-</span></span>
<span id="cb41-4"><a href="#cb41-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop2"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">300</span>, <span class="at">parent =</span> pop1) <span class="sc">%&gt;%</span></span>
<span id="cb41-5"><a href="#cb41-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">100</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1000</span>)</span>
<span id="cb41-6"><a href="#cb41-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-7"><a href="#cb41-7" aria-hidden="true" tabindex="-1"></a>pop3 <span class="ot">&lt;-</span></span>
<span id="cb41-8"><a href="#cb41-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop3"</span>, <span class="at">N =</span> <span class="dv">1000</span>, <span class="at">time =</span> <span class="dv">400</span>, <span class="at">parent =</span> pop2) <span class="sc">%&gt;%</span></span>
<span id="cb41-9"><a href="#cb41-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">2500</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">800</span>)</span>
<span id="cb41-10"><a href="#cb41-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-11"><a href="#cb41-11" aria-hidden="true" tabindex="-1"></a>pop4 <span class="ot">&lt;-</span></span>
<span id="cb41-12"><a href="#cb41-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop4"</span>, <span class="at">N =</span> <span class="dv">1500</span>, <span class="at">time =</span> <span class="dv">500</span>, <span class="at">parent =</span> pop3) <span class="sc">%&gt;%</span></span>
<span id="cb41-13"><a href="#cb41-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">700</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1200</span>, <span class="at">end =</span> <span class="dv">2000</span>)</span>
<span id="cb41-14"><a href="#cb41-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb41-15"><a href="#cb41-15" aria-hidden="true" tabindex="-1"></a>pop5 <span class="ot">&lt;-</span></span>
<span id="cb41-16"><a href="#cb41-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop5"</span>, <span class="at">N =</span> <span class="dv">100</span>, <span class="at">time =</span> <span class="dv">600</span>, <span class="at">parent =</span> pop4) <span class="sc">%&gt;%</span></span>
<span id="cb41-17"><a href="#cb41-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">50</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">900</span>) <span class="sc">%&gt;%</span></span>
<span id="cb41-18"><a href="#cb41-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">250</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">1200</span>) <span class="sc">%&gt;%</span></span>
<span id="cb41-19"><a href="#cb41-19" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">1000</span>, <span class="at">how =</span> <span class="st">"exponential"</span>, <span class="at">time =</span> <span class="dv">1600</span>, <span class="at">end =</span> <span class="dv">2200</span>) <span class="sc">%&gt;%</span></span>
<span id="cb41-20"><a href="#cb41-20" aria-hidden="true" tabindex="-1"></a>  <span class="fu">resize</span>(<span class="at">N =</span> <span class="dv">400</span>, <span class="at">how =</span> <span class="st">"step"</span>, <span class="at">time =</span> <span class="dv">2400</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="remember-each-object-carries-its-history" class="level2">
<h2 class="anchored" data-anchor-id="remember-each-object-carries-its-history">Remember: each object carries its history</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a>pop5</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'population' object 
-------------------------- 
name: pop5 
non-spatial population
stays until the end of the simulation

population history overview:
  - time 600: split from pop4
  - time 900: resize from 100 to 50 individuals
  - time 1200: resize from 50 to 250 individuals
  - time 1600-2200: exponential resize from 250 to 1000 individuals
  - time 2400: resize from 1000 to 400 individuals</code></pre>
</div>
</div>
</section>
<section id="last-step-before-simulation-compilation" class="level2">
<h2 class="anchored" data-anchor-id="last-step-before-simulation-compilation">Last step before simulation: compilation</h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">list</span>(pop1, pop2, pop3, pop4, pop5),</span>
<span id="cb44-3"><a href="#cb44-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>,</span>
<span id="cb44-4"><a href="#cb44-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">simulation_length =</span> <span class="dv">3000</span></span>
<span id="cb44-5"><a href="#cb44-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong>Compilation takes a list of model components, performs internal consistency checks, returns a single model object.</strong>
</center>
</section>
<section id="model-summary" class="level2">
<h2 class="anchored" data-anchor-id="model-summary">Model summary</h2>
<p>Typing the compiled model prints a brief summary:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>model</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>slendr 'model' object 
--------------------- 
populations: pop1, pop2, pop3, pop4, pop5 
geneflow events: [no geneflow]
generation time: 1 
time direction: forward 
total running length: 3000 model time units
model type: non-spatial

configuration files in: /private/var/folders/d_/hblb15pd3b94rg0v35920wd80000gn/T/Rtmp3RrClz/file1610a50f34ba </code></pre>
</div>
</div>
<div class="incremental">
<p>The model is also saved to disk! (The location can be specified via <code>path =</code> argument to <code>compile_model()</code>):</p>
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code> [1] "checksums.tsv"       "description.txt"     "direction.txt"      
 [4] "generation_time.txt" "length.txt"          "orig_length.txt"    
 [7] "populations.tsv"     "ranges.rds"          "resizes.tsv"        
[10] "script.py"           "script.slim"        </code></pre>
</div>
</div>
</div>
</section>
<section id="model-visualization" class="level2">
<h2 class="anchored" data-anchor-id="model-visualization">Model visualization</h2>
<div class="cell" data-layout-align="center">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-30-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="simulating-data-finally" class="level2">
<h2 class="anchored" data-anchor-id="simulating-data-finally">Simulating data (finally…)</h2>
<p>We have the compiled <code>model</code>, how do we simulate data?</p>
<p><em>slendr</em> has two simulation engines already built-in:</p>
<ul>
<li>SLiM engine</li>
<li>msprime engine</li>
</ul>
<div class="incremental">
<center>
<strong>You don’t have to write any msprime/SLiM code!</strong>
</center>
</div>
<div class="incremental">
<p>Take a model object and use the built-in simulation engine:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="incremental">
<br>
<center>
<code>ts</code> is a so-called “tree sequence”
</center>
</div>
</section>
</section>
<section id="the-output-of-a-slendr-simulation-is-a-tree-sequence" class="level1">
<h1>The output of a <em>slendr</em> simulation is a <strong>tree sequence</strong></h1>
<section id="what-is-a-tree-sequence" class="level2">
<h2 class="anchored" data-anchor-id="what-is-a-tree-sequence">What is a tree sequence?</h2>
<center>
<img src="images/tree_sequence_diagram.png" class="img-fluid" style="width:90.0%">
</center>
<ul>
<li>A record of full genetic ancestry of a set of samples</li>
<li>An encoding of DNA sequence carried by those samples</li>
<li>An efficient analysis framework</li>
</ul>
<center>
<a href="https://tskit.dev/learn.html#what">from the <em>tskit</em> project</a>
</center>
</section>
</section>
<section id="why-a-tree-sequence" class="level1">
<h1>Why a tree sequence?</h1>
<section id="what-we-usually-have" class="level2">
<h2 class="anchored" data-anchor-id="what-we-usually-have">What we usually have</h2>
<center>
<img src="images/vcf_screenshot.png" class="img-fluid" style="width:90.0%">
</center>
</section>
<section id="what-we-usually-want" class="level2">
<h2 class="anchored" data-anchor-id="what-we-usually-want">What we usually <em>want</em></h2>
<p>(As full as possible) a representation of our samples’ history:</p>
<center>
<img src="images/tree_sequence_diagram.png" class="img-fluid">
</center>
<div class="fragment">
<center>
<strong>And this is exactly what tree sequences give us.</strong>
</center>
</div>
</section>
<section id="how-does-it-work" class="level2">
<h2 class="anchored" data-anchor-id="how-does-it-work">How does it work?</h2>
<p>This simulates 20 <span class="math inline">\(\times\)</span> 10000 chromosomes of 100 Mb.</p>
<p>In less than 30 seconds.</p>
<p>That’s a crazy amount of data!</p>
<center>
<strong>And it only requires 66.1 Mb of memory!</strong>
</center>
<div class="cell">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">population</span>(<span class="st">"pop"</span>, <span class="at">time =</span> <span class="dv">100000</span>, <span class="at">N =</span> <span class="dv">10000</span>) <span class="sc">%&gt;%</span></span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">compile_model</span>(<span class="at">generation_time =</span> <span class="dv">1</span>, <span class="at">direction =</span> <span class="st">"backward"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb50-4"><a href="#cb50-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(<span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span>
<span id="cb50-5"><a href="#cb50-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb50-6"><a href="#cb50-6" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │     390028║
╟───────────────┼───────────╢
║Sequence Length│  100000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │      20000║
╟───────────────┼───────────╢
║Total Size     │   65.8 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤═══════╤═════════╤════════════╗
║Table      │Rows   │Size     │Has Metadata║
╠═══════════╪═══════╪═════════╪════════════╣
║Edges      │1517108│ 46.3 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Individuals│  10000│273.5 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Migrations │      0│  8 Bytes│          No║
╟───────────┼───────┼─────────┼────────────╢
║Mutations  │      0│ 16 Bytes│          No║
╟───────────┼───────┼─────────┼────────────╢
║Nodes      │ 286124│  7.6 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Populations│      1│222 Bytes│         Yes║
╟───────────┼───────┼─────────┼────────────╢
║Provenances│      1│  1.3 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Sites      │      0│ 16 Bytes│          No║
╚═══════════╧═══════╧═════════╧════════════╝</code></pre>
</div>
</div>
</section>
<section id="how-does-this-work" class="level2">
<h2 class="anchored" data-anchor-id="how-does-this-work">How does this work?!</h2>
<div class="incremental">
<center>
<img src="images/tables.jpeg" class="img-fluid">
<center>
</center></center></div>
</section>
<section id="tree-sequence-tables-tskit-docs" class="level2">
<h2 class="anchored" data-anchor-id="tree-sequence-tables-tskit-docs">Tree-sequence tables (<a href="https://tskit.dev/tutorials/tables_and_editing.html">tskit docs</a> )</h2>
<div class="row">
<div class="columns">
<div class="column" style="width:60%;">
<p>A tree (sequence) can be represented by tables of:</p>
<ul>
<li>nodes (“chromosomes”)</li>
<li>edges (branches) between nodes</li>
<li>mutations on edges <br></li>
<li>individuals (who carry nodes)</li>
<li>populations</li>
</ul>
</div><div class="column" style="width:40%;">
<br>
<center>
<img src="images/tree_diagram.png" class="img-fluid">
</center>
</div>
</div>
</div>
<div class="fragment">
<center>
<h2 class="anchored">
<strong>A set of such tables is a tree sequence.</strong>
</h2>
</center>
</div>
</section>
<section id="tree-sequence-tables-in-practice" class="level2">
<h2 class="anchored" data-anchor-id="tree-sequence-tables-in-practice">Tree-sequence tables in practice</h2>
<div class="columns">
<div class="column" style="width:50%;">
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>Registered S3 method overwritten by 'ggtree':
  method      from 
  identify.gg ggfun</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>ggtree v3.2.1  For help: https://yulab-smu.top/treedata-book/

If you use ggtree in published research, please cite the most appropriate paper(s):

1. Guangchuang Yu. Using ggtree to visualize data on tree-like structures. Current Protocols in Bioinformatics. 2020, 69:e96. doi:10.1002/cpbi.96
2. Guangchuang Yu, Tommy Tsan-Yuk Lam, Huachen Zhu, Yi Guan. Two methods for mapping and visualizing associated data on phylogeny using ggtree. Molecular Biology and Evolution. 2018, 35(12):3041-3043. doi:10.1093/molbev/msy194
3. Guangchuang Yu, David Smith, Huachen Zhu, Yi Guan, Tommy Tsan-Yuk Lam. ggtree: an R package for visualization and annotation of phylogenetic trees with their covariates and other associated data. Methods in Ecology and Evolution. 2017, 8(1):28-36. doi:10.1111/2041-210X.12628</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'ggtree'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:tidyr':

    expand</code></pre>
</div>
<div class="cell-output-display">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div><div class="column" style="width:50%;">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_nodes</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(name, pop, node_id, time) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 4
  name  pop   node_id  time
  &lt;chr&gt; &lt;fct&gt;   &lt;int&gt; &lt;dbl&gt;
1 pop_1 pop         0     0
2 pop_1 pop         1     0
3 pop_2 pop         2     0</code></pre>
</div>
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_edges</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as_tibble</span>() <span class="sc">%&gt;%</span></span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(child_node_id, parent_node_id) <span class="sc">%&gt;%</span> <span class="fu">head</span>(<span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 3 × 2
  child_node_id parent_node_id
          &lt;int&gt;          &lt;int&gt;
1             0          22173
2             1          21227
3             2          22911</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="lets-get-back-to-the-model-we-defined-earlier" class="level2">
<h2 class="anchored" data-anchor-id="lets-get-back-to-the-model-we-defined-earlier">Let’s get back to the model we defined earlier</h2>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="simulating-a-tree-sequence-output" class="level2">
<h2 class="anchored" data-anchor-id="simulating-a-tree-sequence-output">Simulating a tree-sequence output</h2>
<p><br> By default, <code>msprime</code> function automatically loads the tree-sequence that the simulation produced in the background:</p>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>  model,</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">sequence_length =</span> <span class="fl">100e6</span>,</span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">recombination_rate =</span> <span class="fl">1e-8</span></span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p><br></p>
<p>If we type <code>ts</code> into an R console, we get…</p>
</div>
</section>
<section id="a-tree-sequence-content-summary" class="level2">
<h2 class="anchored" data-anchor-id="a-tree-sequence-content-summary">… a tree-sequence content summary</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │     140346║
╟───────────────┼───────────╢
║Sequence Length│  100000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │       9400║
╟───────────────┼───────────╢
║Total Size     │   24.5 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤══════╤═════════╤════════════╗
║Table      │Rows  │Size     │Has Metadata║
╠═══════════╪══════╪═════════╪════════════╣
║Edges      │565381│ 17.3 MiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Individuals│  4700│128.5 KiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Migrations │     0│  8 Bytes│          No║
╟───────────┼──────┼─────────┼────────────╢
║Mutations  │     0│ 16 Bytes│          No║
╟───────────┼──────┼─────────┼────────────╢
║Nodes      │106072│  2.8 MiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Populations│     5│383 Bytes│         Yes║
╟───────────┼──────┼─────────┼────────────╢
║Provenances│     1│  3.9 KiB│          No║
╟───────────┼──────┼─────────┼────────────╢
║Sites      │     0│ 16 Bytes│          No║
╚═══════════╧══════╧═════════╧════════════╝</code></pre>
</div>
</div>
<!-- ## Saving a tree-sequence permanently -->
<!-- <br> -->
<!-- We can either specify an output file: -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- ts <- msprime(model, sequence_length = 1e6, recombination_rate = 1e-8, -->
<!--               output = "<path>") -->
<!-- ``` -->
<!-- <br> -->
<!-- Alternatively, we can always save an already created tree-sequence object: -->
<!-- ```{r} -->
<!-- #| eval: false -->
<!-- ts_save(ts, file = "<path>") -->
<!-- ``` -->
</section>
</section>
<section id="what-can-we-do-with-this" class="level1">
<h1>What can we do with this?</h1>
<section id="r-interface-to-tskit" class="level2">
<h2 class="anchored" data-anchor-id="r-interface-to-tskit">R interface to tskit</h2>
<center>
<p><img src="images/slendr_tskit.png" class="img-fluid" style="width:120.0%"></p>
<p>This <a href="https://www.slendr.net/reference/index.html#tree-sequence-loading-and-processing">R interface</a> links to Python methods implemented in <a href="https://tskit.dev/tskit/docs/stable/python-api.html#statistics"><em>tskit</em></a>.</p>
</center>
</section>
<section id="here-is-the-magic" class="level2">
<h2 class="anchored" data-anchor-id="here-is-the-magic">Here is the magic</h2>
<p>Tree sequences make it possible to directly compute many quantities of interest <em>without going via conversion to a genotype table/VCF</em>!</p>
<center>
<img src="images/tree_sequence_diagram.png" class="img-fluid">
</center>
<div class="aside">
<p><a href="https://academic.oup.com/genetics/article/215/3/779/5930459">Ralph et al.&nbsp;(2020)</a></p>
</div>
</section>
</section>
<section id="how-do-we-use-this-in-practice" class="level1">
<h1>How do we use this in practice?</h1>
</section>
<section id="section-8" class="level1">
<h1></h1>
<h1>
To analyze tree sequences, we need to refer to “samples”
</h1>
<section id="extracting-sample-information" class="level2">
<h2 class="anchored" data-anchor-id="extracting-sample-information">Extracting sample information</h2>
<p>Each “sampled” individual in <em>slendr</em> has a symbolic name, a sampling time, and a population assignment:</p>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-38-1.png" class="img-fluid figure-img" width="768"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="extracting-sample-information-1" class="level2">
<h2 class="anchored" data-anchor-id="extracting-sample-information-1">Extracting sample information</h2>
<p>Each “sampled” individual in <em>slendr</em> has a symbolic name, a sampling time, and a population assignment:</p>
<div class="columns">
<div class="column" style="width:55%;">
<div class="cell" data-output-location="fragment">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 4,700 × 3
   name     time pop  
   &lt;chr&gt;   &lt;dbl&gt; &lt;chr&gt;
 1 pop1_1      0 pop1 
 2 pop1_2      0 pop1 
 3 pop1_3      0 pop1 
 4 pop1_4      0 pop1 
 5 pop1_5      0 pop1 
 6 pop1_6      0 pop1 
 7 pop1_7      0 pop1 
 8 pop1_8      0 pop1 
 9 pop1_9      0 pop1 
10 pop1_10     0 pop1 
# … with 4,690 more rows
# ℹ Use `print(n = ...)` to see more rows</code></pre>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:43%;">
<div class="cell" data-output-location="fragment">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">count</span>(pop)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 2
  pop       n
  &lt;chr&gt; &lt;int&gt;
1 pop1   1000
2 pop2    100
3 pop3   2500
4 pop4    700
5 pop5    400</code></pre>
</div>
</div>
</div>
</div>
</section>
<section id="analyzing-tree-sequences-with-slendr" class="level2">
<h2 class="anchored" data-anchor-id="analyzing-tree-sequences-with-slendr">Analyzing tree sequences with <em>slendr</em></h2>
<p>Let’s say we have the following model and we simulate a tree sequence from it.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1"></a>pop <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"pop"</span>, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">time =</span> <span class="dv">1</span>)</span>
<span id="cb67-2"><a href="#cb67-2"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(pop, <span class="at">generation_time =</span> <span class="dv">1</span>, <span class="at">simulation_length =</span> <span class="dv">10000</span>)</span>
<span id="cb67-3"><a href="#cb67-3"></a></span>
<span id="cb67-4"><a href="#cb67-4"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p><br></p>
<center>
<strong><em>slendr</em> provides a large <a href="https://www.slendr.net/reference/index.html#tree-sequence-loading-and-processing">library of functions</a> for computing population genetic statistics on tree sequences</strong>
</center>
</section>
<section id="example-allele-frequency-spectrum" class="level2">
<h2 class="anchored" data-anchor-id="example-allele-frequency-spectrum">Example: allele frequency spectrum</h2>
<div class="cell">

</div>
<div class="columns">
<div class="column" style="width:55%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb68-1"><a href="#cb68-1"></a><span class="co"># sample 5 individuals</span></span>
<span id="cb68-2"><a href="#cb68-2"></a><span class="co"># (i.e. 10 chromosomes)</span></span>
<span id="cb68-3"><a href="#cb68-3"></a>samples <span class="ot">&lt;-</span></span>
<span id="cb68-4"><a href="#cb68-4"></a>  <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb68-5"><a href="#cb68-5"></a>  <span class="fu">sample_n</span>(<span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb68-6"><a href="#cb68-6"></a>  <span class="fu">pull</span>(name)</span>
<span id="cb68-7"><a href="#cb68-7"></a></span>
<span id="cb68-8"><a href="#cb68-8"></a><span class="co"># compute allele frequency</span></span>
<span id="cb68-9"><a href="#cb68-9"></a><span class="co"># spectrum from the given set</span></span>
<span id="cb68-10"><a href="#cb68-10"></a><span class="co"># of individuals</span></span>
<span id="cb68-11"><a href="#cb68-11"></a>afs1 <span class="ot">&lt;-</span> <span class="fu">ts_afs</span>(</span>
<span id="cb68-12"><a href="#cb68-12"></a>  ts, <span class="fu">list</span>(samples),</span>
<span id="cb68-13"><a href="#cb68-13"></a>  <span class="at">mode =</span> <span class="st">"branch"</span>,</span>
<span id="cb68-14"><a href="#cb68-14"></a>  <span class="at">polarised =</span> <span class="cn">TRUE</span>,</span>
<span id="cb68-15"><a href="#cb68-15"></a>  <span class="at">span_normalise =</span> <span class="cn">TRUE</span></span>
<span id="cb68-16"><a href="#cb68-16"></a>)</span>
<span id="cb68-17"><a href="#cb68-17"></a></span>
<span id="cb68-18"><a href="#cb68-18"></a>afs1</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 40698.682 19734.265 13101.997 10044.762  8166.279  6667.231  5648.618
 [8]  4926.243  4415.790  3824.859</code></pre>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:43%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs1, <span class="at">type =</span> <span class="st">"b"</span>,</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"allele count bin"</span>,</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"frequency"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-44-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="but-wait-we-dont-have-any-mutations" class="level2">
<h2 class="anchored" data-anchor-id="but-wait-we-dont-have-any-mutations">But wait, we don’t have any mutations!</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a>ts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │     391699║
╟───────────────┼───────────╢
║Sequence Length│  100000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │      20000║
╟───────────────┼───────────╢
║Total Size     │   66.1 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤═══════╤═════════╤════════════╗
║Table      │Rows   │Size     │Has Metadata║
╠═══════════╪═══════╪═════════╪════════════╣
║Edges      │1524635│ 46.5 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Individuals│  10000│273.5 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Migrations │      0│  8 Bytes│          No║
╟───────────┼───────┼─────────┼────────────╢
║Mutations  │      0│ 16 Bytes│          No║
╟───────────┼───────┼─────────┼────────────╢
║Nodes      │ 286532│  7.7 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Populations│      1│222 Bytes│         Yes║
╟───────────┼───────┼─────────┼────────────╢
║Provenances│      1│  1.3 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Sites      │      0│ 16 Bytes│          No║
╚═══════════╧═══════╧═════════╧════════════╝</code></pre>
</div>
</div>
</section>
<section id="how-can-we-compute-statistics" class="level2">
<h2 class="anchored" data-anchor-id="how-can-we-compute-statistics">How can we compute statistics?</h2>
<p>There is a duality between mutations and branch lengths in trees (more <a href="https://tskit.dev/tskit/docs/stable/stats.html">here</a>).</p>
<center>
<p><img src="images/tree_sequence_diagram.png" class="img-fluid"></p>
<h3 class="anchored">
But what if we want mutations?
</h3>
</center>
</section>
<section id="coalescent-and-mutation-processes-can-be-decoupled" class="level2">
<h2 class="anchored" data-anchor-id="coalescent-and-mutation-processes-can-be-decoupled">Coalescent and mutation processes can be decoupled!</h2>
<center>
<p><img src="images/tree_sequence_diagram.png" class="img-fluid"></p>
<h3 class="anchored">
This means we can add mutations<br><em>after</em> the simulation.
</h3>
</center>
</section>
<section id="this-allows-efficient-massive-simulations-especially-with-slim" class="level2">
<h2 class="anchored" data-anchor-id="this-allows-efficient-massive-simulations-especially-with-slim">This allows efficient, massive simulations (especially with SLiM)</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a>ts_mutated <span class="ot">&lt;-</span> <span class="fu">ts_mutate</span>(</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>  ts,</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">mutation_rate =</span> <span class="fl">1e-8</span></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment">
<p>Or, with a shortcut:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>  model,</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">sequence_length =</span> <span class="fl">100e6</span>,</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">recombination_rate =</span> <span class="fl">1e-8</span></span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>) <span class="sc">%&gt;%</span></span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(</span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">mutation_rate =</span> <span class="fl">1e-8</span></span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a>  )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</div><div class="column" style="width:1%;">
<p>&nbsp;</p>
</div><div class="column" style="width:59%;">
<div class="fragment">
<div class="cell">
<div class="cell-output cell-output-stdout">
<pre><code>╔═══════════════════════════╗
║TreeSequence               ║
╠═══════════════╤═══════════╣
║Trees          │     393771║
╟───────────────┼───────────╢
║Sequence Length│  100000000║
╟───────────────┼───────────╢
║Time Units     │generations║
╟───────────────┼───────────╢
║Sample Nodes   │      20000║
╟───────────────┼───────────╢
║Total Size     │   91.4 MiB║
╚═══════════════╧═══════════╝
╔═══════════╤═══════╤═════════╤════════════╗
║Table      │Rows   │Size     │Has Metadata║
╠═══════════╪═══════╪═════════╪════════════╣
║Edges      │1534450│ 46.8 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Individuals│  10000│273.5 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Migrations │      0│  8 Bytes│          No║
╟───────────┼───────┼─────────┼────────────╢
║Mutations  │ 420999│ 14.9 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Nodes      │ 288261│  7.7 MiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Populations│      1│222 Bytes│         Yes║
╟───────────┼───────┼─────────┼────────────╢
║Provenances│      2│  2.1 KiB│          No║
╟───────────┼───────┼─────────┼────────────╢
║Sites      │ 420077│ 10.0 MiB│          No║
╚═══════════╧═══════╧═════════╧════════════╝</code></pre>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="example-allele-frequency-spectrum-1" class="level2">
<h2 class="anchored" data-anchor-id="example-allele-frequency-spectrum-1">Example: allele frequency spectrum</h2>
<div class="cell">

</div>
<div class="columns">
<div class="column" style="width:55%;">
<div class="cell">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sample 5 individuals</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a><span class="co"># (i.e. 10 chromosomes)</span></span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span></span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">sample_n</span>(<span class="dv">5</span>) <span class="sc">%&gt;%</span></span>
<span id="cb76-6"><a href="#cb76-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">pull</span>(name)</span>
<span id="cb76-7"><a href="#cb76-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-8"><a href="#cb76-8" aria-hidden="true" tabindex="-1"></a><span class="co"># compute allele frequency</span></span>
<span id="cb76-9"><a href="#cb76-9" aria-hidden="true" tabindex="-1"></a><span class="co"># spectrum from the given set</span></span>
<span id="cb76-10"><a href="#cb76-10" aria-hidden="true" tabindex="-1"></a><span class="co"># of individuals</span></span>
<span id="cb76-11"><a href="#cb76-11" aria-hidden="true" tabindex="-1"></a>afs2 <span class="ot">&lt;-</span> <span class="fu">ts_afs</span>(</span>
<span id="cb76-12"><a href="#cb76-12" aria-hidden="true" tabindex="-1"></a>  ts, <span class="fu">list</span>(samples),</span>
<span id="cb76-13"><a href="#cb76-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">polarised =</span> <span class="cn">TRUE</span></span>
<span id="cb76-14"><a href="#cb76-14" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb76-15"><a href="#cb76-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb76-16"><a href="#cb76-16" aria-hidden="true" tabindex="-1"></a>afs2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code> [1] 39657 19677 13307  9909  7845  6426  5755  4968  4419  4128</code></pre>
</div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:43%;">
<div class="fragment">
<div class="cell">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(afs2, <span class="at">type =</span> <span class="st">"b"</span>,</span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>     <span class="at">xlab =</span> <span class="st">"allele count bin"</span>,</span>
<span id="cb78-3"><a href="#cb78-3" aria-hidden="true" tabindex="-1"></a>     <span class="at">ylab =</span> <span class="st">"frequency"</span>)</span>
<span id="cb78-4"><a href="#cb78-4" aria-hidden="true" tabindex="-1"></a><span class="fu">lines</span>(afs1, <span class="at">type =</span> <span class="st">"b"</span>, <span class="at">col =</span> <span class="st">"red"</span>)</span>
<span id="cb78-5"><a href="#cb78-5" aria-hidden="true" tabindex="-1"></a><span class="fu">legend</span>(<span class="st">"topright"</span>, <span class="at">legend =</span> <span class="fu">c</span>(<span class="st">"mutation-based AFS"</span>, <span class="st">"theoretical branch-based AFS"</span>),</span>
<span id="cb78-6"><a href="#cb78-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">fill =</span> <span class="fu">c</span>(<span class="st">"black"</span>, <span class="st">"red"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-51-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</div>
</section>
<section id="what-we-have-learned-so-far" class="level2">
<h2 class="anchored" data-anchor-id="what-we-have-learned-so-far">What we have learned so far</h2>
<ol type="1">
<li>creating populations – <a href="https://www.slendr.net/reference/population.html"><code>population()</code></a></li>
<li>compiling models – <a href="https://www.slendr.net/reference/compile_model.html"><code>compile_model()</code></a></li>
<li>simulating tree sequences – <a href="https://www.slendr.net/reference/msprime.html"><code>msprime()</code></a></li>
<li>extracting samples – <a href="https://www.slendr.net/reference/ts_samples.html"><code>ts_samples()</code></a></li>
<li>computing AFS – <a href="https://www.slendr.net/reference/ts_afs.html"><code>ts_afs()</code></a></li>
</ol>
<p><br></p>
<center>
<h1>
<strong>Let’s put this to use!</strong>
</h1>
</center>
</section>
</section>
<section id="exercise-1" class="level1">
<h1>Exercise #1</h1>
<p><br></p>
<p>Open your RStudio session:</p>
<ul>
<li><p><strong>On-site participants</strong>: <a href="http://ricco.popgen.dk:3838">http://ricco.popgen.dk:3838</a></p></li>
<li><p><strong>Remote participants</strong>: <a href="http://cloud.popgen.dk:8787">http://cloud.popgen.dk:8787</a></p></li>
</ul>
<p>If you feel the servers are slow, you can also use your own laptop (setup instructions are <a href="https://github.com/bodkan/popgen2022#setting-up-your-local-machine">here</a>).</p>
<section id="workaround-for-an-rstudio-bug-1" class="level2">
<h2 class="anchored" data-anchor-id="workaround-for-an-rstudio-bug-1">Workaround for an RStudio bug</h2>
<p>RStudio sometimes interferes with Python setup that we need for simulation. To fix this, go to <code>Tools</code> -&gt; <code>Global Options</code> in your RStudio and set the following options:</p>
<center>
<img src="images/rstudio_setting.png" class="img-fluid" style="width:40.0%">
</center>
</section>
<section id="exercise-1-1" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1-1">Exercise #1</h2>
<div class="cell">

</div>
<p>Collaborator <a href="https://zelda.fandom.com/wiki/Hestu">Hestu</a> gave you AFS computed from 10 individuals of a sub-species of the <a href="https://www.zeldadungeon.net/wiki/Bushy-Tailed_Squirrel"><em>bushy-tailed squirrel</em></a> discovered in the Forest of Spirits in the land of Hyrule:</p>
<!-- ```{r} -->
<!-- #| echo: false -->
<!-- dput(as.vector(observed_afs)) -->
<!-- ``` -->
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>afs_observed <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="dv">2520</span>, <span class="dv">1449</span>, <span class="dv">855</span>, <span class="dv">622</span>, <span class="dv">530</span>, <span class="dv">446</span>, <span class="dv">365</span>, <span class="dv">334</span>, <span class="dv">349</span>, <span class="dv">244</span>, <span class="dv">264</span>, <span class="dv">218</span>, </span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a><span class="dv">133</span>, <span class="dv">173</span>, <span class="dv">159</span>, <span class="dv">142</span>, <span class="dv">167</span>, <span class="dv">129</span>, <span class="dv">125</span>, <span class="dv">143</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Fossil evidence is consistent with constant size of the population over 100,000 generations of its history. An Oracle you met in the Temple of Time said that the true squirrel <span class="math inline">\(N_e\)</span> has been between 1000 and 30000.</p>
<p>Use <em>slendr</em> to simulate history of this species. Use this to guess the likely value of squirrel’s <span class="math inline">\(N_e\)</span> given the observed AFS.</p>
</section>
<section id="exercise-1-hints" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1-hints">Exercise #1 – hints</h2>
<ol type="1">
<li><p>Write a function that gets <span class="math inline">\(N_e\)</span> as input and returns the AFS.</p></li>
<li><p>Find the <span class="math inline">\(N_e\)</span> value that will give the closest AFS to the one you got from Hestu. Use whatever method you’re comfortable with based on your programming experience:</p>
<ul>
<li><p>i ) Plot simulated AFS for different <span class="math inline">\(N_e\)</span> with the AFS and just eye-ball <span class="math inline">\(N_e\)</span> value that looks correct.</p></li>
<li><p>ii ) Simulate AFS across a grid of <span class="math inline">\(N_e\)</span> values and find the closest matching one (maybe use <a href="https://en.wikipedia.org/wiki/Mean_squared_error">mean-squared error</a>?)</p></li>
<li><p>iii ) Run a mini-Approximate Bayesian Computation, using the Oracle’s range of [10000, 30000] as a uniform prior.</p></li>
</ul></li>
</ol>
</section>
<section id="exercise-1-eye-balling-solution" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1-eye-balling-solution">Exercise #1 – eye-balling solution</h2>
<p><br><br></p>
<center>
<h3 class="anchored">
<a href="https://github.com/bodkan/popgen-2022/blob/main/solutions/exercise_1_simple.R">solution</a>
</h3>
</center>
</section>
<section id="exercise-1-simulations-on-a-grid" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1-simulations-on-a-grid">Exercise #1 – simulations on a grid</h2>
<p><br><br></p>
<center>
<h3 class="anchored">
<a href="https://github.com/bodkan/popgen-2022/blob/main/solutions/exercise_1_grid.R">solution</a>
</h3>
</center>
</section>
<section id="exercise-1-abc-inference" class="level2">
<h2 class="anchored" data-anchor-id="exercise-1-abc-inference">Exercise #1 – ABC inference</h2>
<p><br><br></p>
<center>
<h3 class="anchored">
<a href="https://github.com/bodkan/popgen-2022/blob/main/solutions/exercise_1_abc.R">solution</a>
</h3>
</center>
</section>
<section id="exercise-2" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2">Exercise #2</h2>
<div class="columns">
<div class="column" style="width:70%;">
<p>The squirrels have split into three species with different demographic histories (<em>s1</em>, <em>s2</em>, <em>s3</em> – model on the right). Species <em>s2</em> and <em>s3</em> are daughter populations which split in generation 2 from the original species <em>s1</em>.</p>
<p>Help the Oracle predict the future shape of the AFS of the squirrels after 10000 generations, assuming starting <span class="math inline">\(N_e\)</span> = 6543? Species <em>s1</em> will remain constant, species <em>s2</em> will expand 3X, species <em>s3</em> will get 3X smaller.</p>
</div><div class="column" style="width:30%;">
<div class="cell">
<div class="cell-output-display">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-54-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="exercise-2-solution" class="level2">
<h2 class="anchored" data-anchor-id="exercise-2-solution">Exercise #2 – solution</h2>
<p><br><br></p>
<center>
<h3 class="anchored">
<a href="https://github.com/bodkan/popgen-2022/blob/main/solutions/exercise_2.R">link</a>
</h3>
</center>
</section>
</section>
<section id="what-else-can-slendr-do" class="level1">
<h1>What else can <em>slendr</em> do?</h1>
<section id="gene-flow-events" class="level2">
<h2 class="anchored" data-anchor-id="gene-flow-events">Gene flow events</h2>
<p>Gene flow is programmed using the <code>gene_flow()</code> function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb80"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb80-1"><a href="#cb80-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> pop1, <span class="at">to =</span> pop2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p><br></p>
<p>Multiple gene-flow events can be gathered in a list:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">list</span>(</span>
<span id="cb81-2"><a href="#cb81-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> pop1, <span class="at">to =</span> pop2, <span class="at">start =</span> <span class="dv">500</span>, <span class="at">end =</span> <span class="dv">600</span>, <span class="at">rate =</span> <span class="fl">0.13</span>),</span>
<span id="cb81-3"><a href="#cb81-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">gene_flow</span>(<span class="at">from =</span> pop2, <span class="at">to =</span> pop1, <span class="at">start =</span> <span class="dv">700</span>, <span class="at">end =</span> <span class="dv">750</span>, <span class="at">rate =</span> <span class="fl">0.1</span>)</span>
<span id="cb81-4"><a href="#cb81-4" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="incremental">
<br>
<center>
<strong><code>gene_flow()</code> checks admixture events for consistency</strong>
</center>
<!-- ## Behind the scenes -->
<!-- <br> -->
<!-- The output of `gene_flow()` is nothing but a data frame: -->
<!-- ```{r} -->
<!-- gf -->
<!-- ``` -->
<!-- However, the function does lots of consistency checks behind the scenes, so its better to always use it. -->
</div>
</section>
<section id="lets-build-another-toy-model" class="level2">
<h2 class="anchored" data-anchor-id="lets-build-another-toy-model">Let’s build another toy model</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb82"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb82-1"><a href="#cb82-1" aria-hidden="true" tabindex="-1"></a>o <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"o"</span>, <span class="at">time =</span> <span class="dv">1</span>, <span class="at">N =</span> <span class="dv">100</span>)</span>
<span id="cb82-2"><a href="#cb82-2" aria-hidden="true" tabindex="-1"></a>c <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"c"</span>, <span class="at">time =</span> <span class="dv">2500</span>, <span class="at">N =</span> <span class="dv">500</span>, <span class="at">parent =</span> o)</span>
<span id="cb82-3"><a href="#cb82-3" aria-hidden="true" tabindex="-1"></a>a <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"a"</span>, <span class="at">time =</span> <span class="dv">3000</span>, <span class="at">N =</span> <span class="dv">2000</span>, <span class="at">parent =</span> c)</span>
<span id="cb82-4"><a href="#cb82-4" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"b"</span>, <span class="at">time =</span> <span class="dv">3500</span>, <span class="at">N =</span> <span class="dv">4000</span>, <span class="at">parent =</span> a)</span>
<span id="cb82-5"><a href="#cb82-5" aria-hidden="true" tabindex="-1"></a>x1 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"x1"</span>, <span class="at">time =</span> <span class="dv">3800</span>, <span class="at">N =</span> <span class="dv">8000</span>, <span class="at">parent =</span> c)</span>
<span id="cb82-6"><a href="#cb82-6" aria-hidden="true" tabindex="-1"></a>x2 <span class="ot">&lt;-</span> <span class="fu">population</span>(<span class="st">"x2"</span>, <span class="at">time =</span> <span class="dv">4000</span>, <span class="at">N =</span> <span class="dv">10000</span>, <span class="at">parent =</span> x1)</span>
<span id="cb82-7"><a href="#cb82-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-8"><a href="#cb82-8" aria-hidden="true" tabindex="-1"></a>gf <span class="ot">&lt;-</span> <span class="fu">gene_flow</span>(<span class="at">from =</span> b, <span class="at">to =</span> x1, <span class="at">start =</span> <span class="dv">5500</span>, <span class="at">end =</span> <span class="dv">6000</span>, <span class="at">rate =</span> <span class="fl">0.3</span>)</span>
<span id="cb82-9"><a href="#cb82-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-10"><a href="#cb82-10" aria-hidden="true" tabindex="-1"></a>model <span class="ot">&lt;-</span> <span class="fu">compile_model</span>(</span>
<span id="cb82-11"><a href="#cb82-11" aria-hidden="true" tabindex="-1"></a>  <span class="at">populations =</span> <span class="fu">list</span>(a, b, x1, x2, c, o), <span class="at">gene_flow =</span> gf,</span>
<span id="cb82-12"><a href="#cb82-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">generation_time =</span> <span class="dv">1</span>, <span class="at">simulation_length =</span> <span class="dv">7000</span></span>
<span id="cb82-13"><a href="#cb82-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb82-14"><a href="#cb82-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-15"><a href="#cb82-15" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb82-16"><a href="#cb82-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span>
<span id="cb82-17"><a href="#cb82-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb82-18"><a href="#cb82-18" aria-hidden="true" tabindex="-1"></a><span class="fu">plot_model</span>(model, <span class="at">proportions =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-57-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">

</div>
</section>
<section id="diversity-ts_diversity" class="level2">
<h2 class="anchored" data-anchor-id="diversity-ts_diversity">Diversity – <code>ts_diversity()</code></h2>
<div class="columns">
<div class="column" style="width:48%;">
<p>Extract a list of lists of individuals’ names for each population:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a>samples <span class="ot">&lt;-</span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span></span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">split</span>(., .<span class="sc">$</span>pop) <span class="sc">%&gt;%</span></span>
<span id="cb83-4"><a href="#cb83-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">lapply</span>(pull, <span class="st">"name"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div><div class="column" style="width:2%;">
<p>&nbsp;</p>
</div><div class="column" style="width:50%;">
<div class="fragment">
<p>Compute diversity in each population:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb84"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb84-1"><a href="#cb84-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_diversity</span>(ts, samples) <span class="sc">%&gt;%</span></span>
<span id="cb84-2"><a href="#cb84-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(diversity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 2
  set    diversity
  &lt;chr&gt;      &lt;dbl&gt;
1 o     0.00000408
2 c     0.0000204 
3 a     0.0000538 
4 b     0.0000691 
5 x2    0.0000726 
6 x1    0.0000773 </code></pre>
</div>
</div>
</div>
</div>
</div>
<center>
<strong>Basically every <code>ts_&lt;...&gt;()</code> function of <em>slendr</em> accepts a tree-sequence object as its first argument.</strong>
</center>
</section>
</section>
<section id="exercise-3" class="level1">
<h1>Exercise #3</h1>
<section id="program-the-following-model-in-slendr" class="level2">
<h2 class="anchored" data-anchor-id="program-the-following-model-in-slendr">Program the following model in <em>slendr</em></h2>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
</div>
<div class="cell" data-layout-align="center">
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-62-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="parameter-values" class="level2">
<h2 class="anchored" data-anchor-id="parameter-values">Parameter values</h2>
<ul>
<li><p>“CH” outgroup at time 7 Mya (<span class="math inline">\(N_e\)</span> = 7k)</p></li>
<li><p>“AFR” splitting from “CH” at 6 Mya (<span class="math inline">\(N_e\)</span> = 10k)</p></li>
<li><p>“ALT” splitting from “AFR” at 700 kya (<span class="math inline">\(N_e\)</span> = 500)</p></li>
<li><p>“VI” splitting from “ALT” at 120 kya (<span class="math inline">\(N_e\)</span> = 1k)</p></li>
<li><p>“EUR” splitting from “AFR” at 70 kya (<span class="math inline">\(N_e\)</span> = 5k)</p></li>
<li><p>gene flow from “VI” to “EUR” at 3% over 50-40 kya</p></li>
</ul>
<p>Then simulate 100Mb sequence with <code>msprime()</code>, using recombination rate 1e-8 per bp per generation.</p>
</section>
<section id="exercise-3-solution" class="level2">
<h2 class="anchored" data-anchor-id="exercise-3-solution">Exercise #3 – solution</h2>
<p><br><br></p>
<center>
<h3 class="anchored">
<a href="https://github.com/bodkan/popgen-2022/blob/main/solutions/exercise_3.R">link</a>
</h3>
</center>
</section>
</section>
<section id="exercise-4" class="level1">
<h1>Exercise #4</h1>
<section id="diversity-and-divergence" class="level2">
<h2 class="anchored" data-anchor-id="diversity-and-divergence">Diversity and divergence</h2>
<div class="columns">
<div class="column" style="width:70%;">
<p>Use <code>ts_diversity()</code> to compute diversity in each simulated population (<code>CH</code>, <code>AFR</code>, …).</p>
<p>Does the result match the expectation given demographic history?</p>
<p>What about divergence between all pairs of populations (<code>ts_divergence()</code>)? Do your results recapitulate phylogenetic relationships? How about <code>ts_f3()</code> or <code>ts_fst()</code>?</p>
</div><div class="column" style="width:30%;">
<div class="cell">
<div class="cell-output-display">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-63-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="exercise-4-solution-diversity" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4-solution-diversity">Exercise #4 – solution (diversity)</h2>
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">50e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span>
<span id="cb87-4"><a href="#cb87-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-5"><a href="#cb87-5" aria-hidden="true" tabindex="-1"></a>sample_names <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">split</span>(., .<span class="sc">$</span>pop) <span class="sc">%&gt;%</span> <span class="fu">lapply</span>(pull, <span class="st">"name"</span>)</span>
<span id="cb87-6"><a href="#cb87-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb87-7"><a href="#cb87-7" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_diversity</span>(ts, <span class="at">sample_sets =</span> sample_names) <span class="sc">%&gt;%</span> <span class="fu">arrange</span>(diversity)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 5 × 2
  set   diversity
  &lt;chr&gt;     &lt;dbl&gt;
1 ALT   0.0000198
2 VI    0.0000378
3 CH    0.000282 
4 EUR   0.000369 
5 AFR   0.000390 </code></pre>
</div>
</div>
</section>
<section id="exercise-4-solution-divergence" class="level2">
<h2 class="anchored" data-anchor-id="exercise-4-solution-divergence">Exercise #4 – solution (divergence)</h2>
<div class="columns">
<div class="column" style="width:40%;">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a>divergence <span class="ot">&lt;-</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_divergence</span>(ts, sample_names) <span class="sc">%&gt;%</span></span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">arrange</span>(divergence)</span>
<span id="cb89-4"><a href="#cb89-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb89-5"><a href="#cb89-5" aria-hidden="true" tabindex="-1"></a>divergence</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 10 × 3
   x     y     divergence
   &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt;
 1 ALT   VI      0.000102
 2 AFR   EUR     0.000449
 3 EUR   VI      0.000853
 4 ALT   EUR     0.000854
 5 AFR   ALT     0.000870
 6 AFR   VI      0.000871
 7 AFR   CH      0.00426 
 8 CH    EUR     0.00426 
 9 ALT   CH      0.00427 
10 CH    VI      0.00427 </code></pre>
</div>
</div>
</div><div class="column" style="width:60%;">
<div class="cell">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a>divergence <span class="sc">%&gt;%</span></span>
<span id="cb91-2"><a href="#cb91-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">mutate</span>(<span class="at">pair =</span> <span class="fu">paste</span>(x, <span class="st">"-"</span>, y),</span>
<span id="cb91-3"><a href="#cb91-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">pair =</span> <span class="fu">fct_reorder</span>(pair, divergence)) <span class="sc">%&gt;%</span></span>
<span id="cb91-4"><a href="#cb91-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(pair, divergence)) <span class="sc">+</span></span>
<span id="cb91-5"><a href="#cb91-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>() <span class="sc">+</span></span>
<span id="cb91-6"><a href="#cb91-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">""</span>, <span class="at">y =</span> <span class="st">"divergence"</span>) <span class="sc">+</span></span>
<span id="cb91-7"><a href="#cb91-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"Pairwise genetic divergence between populations"</span>) <span class="sc">+</span></span>
<span id="cb91-8"><a href="#cb91-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-66-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div>
</div>
</section>
<section id="sampling-ancient-dna-time-series" class="level2">
<h2 class="anchored" data-anchor-id="sampling-ancient-dna-time-series">Sampling “ancient DNA” time-series</h2>
<p>By default, <em>slendr</em> records every individual living at the end of the simulation. Sampling can be also scheduled explicitly:</p>
<div class="fragment">
<div class="cell">

</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb92"><pre class="sourceCode numberSource r number-lines code-with-copy"><code class="sourceCode r"><span id="cb92-1"><a href="#cb92-1"></a>x_schedule <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(</span>
<span id="cb92-2"><a href="#cb92-2"></a>  model, <span class="at">times =</span> <span class="fu">seq</span>(<span class="at">from =</span> <span class="dv">7000</span>, <span class="at">to =</span> <span class="dv">4000</span>, <span class="at">by =</span> <span class="sc">-</span><span class="dv">100</span>),</span>
<span id="cb92-3"><a href="#cb92-3"></a>  <span class="fu">list</span>(x1, <span class="dv">5</span>), <span class="fu">list</span>(x2, <span class="dv">5</span>)</span>
<span id="cb92-4"><a href="#cb92-4"></a>)</span>
<span id="cb92-5"><a href="#cb92-5"></a></span>
<span id="cb92-6"><a href="#cb92-6"></a>abco_schedule <span class="ot">&lt;-</span> <span class="fu">schedule_sampling</span>(</span>
<span id="cb92-7"><a href="#cb92-7"></a>  model, <span class="at">times =</span> <span class="dv">7000</span>,</span>
<span id="cb92-8"><a href="#cb92-8"></a>  <span class="fu">list</span>(a, <span class="dv">1</span>), <span class="fu">list</span>(b, <span class="dv">1</span>), <span class="fu">list</span>(c, <span class="dv">1</span>), <span class="fu">list</span>(o, <span class="dv">1</span>)</span>
<span id="cb92-9"><a href="#cb92-9"></a>)</span>
<span id="cb92-10"><a href="#cb92-10"></a></span>
<span id="cb92-11"><a href="#cb92-11"></a>schedule <span class="ot">&lt;-</span> <span class="fu">rbind</span>(x_schedule, abco_schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
<div class="fragment">
<p>The schedule can be used in an <code>msprime()</code> call like this:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>,</span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">samples =</span> schedule)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</div>
</section>
<section id="computing-f_4-statistic-ts_f4" class="level2">
<h2 class="anchored" data-anchor-id="computing-f_4-statistic-ts_f4">Computing <span class="math inline">\(f_4\)</span> statistic – <code>ts_f4()</code></h2>
<div class="columns">
<div class="column" style="width:60%;">
<div class="cell">
<div class="cell-output-display">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-70-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</div><div class="column" style="width:40%;">
<p>Having simulated data from this model, can we detect gene flow from <em>b</em> into <em>x1</em>?</p>
</div>
</div>
</section>
<section id="computing-f_4-statistic-ts_f4-1" class="level2">
<h2 class="anchored" data-anchor-id="computing-f_4-statistic-ts_f4-1">Computing <span class="math inline">\(f_4\)</span> statistic – <code>ts_f4()</code></h2>
<p>Simulate tree sequence, add mutations on it:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb94"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb94-1"><a href="#cb94-1" aria-hidden="true" tabindex="-1"></a>ts <span class="ot">&lt;-</span> <span class="fu">msprime</span>(model, <span class="at">samples =</span> schedule,</span>
<span id="cb94-2"><a href="#cb94-2" aria-hidden="true" tabindex="-1"></a>              <span class="at">sequence_length =</span> <span class="fl">100e6</span>, <span class="at">recombination_rate =</span> <span class="fl">1e-8</span>) <span class="sc">%&gt;%</span></span>
<span id="cb94-3"><a href="#cb94-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ts_mutate</span>(<span class="at">mutation_rate =</span> <span class="fl">1e-8</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="incremental">
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f4</span>(ts, <span class="st">"x1_1"</span>, <span class="st">"x2_1"</span>, <span class="st">"b_1"</span> , <span class="st">"o_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  W     X     Y     Z                f4
  &lt;chr&gt; &lt;chr&gt; &lt;chr&gt; &lt;chr&gt;         &lt;dbl&gt;
1 x1_1  x2_1  b_1   o_1   -0.0000000275</code></pre>
</div>
</div>
</div>
<div class="incremental">
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f4</span>(ts, <span class="st">"x1_100"</span>, <span class="st">"x2_100"</span>, <span class="st">"b_1"</span> , <span class="st">"o_1"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 1 × 5
  W      X      Y     Z             f4
  &lt;chr&gt;  &lt;chr&gt;  &lt;chr&gt; &lt;chr&gt;      &lt;dbl&gt;
1 x1_100 x2_100 b_1   o_1   0.00000103</code></pre>
</div>
</div>
</div>
</section>
<section id="computing-f_4-statistic-for-all-x-individuals" class="level2">
<h2 class="anchored" data-anchor-id="computing-f_4-statistic-for-all-x-individuals">Computing <span class="math inline">\(f_4\)</span> statistic for all <code>x</code> individuals</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract information about samples from populations x1 and x2</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>x_inds <span class="ot">&lt;-</span> <span class="fu">ts_samples</span>(ts) <span class="sc">%&gt;%</span> <span class="fu">filter</span>(pop <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"x1"</span>, <span class="st">"x2"</span>))</span>
<span id="cb99-3"><a href="#cb99-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-4"><a href="#cb99-4" aria-hidden="true" tabindex="-1"></a><span class="co"># create a vector of individual's names</span></span>
<span id="cb99-5"><a href="#cb99-5" aria-hidden="true" tabindex="-1"></a><span class="co"># (used as input in ts_&lt;...&gt;() functions</span></span>
<span id="cb99-6"><a href="#cb99-6" aria-hidden="true" tabindex="-1"></a>x_names <span class="ot">&lt;-</span> x_inds<span class="sc">$</span>name</span>
<span id="cb99-7"><a href="#cb99-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-8"><a href="#cb99-8" aria-hidden="true" tabindex="-1"></a><span class="co"># iterate over all sample names, compute</span></span>
<span id="cb99-9"><a href="#cb99-9" aria-hidden="true" tabindex="-1"></a><span class="co"># f4(X, C; B, O)</span></span>
<span id="cb99-10"><a href="#cb99-10" aria-hidden="true" tabindex="-1"></a>f4_result<span class="ot">&lt;-</span> <span class="fu">map_dfr</span>(</span>
<span id="cb99-11"><a href="#cb99-11" aria-hidden="true" tabindex="-1"></a>  x_names,</span>
<span id="cb99-12"><a href="#cb99-12" aria-hidden="true" tabindex="-1"></a>  <span class="cf">function</span>(w) <span class="fu">ts_f4</span>(ts, <span class="at">W =</span> w, <span class="at">X =</span> <span class="st">"c_1"</span>, <span class="at">Y =</span> <span class="st">"b_1"</span> , <span class="at">Z =</span> <span class="st">"o_1"</span>)</span>
<span id="cb99-13"><a href="#cb99-13" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb99-14"><a href="#cb99-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb99-15"><a href="#cb99-15" aria-hidden="true" tabindex="-1"></a><span class="co"># add the f4 value as a column to the sample information table</span></span>
<span id="cb99-16"><a href="#cb99-16" aria-hidden="true" tabindex="-1"></a>x_inds<span class="sc">$</span>f4 <span class="ot">&lt;-</span> f4_result<span class="sc">$</span>f4</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="computing-f_4-statistic-for-all-x-individuals-1" class="level2">
<h2 class="anchored" data-anchor-id="computing-f_4-statistic-for-all-x-individuals-1">Computing <span class="math inline">\(f_4\)</span> statistic for all <code>x</code> individuals</h2>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(x_inds) <span class="sc">+</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_rect</span>(<span class="fu">aes</span>(<span class="at">xmin =</span> <span class="dv">5500</span>, <span class="at">xmax =</span> <span class="dv">6000</span>, <span class="at">ymin =</span> <span class="sc">-</span><span class="cn">Inf</span>, <span class="at">ymax =</span> <span class="cn">Inf</span>), <span class="at">alpha =</span> <span class="fl">0.5</span>, <span class="at">fill =</span> <span class="st">"gray"</span>) <span class="sc">+</span></span>
<span id="cb100-3"><a href="#cb100-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_jitter</span>(<span class="fu">aes</span>(time, f4, <span class="at">color =</span> pop)) <span class="sc">+</span></span>
<span id="cb100-4"><a href="#cb100-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_line</span>(<span class="at">data =</span> . <span class="sc">%&gt;%</span> <span class="fu">group_by</span>(pop, time) <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="at">f4 =</span> <span class="fu">mean</span>(f4)),</span>
<span id="cb100-5"><a href="#cb100-5" aria-hidden="true" tabindex="-1"></a>            <span class="fu">aes</span>(time, f4, <span class="at">color =</span> pop), <span class="at">size =</span> <span class="dv">2</span>) <span class="sc">+</span></span>
<span id="cb100-6"><a href="#cb100-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_hline</span>(<span class="at">yintercept =</span> <span class="dv">0</span>, <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">size =</span> <span class="dv">1</span>) <span class="sc">+</span></span>
<span id="cb100-7"><a href="#cb100-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output cell-output-stderr">
<pre><code>`summarise()` has grouped output by 'pop'. You can override using the `.groups`
argument.</code></pre>
</div>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-75-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="exercise-5" class="level1">
<h1>Exercise #5</h1>
<section id="neanderthal-ancestry-trajectory" class="level2">
<h2 class="anchored" data-anchor-id="neanderthal-ancestry-trajectory">Neanderthal ancestry trajectory</h2>
<ol type="1">
<li><p>Take your model of Neanderthal introgression</p></li>
<li><p>Implement temporal sampling:</p></li>
</ol>
<ul>
<li>one “European” every 1000 yrs between 40 kya and today</li>
<li>“Altai” (70 ky old) and “Vindija” individuals (40 ky old)</li>
</ul>
<ol start="3" type="1">
<li>Compute <span class="math inline">\(f_4\)</span>-ratio statistic using:</li>
</ol>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ts_f4ratio</span>(ts, <span class="at">X =</span> <span class="st">'&lt;vector of "European" individuals&gt;'</span>,</span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>           <span class="at">A =</span> <span class="st">'&lt;"Altai"&gt;'</span>, <span class="at">C =</span> <span class="st">'&lt;"African"&gt;'</span>, <span class="at">O =</span> <span class="st">'&lt;"Chimp"&gt;'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<ol start="4" type="1">
<li>Plot the estimated trajectory of Neanderthal ancestry in Europe over time.</li>
</ol>
</section>
</section>
<section id="exercise-5-solution" class="level1">
<h1>Exercise #5 – solution</h1>
<center>
<h3 class="anchored" data-anchor-id="exercise-5-solution">
<a href="https://github.com/bodkan/popgen-2022/blob/main/solutions/exercise_5.R">solution</a>
</h3>
</center>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
</div>
</section>
<section id="tree-sequence-simplification" class="level1">
<h1>Tree-sequence simplification</h1>
<section id="sometimes-a-tree-sequence-is-too-big" class="level2">
<h2 class="anchored" data-anchor-id="sometimes-a-tree-sequence-is-too-big">Sometimes a tree sequence is too big</h2>
<div class="cell">
<div class="cell-output cell-output-stderr">
<pre><code>The interface to all required Python modules has been activated.</code></pre>
</div>
<div class="cell-output-display">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-78-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<center>
<strong>Simplification reduces the genealogies only to those nodes that form ancestors of a desired set of samples.</strong>
</center>
</section>
<section id="meet-ts_simplify" class="level2">
<h2 class="anchored" data-anchor-id="meet-ts_simplify">Meet <code>ts_simplify()</code></h2>
<p><br></p>
<div class="cell">
<div class="sourceCode cell-code" id="cb105"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb105-1"><a href="#cb105-1" aria-hidden="true" tabindex="-1"></a>ts_small <span class="ot">&lt;-</span> <span class="fu">ts_simplify</span>(</span>
<span id="cb105-2"><a href="#cb105-2" aria-hidden="true" tabindex="-1"></a>  ts,</span>
<span id="cb105-3"><a href="#cb105-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">simplify_to =</span> <span class="fu">c</span>(</span>
<span id="cb105-4"><a href="#cb105-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"CH_1"</span>,</span>
<span id="cb105-5"><a href="#cb105-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"AFR_1"</span>, <span class="st">"AFR_2"</span>, <span class="st">"AFR_3"</span>,</span>
<span id="cb105-6"><a href="#cb105-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"ALT_1"</span>, <span class="st">"VI_1"</span>,</span>
<span id="cb105-7"><a href="#cb105-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"EUR_1"</span>, <span class="st">"EUR_2"</span>, <span class="st">"EUR_3"</span>, <span class="st">"EUR_4"</span>, <span class="st">"EUR_5"</span></span>
<span id="cb105-8"><a href="#cb105-8" aria-hidden="true" tabindex="-1"></a>  )</span>
<span id="cb105-9"><a href="#cb105-9" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<!-- ```{r} -->
<!-- # library(ggtree) -->
<!-- # pdf("trees.pdf") -->
<!-- # for (i in seq_len(ts_small$num_trees)) { -->
<!-- #   tree <- ts_phylo(ts_small, i, quiet = TRUE) -->
<!-- # nodes <- ts_nodes(tree) %>% -->
<!-- #   as_tibble %>% -->
<!-- #   dplyr::select(node = phylo_id, pop) -->
<!-- #  -->
<!-- # p_tree <- ggtree(tree) %<+% nodes + -->
<!-- #   geom_tiplab(aes(color = pop, fill = pop)) + -->
<!-- #   guides(color = "none") + -->
<!-- #   ggtitle(i) + -->
<!-- #   scale_x_continuous(limits = c(-7e6, 900e3)) -->
<!-- # print(revts(p_tree)) -->
<!-- # } -->
<!-- # dev.off() -->
<!-- ``` -->
<p><br></p>
<center>
<strong>Only the coalescent nodes of genealogies involving<br>these samples will be retained.</strong>
</center>
</section>
<section id="extracting-a-tree-1734" class="level2">
<h2 class="anchored" data-anchor-id="extracting-a-tree-1734">Extracting a tree #1734</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">ts_phylo</span>(ts_small, <span class="dv">1734</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb107"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb107-1"><a href="#cb107-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this needs another R package</span></span>
<span id="cb107-2"><a href="#cb107-2" aria-hidden="true" tabindex="-1"></a><span class="co"># simply using the ape package and running `plot(tree)` will also work</span></span>
<span id="cb107-3"><a href="#cb107-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb107-4"><a href="#cb107-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-5"><a href="#cb107-5" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> <span class="fu">ts_nodes</span>(tree) <span class="sc">%&gt;%</span></span>
<span id="cb107-6"><a href="#cb107-6" aria-hidden="true" tabindex="-1"></a>  as_tibble <span class="sc">%&gt;%</span></span>
<span id="cb107-7"><a href="#cb107-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">node =</span> phylo_id, pop)</span>
<span id="cb107-8"><a href="#cb107-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb107-9"><a href="#cb107-9" aria-hidden="true" tabindex="-1"></a>p_tree <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree) <span class="sc">%&lt;+%</span> nodes <span class="sc">+</span></span>
<span id="cb107-10"><a href="#cb107-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> pop, <span class="at">fill =</span> pop)) <span class="sc">+</span></span>
<span id="cb107-11"><a href="#cb107-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb107-12"><a href="#cb107-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">7e6</span>, <span class="fl">900e3</span>)) <span class="sc">+</span></span>
<span id="cb107-13"><a href="#cb107-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="fl">700e3</span>, <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"red"</span>)</span>
<span id="cb107-14"><a href="#cb107-14" aria-hidden="true" tabindex="-1"></a><span class="fu">revts</span>(p_tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-81-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="extracting-a-tree-1" class="level2">
<h2 class="anchored" data-anchor-id="extracting-a-tree-1">Extracting a tree #1</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">ts_phylo</span>(ts_small, <span class="dv">1</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb109"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb109-1"><a href="#cb109-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this needs another R package</span></span>
<span id="cb109-2"><a href="#cb109-2" aria-hidden="true" tabindex="-1"></a><span class="co"># simply using the ape package and running `plot(tree)` will also work</span></span>
<span id="cb109-3"><a href="#cb109-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb109-4"><a href="#cb109-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb109-5"><a href="#cb109-5" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> <span class="fu">ts_nodes</span>(tree) <span class="sc">%&gt;%</span></span>
<span id="cb109-6"><a href="#cb109-6" aria-hidden="true" tabindex="-1"></a>  as_tibble <span class="sc">%&gt;%</span></span>
<span id="cb109-7"><a href="#cb109-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">node =</span> phylo_id, pop)</span>
<span id="cb109-8"><a href="#cb109-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb109-9"><a href="#cb109-9" aria-hidden="true" tabindex="-1"></a>p_tree <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree) <span class="sc">%&lt;+%</span> nodes <span class="sc">+</span></span>
<span id="cb109-10"><a href="#cb109-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> pop, <span class="at">fill =</span> pop)) <span class="sc">+</span></span>
<span id="cb109-11"><a href="#cb109-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb109-12"><a href="#cb109-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">7e6</span>, <span class="fl">900e3</span>)) <span class="sc">+</span></span>
<span id="cb109-13"><a href="#cb109-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="fl">700e3</span>, <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"red"</span>)</span>
<span id="cb109-14"><a href="#cb109-14" aria-hidden="true" tabindex="-1"></a><span class="fu">revts</span>(p_tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-83-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="fragment">
<center>
<strong>Introgression?</strong>
</center>
</div>
</section>
<section id="extracting-a-tree-903" class="level2">
<h2 class="anchored" data-anchor-id="extracting-a-tree-903">Extracting a tree #903</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a>tree <span class="ot">&lt;-</span> <span class="fu">ts_phylo</span>(ts_small, <span class="dv">903</span>, <span class="at">quiet =</span> <span class="cn">TRUE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell" data-layout-align="center">
<details>
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb111"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb111-1"><a href="#cb111-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this needs another R package</span></span>
<span id="cb111-2"><a href="#cb111-2" aria-hidden="true" tabindex="-1"></a><span class="co"># simply using the ape package and running `plot(tree)` will also work</span></span>
<span id="cb111-3"><a href="#cb111-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb111-4"><a href="#cb111-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb111-5"><a href="#cb111-5" aria-hidden="true" tabindex="-1"></a>nodes <span class="ot">&lt;-</span> <span class="fu">ts_nodes</span>(tree) <span class="sc">%&gt;%</span></span>
<span id="cb111-6"><a href="#cb111-6" aria-hidden="true" tabindex="-1"></a>  as_tibble <span class="sc">%&gt;%</span></span>
<span id="cb111-7"><a href="#cb111-7" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">node =</span> phylo_id, pop)</span>
<span id="cb111-8"><a href="#cb111-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb111-9"><a href="#cb111-9" aria-hidden="true" tabindex="-1"></a>p_tree <span class="ot">&lt;-</span> <span class="fu">ggtree</span>(tree) <span class="sc">%&lt;+%</span> nodes <span class="sc">+</span></span>
<span id="cb111-10"><a href="#cb111-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_tiplab</span>(<span class="fu">aes</span>(<span class="at">color =</span> pop, <span class="at">fill =</span> pop)) <span class="sc">+</span></span>
<span id="cb111-11"><a href="#cb111-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">guides</span>(<span class="at">color =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb111-12"><a href="#cb111-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">7e6</span>, <span class="fl">900e3</span>)) <span class="sc">+</span></span>
<span id="cb111-13"><a href="#cb111-13" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_vline</span>(<span class="at">xintercept =</span> <span class="sc">-</span><span class="fl">700e3</span>, <span class="at">linetype =</span> <span class="dv">2</span>, <span class="at">color =</span> <span class="st">"red"</span>)</span>
<span id="cb111-14"><a href="#cb111-14" aria-hidden="true" tabindex="-1"></a><span class="fu">revts</span>(p_tree)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div class="quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="popgen2022_files/figure-html/unnamed-chunk-85-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<center>
<strong>Introgression?</strong>
</center>
</section>
</section>
<section id="bonus-exercise-2" class="level1">
<h1>Bonus exercise #2</h1>
<section id="try-to-implement-something-relevant-for-your-own-work" class="level2">
<h2 class="anchored" data-anchor-id="try-to-implement-something-relevant-for-your-own-work">Try to implement something relevant for your own work</h2>
<ul>
<li>Program a demographic model for your species of interest</li>
<li>Choose a statistic whose value “true value” you know from the literature (divergence? diversity? f4-statistic? ancestry proportion?)</li>
<li>Try to replicate the statistic in a <em>slendr</em> simulation using some of its <code>ts_&lt;...&gt;()</code> tree-sequence functions!</li>
</ul>
</section>
</section>
<section id="bonus-exercise-3" class="level1">
<h1>Bonus exercise #3</h1>
<section id="try-to-implement-your-own-d-statistic" class="level2">
<h2 class="anchored" data-anchor-id="try-to-implement-your-own-d-statistic">Try to implement your own D-statistic</h2>
<p>You can extract data from a tree sequence in the form of a genotype table using the function <code>ts_genotypes()</code>.</p>
<p>This function returns a simple R data frame:</p>
<ul>
<li>each column for an individual</li>
<li>each row for a site</li>
<li>each cell giving the number of derived alleles in that individual at that site</li>
</ul>
<p>Use the genotype table to compute your own ABBA/BABA statistic on the Neanderthal model data and check if you can recover the same signal as you get from <code>ts_f4()</code>.</p>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    target: function(trigger) {
      return trigger.previousElementSibling;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    setTimeout(function() {
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      let href = ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const cites = ref.parentNode.getAttribute('data-cites').split(' ');
    tippyHover(ref, function() {
      var popup = window.document.createElement('div');
      cites.forEach(function(cite) {
        var citeDiv = window.document.createElement('div');
        citeDiv.classList.add('hanging-indent');
        citeDiv.classList.add('csl-entry');
        var biblioDiv = window.document.getElementById('ref-' + cite);
        if (biblioDiv) {
          citeDiv.innerHTML = biblioDiv.innerHTML;
        }
        popup.appendChild(citeDiv);
      });
      return popup.innerHTML;
    });
  }
});
</script>
</div> <!-- /content -->



</body></html>